{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "49e8ec6c",
   "metadata": {},
   "source": [
    "Add col fix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fd24393-cd6e-4d4e-bb93-29b373448124",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T19:59:21.566736Z",
     "iopub.status.busy": "2025-01-29T19:59:21.566232Z",
     "iopub.status.idle": "2025-01-29T19:59:21.604104Z",
     "shell.execute_reply": "2025-01-29T19:59:21.603750Z",
     "shell.execute_reply.started": "2025-01-29T19:59:21.566717Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d859daee-f96e-4fd5-ae7e-3d82ffa71867",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T19:59:22.698244Z",
     "iopub.status.busy": "2025-01-29T19:59:22.697876Z",
     "iopub.status.idle": "2025-01-29T19:59:32.361175Z",
     "shell.execute_reply": "2025-01-29T19:59:32.360747Z",
     "shell.execute_reply.started": "2025-01-29T19:59:22.698221Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:09:47.819 EDT [4087263] LOG:  starting PostgreSQL 17.2 on x86_64-conda-linux-gnu, compiled by x86_64-conda-linux-gnu-cc (Anaconda gcc) 11.2.0, 64-bit\n",
      "2025-04-03 15:09:47.819 EDT [4087263] LOG:  listening on IPv6 address \"::1\", port 5433\n",
      "2025-04-03 15:09:47.819 EDT [4087263] LOG:  listening on IPv4 address \"127.0.0.1\", port 5433\n",
      "2025-04-03 15:09:47.822 EDT [4087263] LOG:  listening on Unix socket \"/tmp/.s.PGSQL.5433\"\n",
      "2025-04-03 15:09:47.880 EDT [4087272] LOG:  database system was shut down at 2025-04-03 11:57:39 EDT\n",
      "2025-04-03 15:09:47.927 EDT [4087263] LOG:  database system is ready to accept connections\n"
     ]
    }
   ],
   "source": [
    "#start the database\n",
    "process = subprocess.Popen([\"postgres\", \"-D\", \"/home/mcn26/palmer_scratch/db\", \"-p\", \"5433\"])\n",
    "\n",
    "# Load the SQL extension\n",
    "%load_ext sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a744e35a-d4cb-4e8c-8cdf-520aa0efaba9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T19:59:32.362208Z",
     "iopub.status.busy": "2025-01-29T19:59:32.361977Z",
     "iopub.status.idle": "2025-01-29T19:59:32.366982Z",
     "shell.execute_reply": "2025-01-29T19:59:32.366613Z",
     "shell.execute_reply.started": "2025-01-29T19:59:32.362193Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#dealing with a weird bug\n",
    "%config SqlMagic.style = '_DEPRECATED_DEFAULT'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e6fa421-abdf-4431-8d69-49cc9cd004c2",
   "metadata": {},
   "source": [
    "Connect to db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5fe01b0a-f945-4dfa-9fc4-86aebb3610a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:03.836969Z",
     "iopub.status.busy": "2025-01-29T20:00:03.836529Z",
     "iopub.status.idle": "2025-01-29T20:00:04.945641Z",
     "shell.execute_reply": "2025-01-29T20:00:04.945234Z",
     "shell.execute_reply.started": "2025-01-29T20:00:03.836948Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%sql\n",
    "postgresql://mr_root:password@localhost:5433/scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "893a9e79-3914-4739-91e6-ddb7f89e34ef",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:08.044686Z",
     "iopub.status.busy": "2025-01-29T20:00:08.044410Z",
     "iopub.status.idle": "2025-01-29T20:00:08.130074Z",
     "shell.execute_reply": "2025-01-29T20:00:08.129724Z",
     "shell.execute_reply.started": "2025-01-29T20:00:08.044669Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "1 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>table_schema</th>\n",
       "            <th>table_name</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>public</td>\n",
       "            <td>malin_gnomad</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('public', 'malin_gnomad')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT table_schema, table_name\n",
    "FROM information_schema.tables\n",
    "WHERE table_type = 'BASE TABLE' AND table_schema NOT IN ('pg_catalog', 'information_schema');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "64a794b3-d756-4fd7-b15a-0291b550bff7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:18.659282Z",
     "iopub.status.busy": "2025-01-29T20:00:18.658942Z",
     "iopub.status.idle": "2025-01-29T20:00:18.859471Z",
     "shell.execute_reply": "2025-01-29T20:00:18.859109Z",
     "shell.execute_reply.started": "2025-01-29T20:00:18.659265Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--DROP TABLE gwas_AFR;\n",
    "CREATE TABLE gwas_AFR (\n",
    "    rsID TEXT,\n",
    "    pubmedid INT,\n",
    "    study TEXT,\n",
    "    disease_trait TEXT,\n",
    "    p_value TEXT,--non-numeric to avoid precision errors & maintain formatting present in original\n",
    "    broad_ancestral_category TEXT,\n",
    "    ld_buddy TEXT,\n",
    "    r2 TEXT--Needs to be text because some entries are marked as lead SNP & have a text value here instead of numeric\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "52e0e75e-1f8a-4b4a-9cd5-f052712ef31f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:21.765036Z",
     "iopub.status.busy": "2025-01-29T20:00:21.764692Z",
     "iopub.status.idle": "2025-01-29T20:00:28.135627Z",
     "shell.execute_reply": "2025-01-29T20:00:28.135240Z",
     "shell.execute_reply.started": "2025-01-29T20:00:21.765014Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "1513213 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY gwas_AFR\n",
    "FROM PROGRAM 'zcat /home/mcn26/varef/data/GWAS/GWAS_LD_AFR_done.tsv.gz'\n",
    "NULL AS ''\n",
    "HEADER\n",
    "DELIMITER E'\\t';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23eb802e-209f-43b3-8651-6419a9aff479",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:28.136534Z",
     "iopub.status.busy": "2025-01-29T20:00:28.136380Z",
     "iopub.status.idle": "2025-01-29T20:00:28.225429Z",
     "shell.execute_reply": "2025-01-29T20:00:28.225085Z",
     "shell.execute_reply.started": "2025-01-29T20:00:28.136519Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--DROP TABLE gwas_ASN;\n",
    "--DROP TABLE gwas_EUR;\n",
    "CREATE TABLE gwas_ASN (LIKE gwas_AFR);\n",
    "CREATE TABLE gwas_EUR (LIKE gwas_AFR);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "841efcdc-5cfb-4811-9d3c-dcdc95803b3a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:00:28.425335Z",
     "iopub.status.busy": "2025-01-29T20:00:28.425062Z",
     "iopub.status.idle": "2025-01-29T20:01:07.878665Z",
     "shell.execute_reply": "2025-01-29T20:01:07.878221Z",
     "shell.execute_reply.started": "2025-01-29T20:00:28.425318Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:10:17.404 EDT [4087270] LOG:  checkpoint starting: wal\n",
      "2025-04-03 15:10:36.031 EDT [4087270] LOG:  checkpoint complete: wrote 2086 buffers (12.7%); 0 WAL file(s) added, 0 removed, 33 recycled; write=18.286 s, sync=0.011 s, total=18.671 s; sync files=36, longest=0.002 s, average=0.001 s; distance=543413 kB, estimate=543413 kB; lsn=23/8220B390, redo lsn=23/64418788\n",
      "2025-04-03 15:10:36.814 EDT [4087270] LOG:  checkpoints are occurring too frequently (19 seconds apart)\n",
      "2025-04-03 15:10:36.814 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:10:36.814 EDT [4087270] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5756158 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY gwas_ASN\n",
    "FROM PROGRAM 'zcat /home/mcn26/varef/data/GWAS/GWAS_LD_ASN_done.tsv.gz'\n",
    "NULL AS ''\n",
    "HEADER\n",
    "DELIMITER E'\\t';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c9e4f94f-6828-491c-9582-9ec364115db7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:01:07.879441Z",
     "iopub.status.busy": "2025-01-29T20:01:07.879304Z",
     "iopub.status.idle": "2025-01-29T20:02:33.209514Z",
     "shell.execute_reply": "2025-01-29T20:02:33.209140Z",
     "shell.execute_reply.started": "2025-01-29T20:01:07.879427Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:10:47.089 EDT [4087270] LOG:  checkpoint complete: wrote 27 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=9.839 s, sync=0.017 s, total=10.276 s; sync files=5, longest=0.014 s, average=0.004 s; distance=538989 kB, estimate=542971 kB; lsn=23/A2FF9E00, redo lsn=23/85273D08\n",
      "2025-04-03 15:10:47.893 EDT [4087270] LOG:  checkpoints are occurring too frequently (11 seconds apart)\n",
      "2025-04-03 15:10:47.893 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:10:47.893 EDT [4087270] LOG:  checkpoint starting: wal\n",
      "2025-04-03 15:10:59.158 EDT [4087270] LOG:  checkpoint complete: wrote 2030 buffers (12.4%); 0 WAL file(s) added, 0 removed, 33 recycled; write=10.685 s, sync=0.232 s, total=11.265 s; sync files=9, longest=0.226 s, average=0.026 s; distance=538789 kB, estimate=542553 kB; lsn=23/C3F786A8, redo lsn=23/A609D300\n",
      "2025-04-03 15:10:59.980 EDT [4087270] LOG:  checkpoints are occurring too frequently (12 seconds apart)\n",
      "2025-04-03 15:10:59.980 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:10:59.982 EDT [4087270] LOG:  checkpoint starting: wal\n",
      "2025-04-03 15:11:11.313 EDT [4087270] LOG:  checkpoint complete: wrote 27 buffers (0.2%); 0 WAL file(s) added, 1 removed, 32 recycled; write=10.561 s, sync=0.381 s, total=11.334 s; sync files=8, longest=0.253 s, average=0.048 s; distance=544265 kB, estimate=544265 kB; lsn=23/E62944B8, redo lsn=23/C741F8C0\n",
      "2025-04-03 15:11:11.824 EDT [4087270] LOG:  checkpoints are occurring too frequently (12 seconds apart)\n",
      "2025-04-03 15:11:11.824 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:11:11.827 EDT [4087270] LOG:  checkpoint starting: wal\n",
      "2025-04-03 15:11:22.742 EDT [4087270] LOG:  checkpoint complete: wrote 30 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=10.553 s, sync=0.061 s, total=10.918 s; sync files=4, longest=0.054 s, average=0.016 s; distance=540761 kB, estimate=543915 kB; lsn=24/6626738, redo lsn=23/E8435E40\n",
      "2025-04-03 15:11:23.491 EDT [4087270] LOG:  checkpoints are occurring too frequently (12 seconds apart)\n",
      "2025-04-03 15:11:23.491 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:11:23.493 EDT [4087270] LOG:  checkpoint starting: wal\n",
      "2025-04-03 15:11:34.426 EDT [4087270] LOG:  checkpoint complete: wrote 35 buffers (0.2%); 0 WAL file(s) added, 1 removed, 32 recycled; write=10.188 s, sync=0.447 s, total=10.935 s; sync files=6, longest=0.217 s, average=0.075 s; distance=540583 kB, estimate=543581 kB; lsn=24/28416A50, redo lsn=24/941FB58\n",
      "2025-04-03 15:11:34.779 EDT [4087270] LOG:  checkpoints are occurring too frequently (11 seconds apart)\n",
      "2025-04-03 15:11:34.779 EDT [4087270] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-04-03 15:11:34.781 EDT [4087270] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14127460 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY gwas_EUR\n",
    "FROM PROGRAM 'zcat /home/mcn26/varef/data/GWAS/GWAS_LD_EUR_done.tsv.gz'\n",
    "NULL AS ''\n",
    "HEADER\n",
    "DELIMITER E'\\t';"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5c344bf",
   "metadata": {},
   "source": [
    "Hrm. I think that the column headers are actually a little messed up..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b4be2fa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>broad_ancestral_category</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs2834083</td>\n",
       "            <td>0.9</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs28408622</td>\n",
       "            <td>0.88</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.91</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs9983755</td>\n",
       "            <td>0.71</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', 'rs1000005', '7E-12', 'rs1000005', 'tag'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', 'rs1000005', '7E-12', 'rs2834083', '0.9'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', 'rs1000005', '7E-12', 'rs28408622', '0.88'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', 'rs1000005', '7E-12', 'rs7275426', '0.91'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', 'rs1000005', '7E-12', 'rs9983755', '0.71')]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:13:24.332 EDT [4087270] LOG:  checkpoint complete: wrote 23 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=109.177 s, sync=0.170 s, total=109.554 s; sync files=13, longest=0.097 s, average=0.014 s; distance=540598 kB, estimate=543283 kB; lsn=24/3BF64E00, redo lsn=24/2A40D530\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_EUR \n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d22afcc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:15:19.011 EDT [4093209] ERROR:  canceling autovacuum task\n",
      "2025-04-03 15:15:19.011 EDT [4093209] CONTEXT:  while scanning block 268139 of relation \"public.gwas_eur\"\n",
      "\tautomatic vacuum of table \"scratch.public.gwas_eur\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE gwas_EUR\n",
    "    RENAME COLUMN rsid TO tag_snp;\n",
    "ALTER TABLE gwas_EUR\n",
    "    DROP COLUMN p_value;\n",
    "ALTER TABLE gwas_EUR\n",
    "    RENAME COLUMN broad_ancestral_category TO p_value;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "edf2f1ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>tag_snp</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs2834083</td>\n",
       "            <td>0.9</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs28408622</td>\n",
       "            <td>0.88</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.91</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>27863252</td>\n",
       "            <td>The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.</td>\n",
       "            <td>Eosinophil counts</td>\n",
       "            <td>7E-12</td>\n",
       "            <td>rs9983755</td>\n",
       "            <td>0.71</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', '7E-12', 'rs1000005', 'tag'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', '7E-12', 'rs2834083', '0.9'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', '7E-12', 'rs28408622', '0.88'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', '7E-12', 'rs7275426', '0.91'),\n",
       " ('rs1000005', 27863252, 'The Allelic Landscape of Human Blood Cell Trait Variation and Links to Common Complex Disease.', 'Eosinophil counts', '7E-12', 'rs9983755', '0.71')]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:16:34.656 EDT [4087270] LOG:  checkpoint starting: time\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_EUR \n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71c04292",
   "metadata": {},
   "source": [
    "Perfect. Let's check if the others need the same modification:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c9a4a598",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>broad_ancestral_category</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs2834083</td>\n",
       "            <td>0.83</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.96</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000282</td>\n",
       "            <td>34594039</td>\n",
       "            <td>A cross-population atlas of genetic associations for 220 human phenotypes.</td>\n",
       "            <td>Hematocrit</td>\n",
       "            <td>rs10000282</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>rs10000282</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000282</td>\n",
       "            <td>34594039</td>\n",
       "            <td>A cross-population atlas of genetic associations for 220 human phenotypes.</td>\n",
       "            <td>Hematocrit</td>\n",
       "            <td>rs10000282</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>rs10006400</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs1000005', 'tag'),\n",
       " ('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs2834083', '0.83'),\n",
       " ('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs7275426', '0.96'),\n",
       " ('rs10000282', 34594039, 'A cross-population atlas of genetic associations for 220 human phenotypes.', 'Hematocrit', 'rs10000282', '3E-8', 'rs10000282', 'tag'),\n",
       " ('rs10000282', 34594039, 'A cross-population atlas of genetic associations for 220 human phenotypes.', 'Hematocrit', 'rs10000282', '3E-8', 'rs10006400', '1')]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_ASN\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "48ebd582",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>broad_ancestral_category</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.83</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10009710</td>\n",
       "            <td>0.71</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs12640763</td>\n",
       "            <td>0.74</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs1000005', 'tag'),\n",
       " ('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', 'rs1000005', '5E-17', 'rs7275426', '0.83'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs10000225', 'tag'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs10009710', '0.71'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', 'rs10000225', '3E-6', 'rs12640763', '0.74')]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:21:04.684 EDT [4087270] LOG:  checkpoint complete: wrote 6370 buffers (38.9%); 0 WAL file(s) added, 0 removed, 20 recycled; write=269.395 s, sync=0.516 s, total=270.028 s; sync files=39, longest=0.298 s, average=0.014 s; distance=338256 kB, estimate=522780 kB; lsn=24/43055348, redo lsn=24/3EE618E0\n",
      "2025-04-03 15:21:34.929 EDT [4087270] LOG:  checkpoint starting: time\n",
      "2025-04-03 15:22:05.773 EDT [4087270] LOG:  checkpoint complete: wrote 51 buffers (0.3%); 0 WAL file(s) added, 1 removed, 4 recycled; write=30.570 s, sync=0.208 s, total=30.844 s; sync files=13, longest=0.123 s, average=0.016 s; distance=75184 kB, estimate=478021 kB; lsn=24/43F44070, redo lsn=24/437CDC90\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_AFR\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f138bed6",
   "metadata": {},
   "source": [
    "Yes, both need column-name correction..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3d0784bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE gwas_AFR\n",
    "    RENAME COLUMN rsid TO tag_snp;\n",
    "ALTER TABLE gwas_AFR\n",
    "    DROP COLUMN p_value;\n",
    "ALTER TABLE gwas_AFR\n",
    "    RENAME COLUMN broad_ancestral_category TO p_value;\n",
    "\n",
    "ALTER TABLE gwas_ASN\n",
    "    RENAME COLUMN rsid TO tag_snp;\n",
    "ALTER TABLE gwas_ASN\n",
    "    DROP COLUMN p_value;\n",
    "ALTER TABLE gwas_ASN\n",
    "    RENAME COLUMN broad_ancestral_category TO p_value;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e423824d-11ed-48c1-b74d-dddfb7781485",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-29T20:03:33.251792Z",
     "iopub.status.busy": "2025-01-29T20:03:33.251637Z",
     "iopub.status.idle": "2025-01-29T20:03:35.480155Z",
     "shell.execute_reply": "2025-01-29T20:03:35.479623Z",
     "shell.execute_reply.started": "2025-01-29T20:03:33.251778Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "waiting for server to shut down...."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-03 15:24:35.453 EDT [4087263] LOG:  received fast shutdown request\n",
      "2025-04-03 15:24:35.454 EDT [4087263] LOG:  aborting any active transactions\n",
      "2025-04-03 15:24:35.454 EDT [4087292] FATAL:  terminating connection due to administrator command\n",
      "2025-04-03 15:24:35.462 EDT [4087263] LOG:  background worker \"logical replication launcher\" (PID 4087275) exited with exit code 1\n",
      "2025-04-03 15:24:35.462 EDT [4091664] FATAL:  terminating autovacuum process due to administrator command\n",
      "2025-04-03 15:24:35.462 EDT [4091664] CONTEXT:  while scanning block 2712875 of relation \"public.malin_gnomad\"\n",
      "2025-04-03 15:24:36.277 EDT [4087270] LOG:  shutting down\n",
      "2025-04-03 15:24:36.279 EDT [4087270] LOG:  checkpoint starting: shutdown immediate\n",
      "2025-04-03 15:24:36.366 EDT [4087270] LOG:  checkpoint complete: wrote 287 buffers (1.8%); 0 WAL file(s) added, 3 removed, 0 recycled; write=0.059 s, sync=0.003 s, total=0.089 s; sync files=12, longest=0.002 s, average=0.001 s; distance=45402 kB, estimate=434759 kB; lsn=24/46424808, redo lsn=24/46424808\n",
      "2025-04-03 15:24:36.393 EDT [4087263] LOG:  database system is shut down\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " done\n",
      "server stopped\n"
     ]
    }
   ],
   "source": [
    "#stop the sql db\n",
    "!pg_ctl -D ~/palmer_scratch/db stop"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
