---
title: "Analyses of purifying selection in MPAC-gnomAD"
author: "Stephen Rong"
date: "2024-02-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r preamble}
# load libraries
library(tidyverse)
library(data.table)
library(ggthemes)
library(viridis)
library(scales)
library(wrapr)
library(tidyplots)

# load tables
gnomAD_main_table_full_cCREs <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs.rds")
gnomAD_main_table_full_cCREs_in_TF <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_in_TF.rds")
gnomAD_main_table_full_cCREs_in_rep <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_in_rep.rds")
gnomAD_main_table_full_emVar_cat <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_emVar_cat.rds")
gnomAD_main_table_full_cCREs_emVar_cat <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_emVar_cat.rds")
gnomAD_main_table_full_cCREs_pleio <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_pleio.rds")
gnomAD_main_table_full_cCREs_pleio_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_pleio_mean.rds")
gnomAD_main_table_full_CADD <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_CADD.rds")
gnomAD_main_table_full_cCREs_CADD <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_CADD.rds")
gnomAD_main_table_full_cCREs_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean.rds")
gnomAD_main_table_full_cCREs_mean_in_TF <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean_in_TF.rds")
gnomAD_main_table_full_cCREs_mean_in_rep <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean_in_rep.rds")
gnomAD_main_table_full_cCREs_K562 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562.rds")
gnomAD_main_table_full_cCREs_K562_in_TF <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_in_TF.rds")
gnomAD_main_table_full_cCREs_K562_in_rep <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_in_rep.rds")
gnomAD_main_table_full_cCREs_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2.rds")
gnomAD_main_table_full_cCREs_HepG2_in_TF <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_in_TF.rds")
gnomAD_main_table_full_cCREs_HepG2_in_rep <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_in_rep.rds")
gnomAD_main_table_full_cCREs_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH.rds")
gnomAD_main_table_full_cCREs_SKNSH_in_TF <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH_in_TF.rds")
gnomAD_main_table_full_cCREs_SKNSH_in_rep <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH_in_rep.rds")
gnomAD_main_table_full_cCREs_mean_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean_ref.rds")
gnomAD_main_table_full_cCREs_K562_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_ref.rds")
gnomAD_main_table_full_cCREs_HepG2_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_ref.rds")
gnomAD_main_table_full_cCREs_SKNSH_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH_ref.rds")
gnomAD_main_table_full_in_TF_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_mean.rds")
gnomAD_main_table_full_in_TF_K562 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_K562.rds")
gnomAD_main_table_full_in_TF_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_HepG2.rds")
gnomAD_main_table_full_in_TF_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_SKNSH.rds")
gnomAD_main_table_full_in_TF_mean_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_mean_ref.rds")
gnomAD_main_table_full_in_TF_K562_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_K562_ref.rds")
gnomAD_main_table_full_in_TF_HepG2_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_HepG2_ref.rds")
gnomAD_main_table_full_in_TF_SKNSH_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_TF_SKNSH_ref.rds")
gnomAD_main_table_full_in_rep_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_mean.rds")
gnomAD_main_table_full_in_rep_K562 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_K562.rds")
gnomAD_main_table_full_in_rep_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_HepG2.rds")
gnomAD_main_table_full_in_rep_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_SKNSH.rds")
gnomAD_main_table_full_in_rep_mean_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_mean_ref.rds")
gnomAD_main_table_full_in_rep_K562_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_K562_ref.rds")
gnomAD_main_table_full_in_rep_HepG2_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_HepG2_ref.rds")
gnomAD_main_table_full_in_rep_SKNSH_ref <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_in_rep_SKNSH_ref.rds")
gnomAD_main_table_full_cCREs_category <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_category.rds")
gnomAD_main_table_full_cCREs_mean_category <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean_category.rds")
gnomAD_main_table_full_cCREs_K562_category <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_category.rds")
gnomAD_main_table_full_cCREs_HepG2_category <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_category.rds")
gnomAD_main_table_full_cCREs_SKNSH_category <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH_category.rds")
gnomAD_main_table_full_cCREs_mean_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_mean_mean.rds")
gnomAD_main_table_full_cCREs_K562_K562 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_K562.rds")
gnomAD_main_table_full_cCREs_HepG2_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_HepG2.rds")
gnomAD_main_table_full_cCREs_SKNSH_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_SKNSH_SKNSH.rds")
gnomAD_main_table_full_cCREs_emVar_cat_mean <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_emVar_cat_mean.rds")
gnomAD_main_table_full_cCREs_emVar_cat_K562 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_emVar_cat_K562.rds")
gnomAD_main_table_full_cCREs_emVar_cat_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_emVar_cat_HepG2.rds")
gnomAD_main_table_full_cCREs_emVar_cat_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_emVar_cat_SKNSH.rds")
gnomAD_main_table_full_cCREs_K562_HepG2 <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_HepG2.rds")
gnomAD_main_table_full_cCREs_K562_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_K562_SKNSH.rds")
gnomAD_main_table_full_cCREs_HepG2_SKNSH <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table_full_cCREs_HepG2_SKNSH.rds")
```

```{r overall basic statistics}
gnomAD_main_table <- readRDS("../../results/gnomAD_selection_preprocess/gnomAD_main_table.rds")

# how many emVars
gnomAD_main_table_summ_emVar <- gnomAD_main_table %>% 
	group_by(emVar_category) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_emVar %>% filter(emVar_category != "none") %>% 
  tidyplot(x = emVar_category, y = count, color = emVar_category) %>% 
  add_barstack_absolute() %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3) +
  xlab("emVar category") + 
  ylab("Count") + 
  ylim(c(0, 3500000))
ggsave("../../results/gnomAD_selection_analysis/how_many_emVar.pdf")

# how many emVars per cCRE
gnomAD_main_table_summ_emVar_cCRE <- gnomAD_main_table %>% 
  mutate(emVar = ifelse(emVar_category == "none", FALSE, TRUE)) %>% 
	group_by(emVar, ENCODE_cCREs) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_emVar_cCRE %>% 
  filter(ENCODE_cCREs != "Other cCREs", emVar) %>% 
  tidyplot(x = ENCODE_cCREs, y = count, color = emVar) %>% 
  add_mean_bar() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, 
            position = position_dodge(.9), show.legend = FALSE) +
  xlab("ENCODE cCREs") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_emVar_cCRE.pdf")

# how many allelic skew
gnomAD_main_table_summ_K562_skew <- gnomAD_main_table %>% 
	group_by(K562_skew_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_HepG2_skew <- gnomAD_main_table %>% 
	group_by(HepG2_skew_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_SKNSH_skew <- gnomAD_main_table %>% 
	group_by(SKNSH_skew_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_mean_skew <- gnomAD_main_table %>% 
	group_by(mean_skew_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_K562_skew %>% 
  tidyplot(x = K562_skew_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#00A79D") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#00A79D") +
  xlab("K562 allelic skew") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_K562_skew.pdf")

gnomAD_main_table_summ_HepG2_skew %>% 
  tidyplot(x = HepG2_skew_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#FBB040") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#FBB040") +
  xlab("HepG2 allelic skew") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_HepG2_skew.pdf")
  
gnomAD_main_table_summ_SKNSH_skew %>% 
  tidyplot(x = SKNSH_skew_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#ED1C24") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#ED1C24") +
  xlab("SKNSH allelic skew") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_SKNSH_skew.pdf")

gnomAD_main_table_summ_mean_skew %>% 
  tidyplot(x = mean_skew_pred_bin, y = count) %>% 
  add_barstack_absolute() %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3) +
  xlab("Mean allelic skew") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_mean_skew.pdf")

nonsig_skew <- c("[-0.5, -0.2)", "[-0.2, -0.05)", "[-0.05, 0.05)", "[0.05, 0.2)", "[0.2, 0.5)")
a <- sum(filter(gnomAD_main_table, (K562_skew_pred_bin %in% nonsig_skew) & (HepG2_skew_pred_bin %in% nonsig_skew) & (SKNSH_skew_pred_bin %in% nonsig_skew)
)$count)
b <- sum(filter(gnomAD_main_table, !((K562_skew_pred_bin %in% nonsig_skew) & (HepG2_skew_pred_bin %in% nonsig_skew) & (SKNSH_skew_pred_bin %in% nonsig_skew)
))$count)
a
b
b/(a+b)

# how many ref active
gnomAD_main_table_summ_K562_ref <- gnomAD_main_table %>% 
	group_by(K562_ref_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_HepG2_ref <- gnomAD_main_table %>% 
	group_by(HepG2_ref_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_SKNSH_ref <- gnomAD_main_table %>% 
	group_by(SKNSH_ref_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_mean_ref <- gnomAD_main_table %>% 
	group_by(mean_ref_pred_bin) %>% 
	summarise(count = sum(count)) %>% 
	ungroup() %>% 
	mutate(prop = count/sum(count))

gnomAD_main_table_summ_K562_ref %>% 
  tidyplot(x = K562_ref_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#00A79D") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#00A79D") +
  xlab("K562 ref activity") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_K562_ref.pdf")

gnomAD_main_table_summ_HepG2_ref %>% 
  tidyplot(x = HepG2_ref_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#FBB040") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#FBB040") +
  xlab("HepG2 ref activity") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_HepG2_ref.pdf")
  
gnomAD_main_table_summ_SKNSH_ref %>% 
  tidyplot(x = SKNSH_ref_pred_bin, y = count) %>% 
  add_barstack_absolute(fill = "#ED1C24") %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, color = "#ED1C24") +
  xlab("SKNSH ref activity") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_SKNSH_ref.pdf")

gnomAD_main_table_summ_mean_ref %>% 
  tidyplot(x = mean_ref_pred_bin, y = count) %>% 
  add_barstack_absolute() %>% 
  remove_legend() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3) +
  xlab("Mean ref activity") + 
  ylab("Count")
ggsave("../../results/gnomAD_selection_analysis/how_many_mean_ref.pdf")

nonsig_ref <- c("[-Inf, 1)")
a <- sum(filter(gnomAD_main_table, (K562_ref_pred_bin %in% nonsig_ref) & (HepG2_ref_pred_bin %in% nonsig_ref) & (SKNSH_ref_pred_bin %in% nonsig_ref)
)$count)
b <- sum(filter(gnomAD_main_table, !((K562_ref_pred_bin %in% nonsig_ref) & (HepG2_ref_pred_bin %in% nonsig_ref) & (SKNSH_ref_pred_bin %in% nonsig_ref)
))$count)
a
b
b/(a+b)
```

```{r get proportion of emVars}
emVar_bins <- c(" (-Inf, -1.5)", "[-1.5, -1.0)", "[-1.0, -0.5)", "[0.5, 1.0)", "[1.0, 1.5)", "[1.5, Inf)")

sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "PLS") %>% filter((K562_skew_pred_bin %in% emVar_bins) | (HepG2_skew_pred_bin %in% emVar_bins) | (SKNSH_skew_pred_bin %in% emVar_bins)) %>% .$count)/
  sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "PLS") %>% .$count)

sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "pELS") %>% filter((K562_skew_pred_bin %in% emVar_bins) | (HepG2_skew_pred_bin %in% emVar_bins) | (SKNSH_skew_pred_bin %in% emVar_bins)) %>% .$count)/
  sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "pELS") %>% .$count)

sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "dELS") %>% filter((K562_skew_pred_bin %in% emVar_bins) | (HepG2_skew_pred_bin %in% emVar_bins) | (SKNSH_skew_pred_bin %in% emVar_bins)) %>% .$count)/
  sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "dELS") %>% .$count)

sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "non-cCRE") %>% filter((K562_skew_pred_bin %in% emVar_bins) | (HepG2_skew_pred_bin %in% emVar_bins) | (SKNSH_skew_pred_bin %in% emVar_bins)) %>% .$count)/
  sum(gnomAD_main_table %>% filter(ENCODE_cCREs == "non-cCRE") %>% .$count)
```

```{r visualize cons explained}
gnomAD_main_table_full_cCREs_emVar_cat_phyloP <- gnomAD_main_table_full_cCREs_emVar_cat %>% 
  mutate(phyloP_cons = count*phyloP_cons_prop, phyloP_noncons = count*(1-phyloP_cons_prop)) %>% 
  dplyr::select(ENCODE_cCREs, emVar_category, phyloP_cons, phyloP_noncons) %>% 
  pivot_longer(cols = c(phyloP_cons, phyloP_noncons), names_to = "phyloP", values_to = "count") %>% 
  mutate(phyloP = gsub("phyloP_", "", phyloP)) %>% 
  mutate(ENCODE_cCREs_phyloP = paste(ENCODE_cCREs, phyloP, sep="_")) %>% 
  group_by(ENCODE_cCREs_phyloP, ENCODE_cCREs, phyloP) %>% 
  mutate(count_prop = count/sum(count), base = sum(count)) %>% 
  ungroup() %>% 
  filter(emVar_category != "none") %>% 
  mutate(ENCODE_cCREs_phyloP = factor(ENCODE_cCREs_phyloP, levels = c("PLS_noncons", "PLS_cons", "pELS_noncons", "pELS_cons", "dELS_noncons", "dELS_cons", "non-cCRE_noncons", "non-cCRE_cons"))) %>% 
  mutate(emVar_category = factor(emVar_category, levels = c("K562", "HepG2", "SKNSH", "K562+HepG2", "K562+SKNSH", "SKNSH+HepG2", "K562+SKNSH+HepG2")))

write_tsv(gnomAD_main_table_full_cCREs_emVar_cat_phyloP, "../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_cat_phyloP_stacked.txt")

gnomAD_main_table_full_cCREs_emVar_cat_phyloP %>% 
	group_by(ENCODE_cCREs, phyloP, base) %>% 
	summarise(
	 count = sum(count), 
	 count_prop = sum(count_prop)
	) %>% 
	ungroup()

gnomAD_main_table_full_cCREs_emVar_phyloP <- gnomAD_main_table_full_cCREs_emVar_cat %>% 
  mutate(phyloP_cons = count*phyloP_cons_prop, phyloP_noncons = count*(1-phyloP_cons_prop)) %>% 
  dplyr::select(ENCODE_cCREs, emVar_category, phyloP_cons, phyloP_noncons) %>% 
  pivot_longer(cols = c(phyloP_cons, phyloP_noncons), names_to = "phyloP", values_to = "count") %>% 
  mutate(phyloP = gsub("phyloP_", "", phyloP)) %>% 
  group_by(phyloP) %>% 
  mutate(count_prop = count/sum(count), base = sum(count)) %>% 
  ungroup() %>% 
  filter(emVar_category != "none") %>% 
  mutate(emVar_category = factor(emVar_category, levels = c("K562", "HepG2", "SKNSH", "K562+HepG2", "K562+SKNSH", "SKNSH+HepG2", "K562+SKNSH+HepG2")))

write_tsv(gnomAD_main_table_full_cCREs_emVar_phyloP, "../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_stacked.txt")

gnomAD_main_table_full_cCREs_emVar_phyloP %>% 
	group_by(phyloP, base) %>% 
	summarise(
	 count = sum(count), 
	 count_prop = sum(count_prop)
	) %>% 
	ungroup()

gnomAD_main_table_full_cCREs_emVar_cat_phyloP_cCRE <- gnomAD_main_table_full_cCREs_emVar_cat_phyloP %>% 
	group_by(ENCODE_cCREs, phyloP, base) %>% 
	summarise(
	 count = sum(count), 
	 count_prop = sum(count_prop)
	) %>% 
	ungroup()

gnomAD_main_table_full_cCREs_emVar_phyloP_gw <- gnomAD_main_table_full_cCREs_emVar_phyloP %>%
	group_by(phyloP, base) %>%
	summarise(
	 count = sum(count),
	 count_prop = sum(count_prop)
	) %>%
	ungroup() %>%
  mutate(ENCODE_cCREs = "Genome-wide")

gnomAD_main_table_full_cCREs_emVar_phyloP_prop <- bind_rows(
  gnomAD_main_table_full_cCREs_emVar_phyloP_gw,
  gnomAD_main_table_full_cCREs_emVar_cat_phyloP_cCRE
) 
write_tsv(gnomAD_main_table_full_cCREs_emVar_phyloP_prop, "../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_prop.txt")

gnomAD_main_table_full_cCREs_emVar_phyloP_prop %>% 
  mutate(ENCODE_cCREs = factor(ENCODE_cCREs, levels = c("PLS", "pELS", "dELS", "non-cCRE", "Genome-wide"))) %>% 
  mutate(phyloP = ifelse(phyloP == "cons", TRUE, FALSE)) %>% 
  filter(ENCODE_cCREs != "Other cCREs") %>% 
  tidyplot(x = ENCODE_cCREs, y = count_prop, color = phyloP) %>% 
  add_mean_bar() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  xlab("ENCODE cCREs") + 
  ylab("emVar proportion") + 
  labs(fill = "Constrained") + 
  theme(aspect.ratio = 1.6, strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_prop.pdf", scale = 0.7)

gnomAD_main_table_full_cCREs_emVar_phyloP_prop %>% 
  mutate(ENCODE_cCREs = factor(ENCODE_cCREs, levels = c("PLS", "pELS", "dELS", "non-cCRE", "Genome-wide"))) %>% 
  mutate(phyloP = ifelse(phyloP == "cons", TRUE, FALSE)) %>% 
  filter(ENCODE_cCREs != "Other cCREs") %>% 
  tidyplot(x = ENCODE_cCREs, y = count_prop, color = phyloP) %>% 
  add_mean_bar() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  xlab("ENCODE cCREs") + 
  ylab("emVar proportion") + 
  labs(fill = "Constrained") + 
  theme(aspect.ratio = 1/1.01, strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_prop_alt.pdf")

gnomAD_main_table_full_cCREs_emVar_phyloP_prop %>% 
  mutate(ENCODE_cCREs = factor(ENCODE_cCREs, levels = c("PLS", "pELS", "dELS", "non-cCRE", "Genome-wide"))) %>% 
  mutate(phyloP = ifelse(phyloP == "cons", TRUE, FALSE)) %>% 
  filter(ENCODE_cCREs != "Other cCREs") %>% 
  tidyplot(x = ENCODE_cCREs, y = count, color = phyloP) %>% 
  add_mean_bar() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, 
            position = position_dodge(.9), show.legend = FALSE) +
  xlab("ENCODE cCREs") + 
  ylab("emVar count") + 
  labs(fill = "Constrained") + 
    theme(aspect.ratio = 1.6, strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_count.pdf", scale = 0.7)

gnomAD_main_table_full_cCREs_emVar_phyloP_prop %>% 
  mutate(ENCODE_cCREs = factor(ENCODE_cCREs, levels = c("PLS", "pELS", "dELS", "non-cCRE", "Genome-wide"))) %>% 
  mutate(phyloP = ifelse(phyloP == "cons", TRUE, FALSE)) %>% 
  filter(ENCODE_cCREs != "Other cCREs") %>% 
  tidyplot(x = ENCODE_cCREs, y = count, color = phyloP) %>% 
  add_mean_bar() %>% 
  adjust_x_axis(rotate_labels = TRUE) + 
  geom_text(aes(label = count), angle = 90, hjust = -0.1, size = 3, 
            position = position_dodge(.9), show.legend = FALSE) +
  xlab("ENCODE cCREs") + 
  ylab("emVar count") + 
  labs(fill = "Constrained") + 
    theme(aspect.ratio = 1/1.01, strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_phyloP_count_alt.pdf")
```

```{r visualize counts}
ggplot(gnomAD_main_table_full_cCREs, 
	 aes(x="", y=count, fill=ENCODE_cCREs)) +	#	%>% filter(ENCODE_cCREs != "Other cCREs")) +
	geom_col() + 
	geom_text(aes(label = paste0(round(count/1e6, 1), "M")),
	 position = position_stack(vjust = 0.5)) +
	coord_polar("y", start=0) +
	theme_void() + 
	scale_fill_manual(values = c(PLS="#3E0751", pELS="#405187", dELS="#468E8B", `non-cCRE`="#79C66E", `Other cCREs`="grey50")) + 
	labs(fill = "ENCODE cCREs")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_count.pdf", scale=0.6)
```

```{r visualize distributions}
ggplot(gnomAD_main_table_full_cCREs_mean_ref %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_bar(aes(x = mean_ref_pred_bin, y = count, fill = ENCODE_cCREs), stat = "identity", position = "stack") + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank(), axis.line.y = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "Mean ref activity", y = "Variants", color = "ENCODE cCREs") + 
	facet_grid(rows = vars(ENCODE_cCREs), scales = "free") + 
	theme(aspect.ratio = 1/(1.6*4), strip.background = element_blank()) + 
	guides(fill = "none")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_ref_count.pdf", scale=0.5)

ggplot(gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_bar(aes(x = mean_skew_pred_bin, y = count, fill = ENCODE_cCREs), stat = "identity", position = "stack") + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank(), axis.line.y = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "Mean allelic skew", y = "Variants", color = "ENCODE cCREs") + 
	facet_grid(rows = vars(ENCODE_cCREs), scales = "free") +
	theme(aspect.ratio = 1/(1.6*4), strip.background = element_blank()) + 
	guides(fill = "none")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_skew_count.pdf", scale=0.6)

gnomAD_main_table_full_cCREs_emVar_cat <- gnomAD_main_table_full_cCREs_emVar_cat %>% 
	group_by(ENCODE_cCREs) %>% 
	mutate(count_prop = count/sum(count)) %>% 
	ungroup()

ggplot(gnomAD_main_table_full_cCREs_emVar_cat %>% filter(ENCODE_cCREs != "Other cCREs", emVar_category != "none")) + 
	geom_bar(aes(x = emVar_category, y = count_prop, fill = ENCODE_cCREs), stat = "identity", position = "stack") + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank(), axis.line.y = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "emVars (MPAC)", y = "Proportion", color = "ENCODE cCREs") + 
	facet_grid(rows = vars(ENCODE_cCREs), scales = "free") +
	theme(aspect.ratio = 1/(1.6*4), strip.background = element_blank()) + 
	guides(fill = "none")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_emVar_cat_prop.pdf", scale=0.75)

gnomAD_cCREs_emVar_cat_ratio_shared_vs_specific <- gnomAD_main_table_full_cCREs_emVar_cat %>% filter(ENCODE_cCREs != "Other cCREs", emVar_category != "none") %>% 
  dplyr::select(ENCODE_cCREs, emVar_category, count) %>% 
  pivot_wider(names_from = emVar_category, values_from = count) %>% 
  mutate(shared_vs_specific = `K562+SKNSH+HepG2`/(`K562` + `HepG2` + `SKNSH`))

gnomAD_cCREs_emVar_cat_ratio_shared_vs_specific
```

```{r tf footprints plots}
gnomAD_main_table_full_TF_footprints <- read_tsv("../../results/gnomAD_selection_analysis/gnomAD_TF_footprints_mean_allelic_skew.txt") %>% 
  mutate(mean_skew_pred_bin = factor(mean_skew_pred_bin, levels = 
    c("(-Inf, -1.5)", "[-1.5, -1.0)", "[-1.0, -0.5)", "[-0.5, -0.2)", 
    "[-0.2, -0.05)", "[-0.05, 0.05)", "[0.05, 0.2)", "[0.2, 0.5)", 
    "[0.5, 1.0)", "[1.0, 1.5)", "[1.5, Inf)"))
  ) %>% 
  group_by(mean_skew_pred_bin) %>%
  mutate(proportion = count/sum(count)) %>% 
  ungroup()

ggplot(gnomAD_main_table_full_TF_footprints) + 
  geom_bar(aes(x = mean_skew_pred_bin, y = proportion, fill = TF_footprints), stat = "identity", position = "fill") + 
  geom_text(aes(x = mean_skew_pred_bin, 
                y = proportion + 0.15,
                label = ifelse(TF_footprints, sprintf("%.3f", proportion), "")), 
            size = 3, angle = 90, na.rm = TRUE, color = "white") + 
  cowplot::theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
  labs(x = "Mean allelic skew", y = "Proportion", fill = "In TF ChIP-seq") + 
  theme(aspect.ratio = 1/1.6, strip.background = element_blank()) + 
  ylim(c(0,1)) + 
  guides(fill = "none")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_footprints_stack.pdf", scale=0.7)

gnomAD_main_table_full_TF_footprints_gw <- gnomAD_main_table_full_TF_footprints %>% 
  dplyr::select(TF_footprints, mean_skew_pred_bin, count) %>% 
  group_by(TF_footprints) %>% 
  summarise(count = sum(count)) %>% 
  ungroup() %>% 
  pivot_wider(names_prefix = "TF_footprints_", names_from = TF_footprints, values_from = count)

gnomAD_main_table_full_TF_footprints_temp <- gnomAD_main_table_full_TF_footprints %>% 
  dplyr::select(TF_footprints, mean_skew_pred_bin, count) %>% 
  pivot_wider(names_prefix = "TF_footprints_", names_from = TF_footprints, values_from = count)

gnomAD_main_table_full_TF_footprints_fet <- gnomAD_main_table_full_TF_footprints_temp %>% 
  rowwise() %>% 
  mutate(fet_TF_footprints = list(fisher.test(matrix(c(TF_footprints_TRUE+1, TF_footprints_FALSE+1, gnomAD_main_table_full_TF_footprints_gw[["TF_footprints_TRUE"]]-TF_footprints_TRUE+1, gnomAD_main_table_full_TF_footprints_gw[["TF_footprints_FALSE"]]-TF_footprints_FALSE+1), nrow=2)))) %>%
  mutate(fet_TF_footprints_pval = fet_TF_footprints$p.val) %>%
  mutate(fet_TF_footprints_odds = fet_TF_footprints$estimate) %>%
  mutate(fet_TF_footprints_lci = fet_TF_footprints$conf.int[1]) %>%
  mutate(fet_TF_footprints_uci = fet_TF_footprints$conf.int[2]) %>%
  ungroup() %>% 
  mutate(fet_TF_footprints_padj = p.adjust(fet_TF_footprints_pval, method = "fdr")) %>% 
  mutate(fet_TF_footprints_sig = ifelse(fet_TF_footprints_pval < 0.05, "*", ""))

ggplot(gnomAD_main_table_full_TF_footprints_fet %>% filter()) + 
  geom_bar(aes(x = mean_skew_pred_bin, y = (fet_TF_footprints_odds)), stat = "identity", fill = "#56BCC2") + 
  geom_errorbar(aes(x = mean_skew_pred_bin, ymin = (fet_TF_footprints_lci), ymax = (fet_TF_footprints_uci)), stat = "identity", width = 0.2) + 
  geom_text(aes(x = mean_skew_pred_bin, y = (fet_TF_footprints_uci) + 0.5, label = sprintf("%.2f", fet_TF_footprints_odds)), angle = 90, vjust = 0.5, hjust = 0) + 
  geom_hline(yintercept = 1, color = "black", linetype = "dashed", alpha = 0.5) + 
  cowplot::theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
  labs(x = "Mean allelic skew", y = "Odds ratio") + 
  theme(aspect.ratio = 1/1.6, strip.background = element_blank()) + 
  ylim(c(0, 11))
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_footprints_fet.pdf", scale=0.7)

write_tsv(gnomAD_main_table_full_TF_footprints, "../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_footprints_fet.txt")
```

```{r tf chip-seq plots}
gnomAD_main_table_full_TF_ChIP_seq <- read_tsv("../../results/gnomAD_selection_analysis/gnomAD_TF_ChIP_seq_mean_allelic_skew.txt") %>% 
  mutate(mean_skew_pred_bin = factor(mean_skew_pred_bin, levels = 
    c("(-Inf, -1.5)", "[-1.5, -1.0)", "[-1.0, -0.5)", "[-0.5, -0.2)", 
    "[-0.2, -0.05)", "[-0.05, 0.05)", "[0.05, 0.2)", "[0.2, 0.5)", 
    "[0.5, 1.0)", "[1.0, 1.5)", "[1.5, Inf)"))
  ) %>% 
  group_by(mean_skew_pred_bin) %>%
  mutate(proportion = count/sum(count)) %>% 
  ungroup()

ggplot(gnomAD_main_table_full_TF_ChIP_seq) + 
  geom_bar(aes(x = mean_skew_pred_bin, y = proportion, fill = TF_ChIP_seq), stat = "identity", position = "fill") + 
  geom_text(aes(x = mean_skew_pred_bin, 
                y = proportion + 0.15,
                label = ifelse(TF_ChIP_seq, sprintf("%.3f", proportion), "")), 
            size = 3, angle = 90, na.rm = TRUE, color = "white") + 
  cowplot::theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
  labs(x = "Mean allelic skew", y = "Proportion", fill = "In TF ChIP-seq") + 
  theme(aspect.ratio = 1/1.6, strip.background = element_blank()) + 
  ylim(c(0,1)) + 
  guides(fill = "none")
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_ChIP_seq_stack.pdf", scale=0.7)

gnomAD_main_table_full_TF_ChIP_seq_gw <- gnomAD_main_table_full_TF_ChIP_seq %>% 
  dplyr::select(TF_ChIP_seq, mean_skew_pred_bin, count) %>% 
  group_by(TF_ChIP_seq) %>% 
  summarise(count = sum(count)) %>% 
  ungroup() %>% 
  pivot_wider(names_prefix = "TF_ChIP_seq_", names_from = TF_ChIP_seq, values_from = count)

gnomAD_main_table_full_TF_ChIP_seq_temp <- gnomAD_main_table_full_TF_ChIP_seq %>% 
  dplyr::select(TF_ChIP_seq, mean_skew_pred_bin, count) %>% 
  pivot_wider(names_prefix = "TF_ChIP_seq_", names_from = TF_ChIP_seq, values_from = count)

gnomAD_main_table_full_TF_ChIP_seq_fet <- gnomAD_main_table_full_TF_ChIP_seq_temp %>% 
  rowwise() %>% 
  mutate(fet_TF_ChIP_seq = list(fisher.test(matrix(c(TF_ChIP_seq_TRUE+1, TF_ChIP_seq_FALSE+1, gnomAD_main_table_full_TF_ChIP_seq_gw[["TF_ChIP_seq_TRUE"]]-TF_ChIP_seq_TRUE+1, gnomAD_main_table_full_TF_ChIP_seq_gw[["TF_ChIP_seq_FALSE"]]-TF_ChIP_seq_FALSE+1), nrow=2)))) %>%
  mutate(fet_TF_ChIP_seq_pval = fet_TF_ChIP_seq$p.val) %>%
  mutate(fet_TF_ChIP_seq_odds = fet_TF_ChIP_seq$estimate) %>%
  mutate(fet_TF_ChIP_seq_lci = fet_TF_ChIP_seq$conf.int[1]) %>%
  mutate(fet_TF_ChIP_seq_uci = fet_TF_ChIP_seq$conf.int[2]) %>%
  ungroup() %>% 
  mutate(fet_TF_ChIP_seq_padj = p.adjust(fet_TF_ChIP_seq_pval, method = "fdr")) %>% 
  mutate(fet_TF_ChIP_seq_sig = ifelse(fet_TF_ChIP_seq_pval < 0.05, "*", ""))

ggplot(gnomAD_main_table_full_TF_ChIP_seq_fet %>% filter()) + 
  geom_bar(aes(x = mean_skew_pred_bin, y = (fet_TF_ChIP_seq_odds)), stat = "identity", fill = "#56BCC2") + 
  geom_errorbar(aes(x = mean_skew_pred_bin, ymin = (fet_TF_ChIP_seq_lci), ymax = (fet_TF_ChIP_seq_uci)), stat = "identity", width = 0.2) + 
  geom_text(aes(x = mean_skew_pred_bin, y = (fet_TF_ChIP_seq_uci) + 0.5, label = sprintf("%.2f", fet_TF_ChIP_seq_odds)), angle = 90, vjust = 0.5, hjust = 0) + 
  geom_hline(yintercept = 1, color = "black", linetype = "dashed", alpha = 0.5) + 
  cowplot::theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
  labs(x = "Mean allelic skew", y = "Odds ratio") + 
  theme(aspect.ratio = 1/1.6, strip.background = element_blank()) + 
  ylim(c(0, 11))
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_ChIP_seq_fet.pdf", scale=0.7)

write_tsv(gnomAD_main_table_full_TF_ChIP_seq, "../../results/gnomAD_selection_analysis/gnomAD_main_table_full_TF_ChIP_seq_fet.txt")
```

```{r visualize ENCODE cCREs pleio}
ggplot(gnomAD_main_table_full_cCREs_pleio %>% filter(ENCODE_cCREs != "Other cCREs") %>% mutate(pleio = as.factor(pleio))) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_bar(aes(pleio, phyloP_mean, fill = pleio), stat = "identity") + 
	geom_errorbar(aes(pleio, ymin = phyloP_lci, ymax = phyloP_uci), stat = "identity", width = 0.2) + 
	cowplot::theme_cowplot() + 
	theme(axis.line.x = element_blank()) + 
	scale_fill_viridis_d() + 
	scale_color_viridis_d() + 
	labs(x = "emVar pleiotropy", y = "Mean constraint (phyloP)") +
	guides(fill = "none") + 
	facet_wrap(~ENCODE_cCREs) + 
	theme(aspect.ratio = 1, strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_pleio_phyloP.pdf", scale=1)
```

```{r visualize phyloP ref}
ggplot(gnomAD_main_table_full_cCREs_mean_ref %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_ribbon(aes(x = mean_ref_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
	geom_point(aes(x = mean_ref_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
	geom_errorbar(aes(x = mean_ref_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) +   
	scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "Mean ref activitiy", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
	guides(fill = "none") + 
	theme(aspect.ratio = 1/0.99)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_ref_phyloP.pdf")
```

```{r visualize phyloP}
ggplot(gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_ribbon(aes(x = mean_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
	geom_point(aes(x = mean_skew_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
	geom_errorbar(aes(x = mean_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) +   
	scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "Mean allelic skew", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
	guides(fill = "none") + 
	theme(aspect.ratio = 1/0.99)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_phyloP.pdf")

ggplot(gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs != "Other cCREs")) + 
  geom_hline(yintercept = 0, color = "black") + 
  geom_ribbon(aes(x = mean_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
  geom_point(aes(x = mean_skew_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
  geom_errorbar(aes(x = mean_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
  cowplot::theme_cowplot() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
  scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
  scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
  labs(x = "Mean allelic skew", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
  guides(fill = "none") + 
  theme(aspect.ratio = 3/2)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_phyloP_skinny.pdf", scale = 1)

ggplot(gnomAD_main_table_full_cCREs_K562 %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_ribbon(aes(x = K562_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
	geom_point(aes(x = K562_skew_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
	geom_errorbar(aes(x = K562_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) +   
	scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "K562 allelic skew", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
	guides(fill = "none") + 
	theme(aspect.ratio = 1/0.99)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_K562_phyloP.pdf")

ggplot(gnomAD_main_table_full_cCREs_HepG2 %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_ribbon(aes(x = HepG2_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
	geom_point(aes(x = HepG2_skew_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
	geom_errorbar(aes(x = HepG2_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) +   
	scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "HepG2 allelic skew", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
	guides(fill = "none") + 
	theme(aspect.ratio = 1/0.99)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_HepG2_phyloP.pdf")

ggplot(gnomAD_main_table_full_cCREs_SKNSH %>% filter(ENCODE_cCREs != "Other cCREs")) + 
	geom_hline(yintercept = 0, color = "black") + 
	geom_ribbon(aes(x = SKNSH_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, fill = ENCODE_cCREs, group = ENCODE_cCREs), position = position_dodge(width = 0), alpha = 0.5) +
	geom_point(aes(x = SKNSH_skew_pred_bin, y = phyloP_mean, color = ENCODE_cCREs), position = position_dodge(width = 0)) + 
	geom_errorbar(aes(x = SKNSH_skew_pred_bin, ymin = phyloP_lci, ymax = phyloP_uci, color = ENCODE_cCREs), width = 0.5, position = position_dodge(width = 0)) + 
	cowplot::theme_cowplot() + 
	theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), axis.line.x = element_blank()) + 
	scale_fill_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) +   
	scale_color_manual(values = c("#3E0751", "#405187", "#468E8B", "#79C66E", "grey50")) + 
	labs(x = "SKNSH allelic skew", y = "Mean constraint (phyloP)", color = "ENCODE cCREs") + 
	guides(fill = "none") + 
	theme(aspect.ratio = 1/0.99)
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_SKNSH_phyloP.pdf")
```

```{r do as heatmap instead}
ggplot(gnomAD_main_table_full_cCREs_mean_category %>% filter(ENCODE_cCREs != "Other cCREs")) +
  geom_tile(aes(x = mean_skew_pred_bin, y = category_copy, fill = phyloP_mean)) +
  geom_text(aes(x = mean_skew_pred_bin, y = category_copy, 
    label = round(phyloP_mean, 2)),
    color = "white", size = 2) +  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
    axis.line = element_blank()) + 
  scale_fill_viridis_c() + 
  labs(x = "Mean allelic skew", y = "ENCODE cCREs", fill = "Mean constraint (phyloP)") +
  facet_grid(rows = vars(ENCODE_cCREs)) +
  coord_fixed() +  # Ensures square tiles
  theme(strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_mean_category_count_heat.pdf")

ggplot(gnomAD_main_table_full_cCREs_category %>% filter(ENCODE_cCREs != "Other cCREs")) +
  geom_tile(aes(x = 1, y = category_copy, fill = phyloP_mean)) +
  geom_text(aes(x = 1, y = category_copy, 
    label = round(phyloP_mean, 2)),
    color = "white", size = 2) +  
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
    axis.line = element_blank()) + 
  scale_fill_viridis_c(limits = c(
    min(gnomAD_main_table_full_cCREs_mean_category$phyloP_mean, na.rm = T), 
    max(gnomAD_main_table_full_cCREs_mean_category$phyloP_mean, na.rm = T))) + 
  labs(x = "", y = "ENCODE cCREs", fill = "Mean constraint (phyloP)") +
  facet_grid(rows = vars(ENCODE_cCREs)) +
  coord_fixed() +  # Ensures square tiles
  theme(strip.background = element_blank())
ggsave("../../results/gnomAD_selection_analysis/gnomAD_main_table_full_cCREs_category_count_heat.pdf")

gnomAD_main_table_full_rare_common_phyloP_mean <- gnomAD_main_table_full_cCREs_category %>% 
  mutate(category_copy_temp = ifelse(category_copy %in% c("SINGLETON", "ULTRARARE", "RARE"), "RARE", "COMMON")) %>% 
  group_by(category_copy_temp) %>% 
  summarise(
    sum_count = sum(count), 
    sum_phyloP = sum(count * phyloP_mean)
  ) %>% 
  ungroup()

filter(gnomAD_main_table_full_rare_common_phyloP_mean, category_copy_temp == "RARE")$sum_phyloP/
  filter(gnomAD_main_table_full_rare_common_phyloP_mean, category_copy_temp == "RARE")$sum_count

filter(gnomAD_main_table_full_rare_common_phyloP_mean, category_copy_temp == "COMMON")$sum_phyloP/
  filter(gnomAD_main_table_full_rare_common_phyloP_mean, category_copy_temp == "COMMON")$sum_count
```

```{r get mean and confidence interval values}
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "PLS", mean_skew_pred_bin == "(-Inf, -1.5)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, fet_RARE_COMMON_odds, fet_RARE_COMMON_lci, fet_RARE_COMMON_uci)
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "PLS", mean_skew_pred_bin == "[1.0, 1.5)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, fet_RARE_COMMON_odds, fet_RARE_COMMON_lci, fet_RARE_COMMON_uci)
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "PLS", mean_skew_pred_bin == "[1.5, Inf)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, fet_RARE_COMMON_odds, fet_RARE_COMMON_lci, fet_RARE_COMMON_uci)
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "PLS", mean_skew_pred_bin == "(-Inf, -1.5)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "pELS", mean_skew_pred_bin == "(-Inf, -1.5)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "dELS", mean_skew_pred_bin == "(-Inf, -1.5)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "PLS", mean_skew_pred_bin == "[1.5, Inf)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "pELS", mean_skew_pred_bin == "[1.5, Inf)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
gnomAD_main_table_full_cCREs_mean %>% filter(ENCODE_cCREs == "dELS", mean_skew_pred_bin == "[1.5, Inf)") %>% dplyr::select(ENCODE_cCREs, mean_skew_pred_bin, phyloP_mean, phyloP_lci, phyloP_uci) 
```
