These scripts were written by me, Mackenzie Noon (mackenzie.noon@yale.edu) & ChatGPT. Contact me (Mackenzie, that is) with any questions.

## Overview

The goal of this analysis is to .... 

The basic flow:

\[input data\] -> 0. merge -> 1. annotate 2. filter -> 3. count -> 4. graph

(TODO: explain all this once the logic is semi-finalized...)
(this should be changed to a graph instead of a linear flow)

Additionally, manual verification with screenshots...

Additionally, "intuition" ...


## How to reproduce 

- In this directory, a yml file is provided which should allow you to make a conda environment that will facilitate recapitulation most results.
- You can see "conda activate mcn_varef" in the batch files used to submit the relevant 
- A few shell scripts use specific versions of SAMtools & BEDtools, as documented in ...
- These scripts uses file generated by other analysis...
    - Uses gnomad VCF annotated with malinois activity predictions ...
    - gnomad from ...
    - malinous activity predictions produced in ...
    - The filtering step used masking produced in ...

Tange, O. (2021, March 22). GNU Parallel 20210322 ('2002-01-06').
  Zenodo. https://doi.org/10.5281/zenodo.4628277
  
  
- according to ...

TODO: fill all this out once completed...

Worflow also includes some shell code, and so uses awk, wc, other similar utilities. 



### note on encoding in count tables
sometimes presence of variants within a bin is stored in a one-hot fashion : multiple true/false columns, one for each bin. 
This is not generally desierable, so I convert away with "find_true_column"

other times, this is desierable as it is not actually one hot, but multi-hot. Such as presence in genomic region (i.e. some variants might be both in a promoter like sequence & some other cata)

### note on sex chromosomes
gnomad does not include the Y chromosome, so naturally it isn't included ...
to avoid complications, the X chromosome isnt included either. 
- It *is* processed for some of the steps, but the exon file "filter" 

...

AF, AC & AN

(from tooltips on gnomad website)

AC : Allele Count : "Alternate allele count in high-quality genotypes"
AN : Allele Number : "Total high quality genotypes"
AF : Allele Frequency : "Alternate Allele Frequency in high quality genotypes"

AF=AC/AN

I define : MAF (minor allele frequency) = min(AF,1-AF)

gnomad filters 1) filter = pass, 2) maf != 0, 3) AN >= 76,156
- 76,156 is [the number of high-quality genomes in 3.1](https://gnomad.broadinstitute.org/news/2020-10-gnomad-v3-1/)
- Therefore, if a site were called in all individuals it would be 76,156\*2 . Gnomad sometimes issues the following warning

> Warning This variant is covered in fewer than 50% of individuals in gnomAD v3.1.2 genomes. This may indicate a low-quality site.

Example : [rs113653250](https://gnomad.broadinstitute.org/variant/1-434284-T-G?dataset=gnomad_r3)

...


- 

We're using gnomad v3.1.2. 

activity ref OR alt:
(-Inf,1), [1,2), [2,4), [4,6), [6,Inf) (note first bin we would call as not active)
- I am worried that this will lose directional effects. I will make separate sets of bins for REF and ALT. 

allelic skew:
(-Inf, -1.5), [-1.5, 1), [-1, -0.5), [-0.5, 0), [0, 0.5), [0.5, 1), [1, 1.5), [1.5, Inf) (note middle two bins we would not call as emvars)

...

Todo
- [x] Fix genotype number cutoff
- [x] annotated directly from GRCh38-cCREs.V4.bed.gz, make all types
    - [x] Get list of unique 
- [x] chr22 : Re-run annotate on chr 22 with fixed filtering parameters
    - [x] check if slurm can still write to output files in branches (it can't, let's just not track them)
- [~] crunch my way through the rest of the chromosomes
  - [x] rewrite to split by chromosome prior to annotation step. Then run that...
- [x] manually check annotations in IGV. 
- [x] filter for only intronic & intergenic vars prior to subsequent analysis
- [x] regraph
    - [x] modify to put . instead of - in the graphs
    - [x] fix thresholds
- [ ] ~~Add 2x2 : in/out of bin (bin=each bar=combo of skew and ref) vs rare/ common. Lets us put error bars~~
- [ ] Rong: add pseudocounts of 1 to CADD scores (for both rare/common) to distinguish between 0 & inf
- [ ] Conservation of as a function of pleitropy 
- [ ] don't average skews : use individual cell-line skews & group by "skew in 1 cell line" "skew in 2 cell lines..." (how exactly to do latter..?)
  - [ ] Alternative : no bins, just subsample & do quant vs quant for stat test. 
  - [ ] alt: cumulative distro & KS test

- [ ] malinous graphs for all regions : not restricted to promoter, enhancer ... 

- [ ] Add graph of purifying selection as a function of phylop value, numerically instead of thresholded. 

- [ ] modify all graphs to look nice & export to svg or someone

- [ ] move NN E4L9NdjPku1yFLxlZoJtXQ6TaMbewMye to better location

- [ ] move NN BL2jL5SIMor1yx6B21DpQ1jm4oiF82tt to better location

- [ ] If you send a copy of your published article to tange@gnu.org, it will be mentioned in the release notes of next version of GNU Parallel.

0. merge 
Adds malinouis predictions to datasets

1. annotate

2. count
Warning: cre_bool_columns is hardcoded

(interpretations)
CADD

>To simplify interpretation in some contexts, we also defined phred-like scores (scaled C  scores) on the basis of the rank of the C score of each variant relative  to all 8.6 billion possible SNVs, ranging from 1 to 99 (Supplementary  Note). For example, substitutions with the highest 10% (10−1) of all  scores—that is, those least likely to be observed human alleles under  our model—were assigned values of 10 or greater (‘≥C10’), whereas  variants  in  the  highest  1%  (10^−2),  0.1%  (10^−3),  etc.  were  assigned  scores ‘≥C20’, ‘≥C30’, etc

(A general framework for estimating the relative pathogenicity of human genetic variants, nature genetics, 2014)

PhyloP
- 

ensembl VEP
- 

Encode
https://screen.encodeproject.org/

>1. cCREs with promoter-like signatures (cCRE-PLS) fall within 200 bp (center to center) of an annotated GENCODE TSS and have high DNase and H3K4me3 signals (evaluated as DNase and H3K4me3 max-Z scores, defined as the maximal DNase or H3K4me3 Z scores across all biosamples with data; see Methods).
>2. cCREs with enhancer-like signatures (cCRE-ELS) have high DNase and H3K27ac max-Z scores and must additionally have a low H3K4me3 max-Z score if they are within 200 bp of an annotated TSS. The subset of cCREs-ELS within 2 kb of a TSS is denoted proximal (cCRE-pELS), while the remaining subset is denoted distal (cCRE-dELS).
>3. DNase-H3K4me3 cCREs have high H3K4me3 max-Z scores but low H3K27ac max-Z scores and do not fall within 200 bp of a TSS.
>4. CTCF-only cCREs have high DNase and CTCF max-Z scores and low H3K4me3 and H3K27ac max-Z scores. 
