{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:09:18.658 EST [4146906] LOG:  starting PostgreSQL 17.2 on x86_64-conda-linux-gnu, compiled by x86_64-conda-linux-gnu-cc (Anaconda gcc) 11.2.0, 64-bit\n",
      "2025-02-10 20:09:18.658 EST [4146906] LOG:  listening on IPv6 address \"::1\", port 5433\n",
      "2025-02-10 20:09:18.658 EST [4146906] LOG:  listening on IPv4 address \"127.0.0.1\", port 5433\n",
      "2025-02-10 20:09:18.661 EST [4146906] LOG:  listening on Unix socket \"/tmp/.s.PGSQL.5433\"\n",
      "2025-02-10 20:09:18.780 EST [4146911] LOG:  database system was shut down at 2025-02-10 18:37:31 EST\n",
      "2025-02-10 20:09:18.860 EST [4146906] LOG:  database system is ready to accept connections\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import requests\n",
    "#start the database\n",
    "process = subprocess.Popen([\"postgres\", \"-D\", \"/home/mcn26/palmer_scratch/db\", \"-p\", \"5433\"])\n",
    "# Load the SQL extension\n",
    "%load_ext sql\n",
    "#dealing with a weird bug\n",
    "%config SqlMagic.style = '_DEPRECATED_DEFAULT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "postgresql://mr_root:password@localhost:5433/scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--DROP TABLE AFR_broken;\n",
    "CREATE TABLE AFR_broken (\n",
    "    rsID TEXT,\n",
    "    pubmedid INT,\n",
    "    study TEXT,\n",
    "    disease_trait TEXT,\n",
    "    p_value TEXT,--non-numeric to avoid precision errors & maintain formatting present in original\n",
    "    ld_buddy TEXT,\n",
    "    r2 TEXT,--Needs to be text because some entries are marked as lead SNP & have a text value here instead of numeric\n",
    "    chromosome_id TEXT,\n",
    "    pos INT,\n",
    "    ref TEXT,\n",
    "    alt TEXT,\n",
    "    K562__ref FLOAT,\n",
    "    HepG2__ref FLOAT,\n",
    "    SKNSH__ref FLOAT,\n",
    "    K562__alt FLOAT,\n",
    "    HepG2__alt FLOAT,\n",
    "    SKNSH__alt FLOAT,\n",
    "    K562__skew FLOAT,\n",
    "    HepG2__skew FLOAT,\n",
    "    SKNSH__skew FLOAT\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--DROP TABLE ASN_broken;\n",
    "CREATE TABLE ASN_broken (LIKE AFR_broken);\n",
    "--DROP TABLE EUR_broken;\n",
    "CREATE TABLE EUR_broken (LIKE AFR_broken);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:49:19.651 EST [4146909] LOG:  checkpoint starting: time\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1262917 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY AFR_broken\n",
    "FROM '/home/mcn26/varef/data/GWAS_malin/AFR_broken.tsv' \n",
    "WITH (FORMAT csv, DELIMITER E'\\t', HEADER);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:49:51.703 EST [4146909] LOG:  checkpoint complete: wrote 52 buffers (0.3%); 0 WAL file(s) added, 0 removed, 0 recycled; write=31.849 s, sync=0.053 s, total=32.052 s; sync files=36, longest=0.012 s, average=0.002 s; distance=7780 kB, estimate=7780 kB; lsn=8F/29DA1B8, redo lsn=8E/E4BACEC8\n",
      "2025-02-10 20:49:53.149 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:50:09.703 EST [4146909] LOG:  checkpoint complete: wrote 2049 buffers (12.5%); 0 WAL file(s) added, 0 removed, 33 recycled; write=15.865 s, sync=0.017 s, total=16.560 s; sync files=6, longest=0.009 s, average=0.003 s; distance=532852 kB, estimate=532852 kB; lsn=8F/2322ED98, redo lsn=8F/5409F58\n",
      "2025-02-10 20:50:10.623 EST [4146909] LOG:  checkpoints are occurring too frequently (17 seconds apart)\n",
      "2025-02-10 20:50:10.623 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:50:10.623 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:50:28.893 EST [4146909] LOG:  checkpoint complete: wrote 27 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=17.538 s, sync=0.122 s, total=18.270 s; sync files=5, longest=0.078 s, average=0.025 s; distance=536631 kB, estimate=536631 kB; lsn=8F/43E00DD8, redo lsn=8F/26017D50\n",
      "2025-02-10 20:50:30.549 EST [4146909] LOG:  checkpoints are occurring too frequently (20 seconds apart)\n",
      "2025-02-10 20:50:30.549 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:50:30.552 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:50:49.747 EST [4146909] LOG:  checkpoint complete: wrote 25 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=18.604 s, sync=0.021 s, total=19.199 s; sync files=9, longest=0.010 s, average=0.003 s; distance=544691 kB, estimate=544691 kB; lsn=8F/6509D5E0, redo lsn=8F/47404A48\n",
      "2025-02-10 20:50:51.135 EST [4146909] LOG:  checkpoints are occurring too frequently (21 seconds apart)\n",
      "2025-02-10 20:50:51.135 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:50:51.139 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:51:09.933 EST [4146909] LOG:  checkpoint complete: wrote 2063 buffers (12.6%); 0 WAL file(s) added, 0 removed, 33 recycled; write=18.076 s, sync=0.121 s, total=18.798 s; sync files=15, longest=0.068 s, average=0.009 s; distance=540714 kB, estimate=544293 kB; lsn=8F/864466F8, redo lsn=8F/6840F608\n",
      "2025-02-10 20:51:11.145 EST [4146909] LOG:  checkpoints are occurring too frequently (20 seconds apart)\n",
      "2025-02-10 20:51:11.145 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:51:11.154 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:51:30.480 EST [4146909] LOG:  checkpoint complete: wrote 33 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=18.767 s, sync=0.030 s, total=19.335 s; sync files=3, longest=0.012 s, average=0.010 s; distance=538696 kB, estimate=543733 kB; lsn=8F/A6DD3AC8, redo lsn=8F/89221698\n",
      "2025-02-10 20:51:32.111 EST [4146909] LOG:  checkpoints are occurring too frequently (21 seconds apart)\n",
      "2025-02-10 20:51:32.111 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:51:32.114 EST [4146909] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11702163 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY EUR_broken\n",
    "FROM '/home/mcn26/varef/data/GWAS_malin/EUR_broken.tsv' \n",
    "WITH (FORMAT csv, DELIMITER E'\\t', HEADER);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:51:49.140 EST [4146909] LOG:  checkpoint complete: wrote 27 buffers (0.2%); 0 WAL file(s) added, 0 removed, 33 recycled; write=16.409 s, sync=0.018 s, total=17.030 s; sync files=5, longest=0.014 s, average=0.004 s; distance=542669 kB, estimate=543627 kB; lsn=8F/C81ED538, redo lsn=8F/AA414D20\n",
      "2025-02-10 20:51:50.561 EST [4146909] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 20:51:50.561 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:51:50.561 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:52:08.309 EST [4146909] LOG:  checkpoint complete: wrote 2010 buffers (12.3%); 0 WAL file(s) added, 0 removed, 33 recycled; write=17.099 s, sync=0.012 s, total=17.749 s; sync files=7, longest=0.006 s, average=0.002 s; distance=538059 kB, estimate=543070 kB; lsn=8F/E8EBA750, redo lsn=8F/CB187C48\n",
      "2025-02-10 20:52:09.500 EST [4146909] LOG:  checkpoints are occurring too frequently (19 seconds apart)\n",
      "2025-02-10 20:52:09.500 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:52:09.509 EST [4146909] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4804689 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "COPY ASN_broken\n",
    "FROM '/home/mcn26/varef/data/GWAS_malin/ASN_broken.tsv' \n",
    "WITH (FORMAT csv, DELIMITER E'\\t', HEADER);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:53:24.536 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:53:34.563 EST [4146909] LOG:  checkpoint complete: wrote 2069 buffers (12.6%); 0 WAL file(s) added, 0 removed, 33 recycled; write=9.325 s, sync=0.037 s, total=10.031 s; sync files=33, longest=0.018 s, average=0.002 s; distance=543235 kB, estimate=543235 kB; lsn=90/2B47C370, redo lsn=90/D398E08\n",
      "2025-02-10 20:53:34.983 EST [4146909] LOG:  checkpoints are occurring too frequently (10 seconds apart)\n",
      "2025-02-10 20:53:34.983 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:53:34.986 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:53:45.119 EST [4146909] LOG:  checkpoint complete: wrote 11 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=9.420 s, sync=0.225 s, total=10.137 s; sync files=6, longest=0.098 s, average=0.038 s; distance=541242 kB, estimate=543036 kB; lsn=90/4C49CAC0, redo lsn=90/2E427660\n",
      "2025-02-10 20:53:46.133 EST [4146909] LOG:  checkpoints are occurring too frequently (12 seconds apart)\n",
      "2025-02-10 20:53:46.133 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:53:46.137 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:54:02.658 EST [4146909] LOG:  checkpoint complete: wrote 8 buffers (0.0%); 0 WAL file(s) added, 0 removed, 33 recycled; write=15.896 s, sync=0.127 s, total=16.526 s; sync files=6, longest=0.065 s, average=0.022 s; distance=540796 kB, estimate=542812 kB; lsn=90/6D6BCB00, redo lsn=90/4F4466A0\n",
      "2025-02-10 20:54:03.922 EST [4146909] LOG:  checkpoints are occurring too frequently (17 seconds apart)\n",
      "2025-02-10 20:54:03.922 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:54:03.925 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:54:20.231 EST [4146909] LOG:  checkpoint complete: wrote 15 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=15.609 s, sync=0.165 s, total=16.310 s; sync files=5, longest=0.087 s, average=0.033 s; distance=539588 kB, estimate=542490 kB; lsn=90/8E3ADA80, redo lsn=90/70337A30\n",
      "2025-02-10 20:54:21.175 EST [4146909] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 20:54:21.175 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:54:21.178 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:54:32.330 EST [4146909] LOG:  checkpoint complete: wrote 17 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=10.518 s, sync=0.121 s, total=11.155 s; sync files=6, longest=0.072 s, average=0.021 s; distance=541231 kB, estimate=542364 kB; lsn=90/AF2F7520, redo lsn=90/913C37D0\n",
      "2025-02-10 20:54:32.981 EST [4146909] LOG:  checkpoints are occurring too frequently (11 seconds apart)\n",
      "2025-02-10 20:54:32.981 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:54:33.009 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:54:43.895 EST [4146909] LOG:  checkpoint complete: wrote 20 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=10.344 s, sync=0.076 s, total=10.914 s; sync files=5, longest=0.038 s, average=0.016 s; distance=541554 kB, estimate=542283 kB; lsn=90/D041B828, redo lsn=90/B24A0340\n",
      "2025-02-10 20:54:44.549 EST [4146909] LOG:  checkpoints are occurring too frequently (12 seconds apart)\n",
      "2025-02-10 20:54:44.549 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:54:44.553 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:55:00.118 EST [4146909] LOG:  checkpoint complete: wrote 16 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=14.643 s, sync=0.278 s, total=15.569 s; sync files=8, longest=0.162 s, average=0.035 s; distance=540272 kB, estimate=542082 kB; lsn=90/F150AAA8, redo lsn=90/D343C5B8\n",
      "2025-02-10 20:55:01.146 EST [4146909] LOG:  checkpoints are occurring too frequently (17 seconds apart)\n",
      "2025-02-10 20:55:01.146 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:55:01.151 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:55:19.086 EST [4146909] LOG:  checkpoint complete: wrote 5 buffers (0.0%); 0 WAL file(s) added, 0 removed, 33 recycled; write=16.972 s, sync=0.173 s, total=17.940 s; sync files=6, longest=0.082 s, average=0.029 s; distance=538460 kB, estimate=541720 kB; lsn=91/123C1B20, redo lsn=90/F4213888\n",
      "2025-02-10 20:55:19.933 EST [4146909] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 20:55:19.933 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:55:19.936 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:55:35.563 EST [4146909] LOG:  checkpoint complete: wrote 4011 buffers (24.5%); 0 WAL file(s) added, 0 removed, 33 recycled; write=15.070 s, sync=0.011 s, total=15.630 s; sync files=15, longest=0.008 s, average=0.001 s; distance=542578 kB, estimate=542578 kB; lsn=91/330998C0, redo lsn=91/153F00B8\n",
      "2025-02-10 20:55:36.612 EST [4146909] LOG:  checkpoints are occurring too frequently (17 seconds apart)\n",
      "2025-02-10 20:55:36.612 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:55:36.612 EST [4146909] LOG:  checkpoint starting: wal\n",
      "2025-02-10 20:55:52.823 EST [4146909] LOG:  checkpoint complete: wrote 15 buffers (0.1%); 0 WAL file(s) added, 0 removed, 33 recycled; write=15.660 s, sync=0.019 s, total=16.212 s; sync files=10, longest=0.013 s, average=0.002 s; distance=540486 kB, estimate=542368 kB; lsn=91/53FFAD18, redo lsn=91/363C1A20\n",
      "2025-02-10 20:55:54.172 EST [4146909] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 20:55:54.172 EST [4146909] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 20:55:54.176 EST [4146909] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE TABLE stacked_table AS\n",
    "SELECT 'AFR' AS source_table, * FROM AFR_broken\n",
    "UNION ALL\n",
    "SELECT 'EUR' AS source_table, * FROM EUR_broken\n",
    "UNION ALL\n",
    "SELECT 'ASN' AS source_table, * FROM ASN_broken;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>pos</th>\n",
       "            <th>ref</th>\n",
       "            <th>alt</th>\n",
       "            <th>k562__ref</th>\n",
       "            <th>hepg2__ref</th>\n",
       "            <th>sknsh__ref</th>\n",
       "            <th>k562__alt</th>\n",
       "            <th>hepg2__alt</th>\n",
       "            <th>sknsh__alt</th>\n",
       "            <th>k562__skew</th>\n",
       "            <th>hepg2__skew</th>\n",
       "            <th>sknsh__skew</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.83</td>\n",
       "            <td>chr21</td>\n",
       "            <td>33060745</td>\n",
       "            <td>G</td>\n",
       "            <td>C</td>\n",
       "            <td>0.4658757</td>\n",
       "            <td>0.5130802</td>\n",
       "            <td>0.39136338</td>\n",
       "            <td>0.4494594</td>\n",
       "            <td>0.51362103</td>\n",
       "            <td>0.34879774</td>\n",
       "            <td>-0.016416281</td>\n",
       "            <td>0.0005408189</td>\n",
       "            <td>-0.042565674</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr21</td>\n",
       "            <td>33060745</td>\n",
       "            <td>G</td>\n",
       "            <td>C</td>\n",
       "            <td>0.4658757</td>\n",
       "            <td>0.5130802</td>\n",
       "            <td>0.39136338</td>\n",
       "            <td>0.4494594</td>\n",
       "            <td>0.51362103</td>\n",
       "            <td>0.34879774</td>\n",
       "            <td>-0.016416281</td>\n",
       "            <td>0.0005408189</td>\n",
       "            <td>-0.042565674</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10009710</td>\n",
       "            <td>0.71</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs12640763</td>\n",
       "            <td>0.74</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs1505771</td>\n",
       "            <td>0.98</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 'rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs7275426', '0.83', 'chr21', 33060745, 'G', 'C', 0.4658757, 0.5130802, 0.39136338, 0.4494594, 0.51362103, 0.34879774, -0.016416281, 0.0005408189, -0.042565674),\n",
       " ('AFR', 'rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs1000005', 'tag', 'chr21', 33060745, 'G', 'C', 0.4658757, 0.5130802, 0.39136338, 0.4494594, 0.51362103, 0.34879774, -0.016416281, 0.0005408189, -0.042565674),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs10009710', '0.71', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs12640763', '0.74', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs1505771', '0.98', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599)]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM stacked_table\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DONE\n",
      "waiting for server to shut down...."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:56:37.866 EST [4146906] LOG:  received fast shutdown request\n",
      "2025-02-10 20:56:37.872 EST [4146906] LOG:  aborting any active transactions\n",
      "2025-02-10 20:56:37.872 EST [4147044] FATAL:  terminating connection due to administrator command\n",
      "2025-02-10 20:56:37.887 EST [4146906] LOG:  background worker \"logical replication launcher\" (PID 4146914) exited with exit code 1\n",
      "2025-02-10 20:56:37.888 EST [4168962] FATAL:  terminating autovacuum process due to administrator command\n",
      "2025-02-10 20:56:37.888 EST [4168962] CONTEXT:  while scanning block 69442 of relation \"public.stacked_table\"\n",
      "2025-02-10 20:56:38.479 EST [4146909] LOG:  checkpoint complete: wrote 7 buffers (0.0%); 0 WAL file(s) added, 0 removed, 33 recycled; write=43.838 s, sync=0.002 s, total=44.307 s; sync files=4, longest=0.002 s, average=0.001 s; distance=537849 kB, estimate=541916 kB; lsn=91/5A66B028, redo lsn=91/570FFFB0\n",
      "2025-02-10 20:56:38.487 EST [4146909] LOG:  shutting down\n",
      "2025-02-10 20:56:38.493 EST [4146909] LOG:  checkpoint starting: shutdown immediate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " done\n",
      "server stopped\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:56:38.751 EST [4146909] LOG:  checkpoint complete: wrote 2280 buffers (13.9%); 0 WAL file(s) added, 0 removed, 3 recycled; write=0.145 s, sync=0.038 s, total=0.265 s; sync files=7, longest=0.023 s, average=0.006 s; distance=54700 kB, estimate=493195 kB; lsn=91/5A66B0A0, redo lsn=91/5A66B0A0\n",
      "2025-02-10 20:56:38.789 EST [4146906] LOG:  database system is shut down\n"
     ]
    }
   ],
   "source": [
    "print(\"DONE\")\n",
    "!pg_ctl -D ~/palmer_scratch/db stop"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mcn_varef",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
