{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c21df2e2-9320-4f58-8c47-ad7380b6c855",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T21:21:32.037780Z",
     "iopub.status.busy": "2025-01-07T21:21:32.037358Z",
     "iopub.status.idle": "2025-01-07T21:21:32.076280Z",
     "shell.execute_reply": "2025-01-07T21:21:32.075882Z",
     "shell.execute_reply.started": "2025-01-07T21:21:32.037761Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "28b7f74b-b841-4531-ba69-3e347ef2a216",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T21:21:32.733353Z",
     "iopub.status.busy": "2025-01-07T21:21:32.732996Z",
     "iopub.status.idle": "2025-01-07T21:21:42.928958Z",
     "shell.execute_reply": "2025-01-07T21:21:42.928525Z",
     "shell.execute_reply.started": "2025-01-07T21:21:32.733336Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 16:21:33.302 EST [1851638] LOG:  starting PostgreSQL 17.2 on x86_64-conda-linux-gnu, compiled by x86_64-conda-linux-gnu-cc (Anaconda gcc) 11.2.0, 64-bit\n",
      "2025-01-07 16:21:33.302 EST [1851638] LOG:  listening on IPv6 address \"::1\", port 5433\n",
      "2025-01-07 16:21:33.302 EST [1851638] LOG:  listening on IPv4 address \"127.0.0.1\", port 5433\n",
      "2025-01-07 16:21:33.306 EST [1851638] LOG:  listening on Unix socket \"/tmp/.s.PGSQL.5433\"\n",
      "2025-01-07 16:21:33.426 EST [1851644] LOG:  database system was interrupted; last known up at 2025-01-07 13:10:31 EST\n",
      "2025-01-07 16:21:36.236 EST [1851644] LOG:  database system was not properly shut down; automatic recovery in progress\n",
      "2025-01-07 16:21:36.291 EST [1851644] LOG:  redo starts at 27/2EAB3728\n",
      "2025-01-07 16:21:36.341 EST [1851644] LOG:  invalid record length at 27/2EAE1C98: expected at least 24, got 0\n",
      "2025-01-07 16:21:36.341 EST [1851644] LOG:  redo done at 27/2EAE1C60 system usage: CPU: user: 0.00 s, system: 0.00 s, elapsed: 0.05 s\n",
      "2025-01-07 16:21:36.351 EST [1851642] LOG:  checkpoint starting: end-of-recovery immediate wait\n",
      "2025-01-07 16:21:36.412 EST [1851642] LOG:  checkpoint complete: wrote 39 buffers (0.2%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.020 s, sync=0.009 s, total=0.063 s; sync files=30, longest=0.003 s, average=0.001 s; distance=185 kB, estimate=185 kB; lsn=27/2EAE1C98, redo lsn=27/2EAE1C98\n",
      "2025-01-07 16:21:36.421 EST [1851638] LOG:  database system is ready to accept connections\n"
     ]
    }
   ],
   "source": [
    "#start the database\n",
    "process = subprocess.Popen([\"postgres\", \"-D\", \"/home/mcn26/palmer_scratch/db\", \"-p\", \"5433\"])\n",
    "\n",
    "# Load the SQL extension\n",
    "%load_ext sql"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba8a6d13-0d89-4905-b50f-f97b3f658643",
   "metadata": {},
   "source": [
    "connect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e2c3dd95-f87a-4046-8faa-066297c01204",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T21:21:42.929920Z",
     "iopub.status.busy": "2025-01-07T21:21:42.929707Z",
     "iopub.status.idle": "2025-01-07T21:21:43.972378Z",
     "shell.execute_reply": "2025-01-07T21:21:43.971822Z",
     "shell.execute_reply.started": "2025-01-07T21:21:42.929906Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%sql\n",
    "postgresql://mr_root:password@localhost:5433/scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1fed16fb-2f22-434c-8fa7-aec2e153ead9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T21:21:43.974882Z",
     "iopub.status.busy": "2025-01-07T21:21:43.973129Z",
     "iopub.status.idle": "2025-01-07T21:21:43.978720Z",
     "shell.execute_reply": "2025-01-07T21:21:43.978342Z",
     "shell.execute_reply.started": "2025-01-07T21:21:43.974860Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "#dealing with a weird bug\n",
    "%config SqlMagic.style = '_DEPRECATED_DEFAULT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3e4d31b0-73a5-4235-8196-83e425174bd4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T15:38:05.020522Z",
     "iopub.status.busy": "2025-01-07T15:38:05.020201Z",
     "iopub.status.idle": "2025-01-07T15:38:05.043264Z",
     "shell.execute_reply": "2025-01-07T15:38:05.042891Z",
     "shell.execute_reply.started": "2025-01-07T15:38:05.020504Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>table_name</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>cars</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>gwas_cata</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>malin_gnomad</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('cars',), ('gwas_cata',), ('malin_gnomad',)]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 10:40:58.481 EST [1936925] LOG:  checkpoint starting: time\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT table_name\n",
    "FROM information_schema.tables\n",
    "WHERE table_schema = 'public';"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc28ff4d-e069-4cb6-8c74-f98f240a0cdf",
   "metadata": {},
   "source": [
    "(\"Cars\" was a test table)\n",
    "\n",
    "Let's check out the fields, to see what we should merge on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "70280143-75ca-45a1-82b3-444e32bd7442",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T15:42:10.963640Z",
     "iopub.status.busy": "2025-01-07T15:42:10.963164Z",
     "iopub.status.idle": "2025-01-07T15:42:11.026444Z",
     "shell.execute_reply": "2025-01-07T15:42:11.026090Z",
     "shell.execute_reply.started": "2025-01-07T15:42:10.963614Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>date_added</th>\n",
       "            <th>pubmed_id</th>\n",
       "            <th>first_author</th>\n",
       "            <th>date</th>\n",
       "            <th>journal</th>\n",
       "            <th>link</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_or_trait</th>\n",
       "            <th>initial_sample_size</th>\n",
       "            <th>replication_sample_size</th>\n",
       "            <th>chromosome_region</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>chromosome_pos</th>\n",
       "            <th>reported_genes</th>\n",
       "            <th>mapped_gene</th>\n",
       "            <th>upstream_gene_id</th>\n",
       "            <th>downstream_gene_id</th>\n",
       "            <th>snp_gene_ids</th>\n",
       "            <th>upstream_gene_distance</th>\n",
       "            <th>downstream_gene_distance</th>\n",
       "            <th>strongest_snp_risk_allele</th>\n",
       "            <th>snps</th>\n",
       "            <th>merged</th>\n",
       "            <th>snp_id_current</th>\n",
       "            <th>context</th>\n",
       "            <th>intergenic</th>\n",
       "            <th>risk_allele_frequency</th>\n",
       "            <th>p_value</th>\n",
       "            <th>p_value_mlog</th>\n",
       "            <th>p_value_text</th>\n",
       "            <th>odds_ratio_or_beta</th>\n",
       "            <th>perc_95_ci</th>\n",
       "            <th>platform_snps_passing_qc</th>\n",
       "            <th>cnv</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>1q21.3</td>\n",
       "            <td>1</td>\n",
       "            <td>152784619</td>\n",
       "            <td>LCE1E</td>\n",
       "            <td>LCE1F - LCE1E</td>\n",
       "            <td>ENSG00000240386</td>\n",
       "            <td>ENSG00000186226</td>\n",
       "            <td>None</td>\n",
       "            <td>7595</td>\n",
       "            <td>1595</td>\n",
       "            <td>rs77199844-?</td>\n",
       "            <td>rs77199844</td>\n",
       "            <td>False</td>\n",
       "            <td>77199844</td>\n",
       "            <td>intergenic_variant</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-17</td>\n",
       "            <td>16.69897000433602</td>\n",
       "            <td>(Atopic dermatitis)</td>\n",
       "            <td>1.23</td>\n",
       "            <td>[1.15–1.32]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>6q23.3</td>\n",
       "            <td>6</td>\n",
       "            <td>137874556</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>ENSG00000118503</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs643177-?</td>\n",
       "            <td>rs643177</td>\n",
       "            <td>False</td>\n",
       "            <td>643177</td>\n",
       "            <td>intron_variant</td>\n",
       "            <td>False</td>\n",
       "            <td>NR</td>\n",
       "            <td>9E-16</td>\n",
       "            <td>15.045757490560675</td>\n",
       "            <td>(Psoriasis)</td>\n",
       "            <td>1.27</td>\n",
       "            <td>[1.20–1.34]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '1q21.3', '1', '152784619', 'LCE1E', 'LCE1F - LCE1E', 'ENSG00000240386', 'ENSG00000186226', None, 7595, 1595, 'rs77199844-?', 'rs77199844', False, '77199844', 'intergenic_variant', True, 'NR', '2E-17', 16.69897000433602, '(Atopic dermatitis)', 1.23, '[1.15–1.32]', 'Affymetrix, Illumina [~ 5200000]', 'N'),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '6q23.3', '6', '137874556', 'TNFAIP3', 'TNFAIP3', None, None, 'ENSG00000118503', None, None, 'rs643177-?', 'rs643177', False, '643177', 'intron_variant', False, 'NR', '9E-16', 15.045757490560675, '(Psoriasis)', 1.27, '[1.20–1.34]', 'Affymetrix, Illumina [~ 5200000]', 'N')]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_cata\n",
    "LIMIT 2;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fe449c51-0b61-4b13-8247-8941a2fd8ac2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T15:43:55.071140Z",
     "iopub.status.busy": "2025-01-07T15:43:55.070916Z",
     "iopub.status.idle": "2025-01-07T15:43:55.631121Z",
     "shell.execute_reply": "2025-01-07T15:43:55.630755Z",
     "shell.execute_reply.started": "2025-01-07T15:43:55.071124Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>pos</th>\n",
       "            <th>id</th>\n",
       "            <th>ref</th>\n",
       "            <th>alt</th>\n",
       "            <th>info</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>chr10</td>\n",
       "            <td>10265</td>\n",
       "            <td>.</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>K562__ref=0.06947866;HepG2__ref=-0.1994611;SKNSH__ref=-0.60886025;K562__alt=0.07767675;HepG2__alt=-0.2021158;SKNSH__alt=-0.62348086;K562__skew=0.0081980955;HepG2__skew=-0.0026547052;SKNSH__skew=-0.01462063</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr10</td>\n",
       "            <td>10266</td>\n",
       "            <td>.</td>\n",
       "            <td>T</td>\n",
       "            <td>A</td>\n",
       "            <td>K562__ref=0.036172275;HepG2__ref=-0.20694233;SKNSH__ref=-0.6183255;K562__alt=0.020159498;HepG2__alt=-0.22194892;SKNSH__alt=-0.63133883;K562__skew=-0.016012775;HepG2__skew=-0.015006584;SKNSH__skew=-0.013013384</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('chr10', 10265, '.', 'C', 'T', 'K562__ref=0.06947866;HepG2__ref=-0.1994611;SKNSH__ref=-0.60886025;K562__alt=0.07767675;HepG2__alt=-0.2021158;SKNSH__alt=-0.62348086;K562__skew=0.0081980955;HepG2__skew=-0.0026547052;SKNSH__skew=-0.01462063'),\n",
       " ('chr10', 10266, '.', 'T', 'A', 'K562__ref=0.036172275;HepG2__ref=-0.20694233;SKNSH__ref=-0.6183255;K562__alt=0.020159498;HepG2__alt=-0.22194892;SKNSH__alt=-0.63133883;K562__skew=-0.016012775;HepG2__skew=-0.015006584;SKNSH__skew=-0.013013384')]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 10:45:58.428 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 10:46:30.279 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.453 s, sync=2.042 s, total=31.851 s; sync files=12, longest=0.989 s, average=0.171 s; distance=76383 kB, estimate=76383 kB; lsn=26/D1149D68, redo lsn=26/D09DE798\n",
      "2025-01-07 10:50:58.009 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 10:51:29.353 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.452 s, sync=1.489 s, total=31.345 s; sync files=11, longest=1.184 s, average=0.136 s; distance=75364 kB, estimate=76281 kB; lsn=26/D5AF3BD0, redo lsn=26/D53778B8\n",
      "2025-01-07 10:55:58.565 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 10:56:30.374 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 4 removed, 0 recycled; write=29.462 s, sync=2.008 s, total=31.810 s; sync files=11, longest=1.311 s, average=0.183 s; distance=75978 kB, estimate=76251 kB; lsn=26/DA54C1C8, redo lsn=26/D9DAA298\n",
      "2025-01-07 11:00:58.311 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:01:29.817 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.588 s, sync=1.532 s, total=31.506 s; sync files=12, longest=0.947 s, average=0.128 s; distance=76117 kB, estimate=76238 kB; lsn=26/DEF92B08, redo lsn=26/DE7FF8D0\n",
      "2025-01-07 11:05:58.805 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:06:30.055 EST [1936925] LOG:  checkpoint complete: wrote 37 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.350 s, sync=1.455 s, total=31.250 s; sync files=11, longest=0.722 s, average=0.133 s; distance=76046 kB, estimate=76219 kB; lsn=26/E39D47E0, redo lsn=26/E3243300\n",
      "2025-01-07 11:10:58.979 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:11:29.982 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 4 removed, 0 recycled; write=29.427 s, sync=1.284 s, total=31.003 s; sync files=11, longest=0.768 s, average=0.117 s; distance=75533 kB, estimate=76150 kB; lsn=26/E83475F8, redo lsn=26/E7C069E0\n",
      "2025-01-07 11:15:58.273 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:16:29.697 EST [1936925] LOG:  checkpoint complete: wrote 37 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.561 s, sync=1.536 s, total=31.424 s; sync files=11, longest=0.774 s, average=0.140 s; distance=75225 kB, estimate=76058 kB; lsn=26/ECCED530, redo lsn=26/EC57D068\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM malin_gnomad\n",
    "LIMIT 2;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "019e4b55-1d86-4467-a339-3b583a31b64b",
   "metadata": {},
   "source": [
    "Examining the documentation for the gwas catalog [retrieved 2024-01-07](https://www.ebi.ac.uk/gwas/docs/fileheaders#_file_headers_for_catalog_v1_0):\n",
    "\n",
    "> STRONGEST SNP-RISK ALLELE*: SNP(s) most strongly associated with trait + risk allele (? for unknown risk allele). May also refer to a haplotype.\n",
    "\n",
    "> SNPS*: Strongest SNP; if a haplotype it may include more than one rs number (multiple SNPs comprising the haplotype)\n",
    "\n",
    "> MERGED*: denotes whether the SNP has been merged into a subsequent rs record (0 = no; 1 = yes;)\n",
    "\n",
    "> SNP_ID_CURRENT*: current rs number (will differ from strongest SNP when merged = 1)\n",
    "\n",
    "So \"SNP_ID_CURRENT\" is probably what we want: column 24. This column does not have the risk ID, but \n",
    "\n",
    "Unfortunately paging through the data there seem to be some with null values. The rest appear well-formatted. Let's check out the nulls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cde6eb98-d1f5-4430-aa68-e31dab0ef218",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T16:17:33.075904Z",
     "iopub.status.busy": "2025-01-07T16:17:33.075578Z",
     "iopub.status.idle": "2025-01-07T16:17:33.134421Z",
     "shell.execute_reply": "2025-01-07T16:17:33.134016Z",
     "shell.execute_reply.started": "2025-01-07T16:17:33.075886Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>date_added</th>\n",
       "            <th>pubmed_id</th>\n",
       "            <th>first_author</th>\n",
       "            <th>date</th>\n",
       "            <th>journal</th>\n",
       "            <th>link</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_or_trait</th>\n",
       "            <th>initial_sample_size</th>\n",
       "            <th>replication_sample_size</th>\n",
       "            <th>chromosome_region</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>chromosome_pos</th>\n",
       "            <th>reported_genes</th>\n",
       "            <th>mapped_gene</th>\n",
       "            <th>upstream_gene_id</th>\n",
       "            <th>downstream_gene_id</th>\n",
       "            <th>snp_gene_ids</th>\n",
       "            <th>upstream_gene_distance</th>\n",
       "            <th>downstream_gene_distance</th>\n",
       "            <th>strongest_snp_risk_allele</th>\n",
       "            <th>snps</th>\n",
       "            <th>merged</th>\n",
       "            <th>snp_id_current</th>\n",
       "            <th>context</th>\n",
       "            <th>intergenic</th>\n",
       "            <th>risk_allele_frequency</th>\n",
       "            <th>p_value</th>\n",
       "            <th>p_value_mlog</th>\n",
       "            <th>p_value_text</th>\n",
       "            <th>odds_ratio_or_beta</th>\n",
       "            <th>perc_95_ci</th>\n",
       "            <th>platform_snps_passing_qc</th>\n",
       "            <th>cnv</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>2015-01-13</td>\n",
       "            <td>24940741</td>\n",
       "            <td>Chang SW</td>\n",
       "            <td>2014-06-18</td>\n",
       "            <td>PLoS One</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/24940741</td>\n",
       "            <td>A genome-wide association study on chronic HBV infection and its clinical progression in male Han-Taiwanese.</td>\n",
       "            <td>Chronic hepatitis B infection</td>\n",
       "            <td>321 Chinese ancestry cases, 304 Chinese ancestry controls</td>\n",
       "            <td>1,302 Chinese ancestry cases, 761 Chinese ancestry controls</td>\n",
       "            <td>6p21.32</td>\n",
       "            <td>6;6;6;6;6</td>\n",
       "            <td>32756140;32762235;33136398;32739518;33087084</td>\n",
       "            <td>HLA-DQA2, HLA-DQB2, HLA-DPB1, HLA-DPA3</td>\n",
       "            <td>HLA-DQB2; HLA-DQB2; HLA-DPA3; HLA-DQB3 - HLA-DQA2; HLA-DPB1</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs7756516-T; rs7453920-G; rs9366816-T; rs9276370-T; rs9277535-G</td>\n",
       "            <td>rs7756516; rs7453920; rs9366816; rs9276370; rs9277535</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>3_prime_UTR_variant; intron_variant; intron_variant; intergenic_variant; 3_prime_UTR_variant</td>\n",
       "            <td>None</td>\n",
       "            <td>0.4305</td>\n",
       "            <td>1E-12</td>\n",
       "            <td>12.0</td>\n",
       "            <td>None</td>\n",
       "            <td>1.49</td>\n",
       "            <td>[1.33-1.66]</td>\n",
       "            <td>Illumina [456262]</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2017-03-20</td>\n",
       "            <td>27629089</td>\n",
       "            <td>Sailer A</td>\n",
       "            <td>2016-09-14</td>\n",
       "            <td>Neurology</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/27629089</td>\n",
       "            <td>A genome-wide association study in multiple system atrophy.</td>\n",
       "            <td>Multiple system atrophy (pathologically confirmed)</td>\n",
       "            <td>295 European ancestry cases, 3,864 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>NDE1</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr16:15689134-?</td>\n",
       "            <td>chr16:15689134</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-6</td>\n",
       "            <td>5.698970004336019</td>\n",
       "            <td>None</td>\n",
       "            <td>7.692308</td>\n",
       "            <td>NR</td>\n",
       "            <td>Illumina [4903804] (imputed)</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2014-11-11</td>\n",
       "            <td>24792382</td>\n",
       "            <td>Park TJ</td>\n",
       "            <td>2014-04-30</td>\n",
       "            <td>Clin Chim Acta</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/24792382</td>\n",
       "            <td>Genome-wide association study identifies ALLC polymorphisms correlated with FEV₁ change by corticosteroid.</td>\n",
       "            <td>Response to inhaled corticosteroid treatment in asthma (change in FEV1)</td>\n",
       "            <td>189 Korean ancestry cases</td>\n",
       "            <td>NA</td>\n",
       "            <td>2p25.3</td>\n",
       "            <td>2;2;2;2;2;2;2;2;2;2;2;2</td>\n",
       "            <td>3659833;3659032;3661518;3666539;3653746;3655451;3650781;3655143;3657240;3667425;3666068;3666671</td>\n",
       "            <td>ALLC</td>\n",
       "            <td>ALLC; ALLC; ALLC; ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; ALLC; ALLC; ALLC</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs6754459-T; rs13384889-C; rs1965732-G; rs10173297-A; rs4849975-G; rs17445240-G; rs13031619-G; rs11902059-A; rs13418767-T; rs13395090-T; rs17017879-C; rs7558370-C</td>\n",
       "            <td>rs6754459; rs13384889; rs1965732; rs10173297; rs4849975; rs17445240; rs13031619; rs11902059; rs13418767; rs13395090; rs17017879; rs7558370</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>intron_variant; intron_variant; intron_variant; intron_variant; intergenic_variant; intergenic_variant; regulatory_region_variant; intergenic_variant; regulatory_region_variant; intron_variant; intron_variant; intron_variant</td>\n",
       "            <td>None</td>\n",
       "            <td>0.093</td>\n",
       "            <td>4E-6</td>\n",
       "            <td>5.3979400086720375</td>\n",
       "            <td>None</td>\n",
       "            <td>0.329</td>\n",
       "            <td>[NR] unit increase</td>\n",
       "            <td>Illumina [430487]</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2017-03-20</td>\n",
       "            <td>27629089</td>\n",
       "            <td>Sailer A</td>\n",
       "            <td>2016-09-14</td>\n",
       "            <td>Neurology</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/27629089</td>\n",
       "            <td>A genome-wide association study in multiple system atrophy.</td>\n",
       "            <td>Multiple system atrophy (pathologically confirmed)</td>\n",
       "            <td>295 European ancestry cases, 3,864 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>LOC283867</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr16:64525400-?</td>\n",
       "            <td>chr16:64525400</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>9E-6</td>\n",
       "            <td>5.045757490560675</td>\n",
       "            <td>None</td>\n",
       "            <td>1.97</td>\n",
       "            <td>NR</td>\n",
       "            <td>Illumina [4903804] (imputed)</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2024-09-20</td>\n",
       "            <td>37794183</td>\n",
       "            <td>Dhindsa RS</td>\n",
       "            <td>2023-10-04</td>\n",
       "            <td>Nature</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/37794183</td>\n",
       "            <td>Rare variant associations with plasma protein levels in the UK Biobank.</td>\n",
       "            <td>Poliovirus receptor levels</td>\n",
       "            <td>46,327 European ancestry individuals</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr19:43873462-?</td>\n",
       "            <td>chr19:43873462</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>9E-12</td>\n",
       "            <td>11.045757490560675</td>\n",
       "            <td>(genotypic)</td>\n",
       "            <td>0.1741963</td>\n",
       "            <td>[-0.224270228--0.124122374] unit decrease</td>\n",
       "            <td>NR [617073]</td>\n",
       "            <td>N</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('2015-01-13', 24940741, 'Chang SW', '2014-06-18', 'PLoS One', 'www.ncbi.nlm.nih.gov/pubmed/24940741', 'A genome-wide association study on chronic HBV infection and its clinical progression in male Han-Taiwanese.', 'Chronic hepatitis B infection', '321 Chinese ancestry cases, 304 Chinese ancestry controls', '1,302 Chinese ancestry cases, 761 Chinese ancestry controls', '6p21.32', '6;6;6;6;6', '32756140;32762235;33136398;32739518;33087084', 'HLA-DQA2, HLA-DQB2, HLA-DPB1, HLA-DPA3', 'HLA-DQB2; HLA-DQB2; HLA-DPA3; HLA-DQB3 - HLA-DQA2; HLA-DPB1', None, None, None, None, None, 'rs7756516-T; rs7453920-G; rs9366816-T; rs9276370-T; rs9277535-G', 'rs7756516; rs7453920; rs9366816; rs9276370; rs9277535', False, None, '3_prime_UTR_variant; intron_variant; intron_variant; intergenic_variant; 3_prime_UTR_variant', None, '0.4305', '1E-12', 12.0, None, 1.49, '[1.33-1.66]', 'Illumina [456262]', 'N'),\n",
       " ('2017-03-20', 27629089, 'Sailer A', '2016-09-14', 'Neurology', 'www.ncbi.nlm.nih.gov/pubmed/27629089', 'A genome-wide association study in multiple system atrophy.', 'Multiple system atrophy (pathologically confirmed)', '295 European ancestry cases, 3,864 European ancestry controls', 'NA', None, None, None, 'NDE1', None, None, None, None, None, None, 'chr16:15689134-?', 'chr16:15689134', False, None, None, True, 'NR', '2E-6', 5.698970004336019, None, 7.692308, 'NR', 'Illumina [4903804] (imputed)', 'N'),\n",
       " ('2014-11-11', 24792382, 'Park TJ', '2014-04-30', 'Clin Chim Acta', 'www.ncbi.nlm.nih.gov/pubmed/24792382', 'Genome-wide association study identifies ALLC polymorphisms correlated with FEV₁ change by corticosteroid.', 'Response to inhaled corticosteroid treatment in asthma (change in FEV1)', '189 Korean ancestry cases', 'NA', '2p25.3', '2;2;2;2;2;2;2;2;2;2;2;2', '3659833;3659032;3661518;3666539;3653746;3655451;3650781;3655143;3657240;3667425;3666068;3666671', 'ALLC', 'ALLC; ALLC; ALLC; ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; COLEC11 - ALLC; ALLC; ALLC; ALLC', None, None, None, None, None, 'rs6754459-T; rs13384889-C; rs1965732-G; rs10173297-A; rs4849975-G; rs17445240-G; rs13031619-G; rs11902059-A; rs13418767-T; rs13395090-T; rs17017879-C; rs7558370-C', 'rs6754459; rs13384889; rs1965732; rs10173297; rs4849975; rs17445240; rs13031619; rs11902059; rs13418767; rs13395090; rs17017879; rs7558370', False, None, 'intron_variant; intron_variant; intron_variant; intron_variant; intergenic_variant; intergenic_variant; regulatory_region_variant; intergenic_variant; regulatory_region_variant; intron_variant; intron_variant; intron_variant', None, '0.093', '4E-6', 5.3979400086720375, None, 0.329, '[NR] unit increase', 'Illumina [430487]', 'N'),\n",
       " ('2017-03-20', 27629089, 'Sailer A', '2016-09-14', 'Neurology', 'www.ncbi.nlm.nih.gov/pubmed/27629089', 'A genome-wide association study in multiple system atrophy.', 'Multiple system atrophy (pathologically confirmed)', '295 European ancestry cases, 3,864 European ancestry controls', 'NA', None, None, None, 'LOC283867', None, None, None, None, None, None, 'chr16:64525400-?', 'chr16:64525400', False, None, None, True, 'NR', '9E-6', 5.045757490560675, None, 1.97, 'NR', 'Illumina [4903804] (imputed)', 'N'),\n",
       " ('2024-09-20', 37794183, 'Dhindsa RS', '2023-10-04', 'Nature', 'www.ncbi.nlm.nih.gov/pubmed/37794183', 'Rare variant associations with plasma protein levels in the UK Biobank.', 'Poliovirus receptor levels', '46,327 European ancestry individuals', 'NA', None, None, None, None, None, None, None, None, None, None, 'chr19:43873462-?', 'chr19:43873462', False, None, None, True, 'NR', '9E-12', 11.045757490560675, '(genotypic)', 0.1741963, '[-0.224270228--0.124122374] unit decrease', 'NR [617073]', 'N')]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 11:20:58.640 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:21:30.111 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 4 removed, 0 recycled; write=29.476 s, sync=1.797 s, total=31.471 s; sync files=11, longest=1.201 s, average=0.164 s; distance=75385 kB, estimate=75990 kB; lsn=26/F16A50D0, redo lsn=26/F0F1B5C0\n",
      "2025-01-07 11:25:58.480 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:26:30.078 EST [1936925] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.975 s, sync=1.253 s, total=31.598 s; sync files=11, longest=1.199 s, average=0.114 s; distance=75461 kB, estimate=75937 kB; lsn=26/F602FA08, redo lsn=26/F58CCBA0\n",
      "2025-01-07 11:30:58.459 EST [1936925] LOG:  checkpoint starting: time\n",
      "2025-01-07 11:31:29.300 EST [1936925] LOG:  checkpoint complete: wrote 37 buffers (0.2%); 0 WAL file(s) added, 5 removed, 0 recycled; write=29.654 s, sync=0.836 s, total=30.842 s; sync files=11, longest=0.836 s, average=0.076 s; distance=74597 kB, estimate=75803 kB; lsn=26/FA8FDE68, redo lsn=26/FA1A6058\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_cata\n",
    "WHERE snp_id_current IS NULL\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65a31ed-0f76-412d-9638-5701b5453a22",
   "metadata": {},
   "source": [
    "Let's add a \"gwas_catalog_id\" to uniquely identify entries in the gwas catalog. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "84a2bd1b-4f64-402a-88e6-9074d7090f14",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:04:51.715222Z",
     "iopub.status.busy": "2025-01-07T18:04:51.714704Z",
     "iopub.status.idle": "2025-01-07T18:05:24.299813Z",
     "shell.execute_reply": "2025-01-07T18:05:24.299228Z",
     "shell.execute_reply.started": "2025-01-07T18:04:51.715194Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--ALTER TABLE gwas_cata DROP COLUMN id;\n",
    "ALTER TABLE gwas_cata ADD COLUMN id SERIAL PRIMARY KEY;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e66d62b-7655-4846-ab14-8cf1ad237562",
   "metadata": {},
   "source": [
    "Ok, so examing a few rows, I see a null in the \"snp_id_current\" row when \"snps\" is ambiguious, and when \"merged\" is false. \n",
    "My suspicion is that the former is the cause of the nulls: according to the docs linked above, I see that merged=false should only result in the field strongest snp being duplicated.\n",
    "So my guess is : weird formatting -> failure to merge into new SNPs -> false. So there should be some rows where merged is false but snp_id_current is not null. Let's check:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "0f0ba091-f58c-4a9d-8e6a-31ad58adbc59",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:05:24.302505Z",
     "iopub.status.busy": "2025-01-07T18:05:24.302219Z",
     "iopub.status.idle": "2025-01-07T18:05:24.335055Z",
     "shell.execute_reply": "2025-01-07T18:05:24.334589Z",
     "shell.execute_reply.started": "2025-01-07T18:05:24.302487Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>date_added</th>\n",
       "            <th>pubmed_id</th>\n",
       "            <th>first_author</th>\n",
       "            <th>date</th>\n",
       "            <th>journal</th>\n",
       "            <th>link</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_or_trait</th>\n",
       "            <th>initial_sample_size</th>\n",
       "            <th>replication_sample_size</th>\n",
       "            <th>chromosome_region</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>chromosome_pos</th>\n",
       "            <th>reported_genes</th>\n",
       "            <th>mapped_gene</th>\n",
       "            <th>upstream_gene_id</th>\n",
       "            <th>downstream_gene_id</th>\n",
       "            <th>snp_gene_ids</th>\n",
       "            <th>upstream_gene_distance</th>\n",
       "            <th>downstream_gene_distance</th>\n",
       "            <th>strongest_snp_risk_allele</th>\n",
       "            <th>snps</th>\n",
       "            <th>merged</th>\n",
       "            <th>snp_id_current</th>\n",
       "            <th>context</th>\n",
       "            <th>intergenic</th>\n",
       "            <th>risk_allele_frequency</th>\n",
       "            <th>p_value</th>\n",
       "            <th>p_value_mlog</th>\n",
       "            <th>p_value_text</th>\n",
       "            <th>odds_ratio_or_beta</th>\n",
       "            <th>perc_95_ci</th>\n",
       "            <th>platform_snps_passing_qc</th>\n",
       "            <th>cnv</th>\n",
       "            <th>id</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>1q21.3</td>\n",
       "            <td>1</td>\n",
       "            <td>152784619</td>\n",
       "            <td>LCE1E</td>\n",
       "            <td>LCE1F - LCE1E</td>\n",
       "            <td>ENSG00000240386</td>\n",
       "            <td>ENSG00000186226</td>\n",
       "            <td>None</td>\n",
       "            <td>7595</td>\n",
       "            <td>1595</td>\n",
       "            <td>rs77199844-?</td>\n",
       "            <td>rs77199844</td>\n",
       "            <td>False</td>\n",
       "            <td>77199844</td>\n",
       "            <td>intergenic_variant</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-17</td>\n",
       "            <td>16.69897000433602</td>\n",
       "            <td>(Atopic dermatitis)</td>\n",
       "            <td>1.23</td>\n",
       "            <td>[1.15–1.32]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>6q23.3</td>\n",
       "            <td>6</td>\n",
       "            <td>137874556</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>ENSG00000118503</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs643177-?</td>\n",
       "            <td>rs643177</td>\n",
       "            <td>False</td>\n",
       "            <td>643177</td>\n",
       "            <td>intron_variant</td>\n",
       "            <td>False</td>\n",
       "            <td>NR</td>\n",
       "            <td>9E-16</td>\n",
       "            <td>15.045757490560675</td>\n",
       "            <td>(Psoriasis)</td>\n",
       "            <td>1.27</td>\n",
       "            <td>[1.20–1.34]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>2</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>1q21.3</td>\n",
       "            <td>1</td>\n",
       "            <td>152560266</td>\n",
       "            <td>LCE3E</td>\n",
       "            <td>CRCT1 - LCE3E</td>\n",
       "            <td>ENSG00000169509</td>\n",
       "            <td>ENSG00000185966</td>\n",
       "            <td>None</td>\n",
       "            <td>44258</td>\n",
       "            <td>5388</td>\n",
       "            <td>rs10888499-?</td>\n",
       "            <td>rs10888499</td>\n",
       "            <td>False</td>\n",
       "            <td>10888499</td>\n",
       "            <td>intergenic_variant</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>5E-25</td>\n",
       "            <td>24.30102999566398</td>\n",
       "            <td>(Atopic dermatitis)</td>\n",
       "            <td>1.49</td>\n",
       "            <td>[1.38–1.61]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>3</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>5q33.3</td>\n",
       "            <td>5</td>\n",
       "            <td>159231004</td>\n",
       "            <td>CTB-11I22.1</td>\n",
       "            <td>LINC01932</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>ENSG00000253315</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs10515778-?</td>\n",
       "            <td>rs10515778</td>\n",
       "            <td>False</td>\n",
       "            <td>10515778</td>\n",
       "            <td>intron_variant</td>\n",
       "            <td>False</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-14</td>\n",
       "            <td>13.698970004336019</td>\n",
       "            <td>(Psoriasis)</td>\n",
       "            <td>1.29</td>\n",
       "            <td>[1.21–1.38]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>4</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>1q21.3</td>\n",
       "            <td>1</td>\n",
       "            <td>152481779</td>\n",
       "            <td>RP1-91G5.3, LCE5A</td>\n",
       "            <td>FLG-AS1 - LCE5A</td>\n",
       "            <td>ENSG00000237975</td>\n",
       "            <td>ENSG00000186207</td>\n",
       "            <td>None</td>\n",
       "            <td>36323</td>\n",
       "            <td>29024</td>\n",
       "            <td>rs471144-?</td>\n",
       "            <td>rs471144</td>\n",
       "            <td>False</td>\n",
       "            <td>471144</td>\n",
       "            <td>intergenic_variant</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-12</td>\n",
       "            <td>11.698970004336019</td>\n",
       "            <td>(Atopic dermatitis)</td>\n",
       "            <td>1.54</td>\n",
       "            <td>[1.37–1.73]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>5</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '1q21.3', '1', '152784619', 'LCE1E', 'LCE1F - LCE1E', 'ENSG00000240386', 'ENSG00000186226', None, 7595, 1595, 'rs77199844-?', 'rs77199844', False, '77199844', 'intergenic_variant', True, 'NR', '2E-17', 16.69897000433602, '(Atopic dermatitis)', 1.23, '[1.15–1.32]', 'Affymetrix, Illumina [~ 5200000]', 'N', 1),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '6q23.3', '6', '137874556', 'TNFAIP3', 'TNFAIP3', None, None, 'ENSG00000118503', None, None, 'rs643177-?', 'rs643177', False, '643177', 'intron_variant', False, 'NR', '9E-16', 15.045757490560675, '(Psoriasis)', 1.27, '[1.20–1.34]', 'Affymetrix, Illumina [~ 5200000]', 'N', 2),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '1q21.3', '1', '152560266', 'LCE3E', 'CRCT1 - LCE3E', 'ENSG00000169509', 'ENSG00000185966', None, 44258, 5388, 'rs10888499-?', 'rs10888499', False, '10888499', 'intergenic_variant', True, 'NR', '5E-25', 24.30102999566398, '(Atopic dermatitis)', 1.49, '[1.38–1.61]', 'Affymetrix, Illumina [~ 5200000]', 'N', 3),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '5q33.3', '5', '159231004', 'CTB-11I22.1', 'LINC01932', None, None, 'ENSG00000253315', None, None, 'rs10515778-?', 'rs10515778', False, '10515778', 'intron_variant', False, 'NR', '2E-14', 13.698970004336019, '(Psoriasis)', 1.29, '[1.21–1.38]', 'Affymetrix, Illumina [~ 5200000]', 'N', 4),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '1q21.3', '1', '152481779', 'RP1-91G5.3, LCE5A', 'FLG-AS1 - LCE5A', 'ENSG00000237975', 'ENSG00000186207', None, 36323, 29024, 'rs471144-?', 'rs471144', False, '471144', 'intergenic_variant', True, 'NR', '2E-12', 11.698970004336019, '(Atopic dermatitis)', 1.54, '[1.37–1.73]', 'Affymetrix, Illumina [~ 5200000]', 'N', 5)]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 13:07:06.861 EST [1949987] LOG:  checkpoint starting: time\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_cata\n",
    "WHERE merged =False\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddf9af26-f6e5-4c1b-b781-c1d3accb52d2",
   "metadata": {},
   "source": [
    "Yes, there are plenty of merged=false where the snp_id_current is not null. Unsurprisingly, those sampled correspond to rows with \"snps\" fields that are nice and clean (single snp in RSID format)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "362a4af7-bf9b-43fe-87a8-5273f85563c9",
   "metadata": {},
   "source": [
    "Let's produce views of the gwas catalog that reflect different kinds of values present in \"snps\"\n",
    "1. gwas_snplist : has a list of rsIDs\n",
    "2. gwas_chrpos : has a chromosome and position\n",
    "3. gwas_other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "1263c39b-c6ec-4bd6-a47b-5a1500d8d53d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:08:22.306972Z",
     "iopub.status.busy": "2025-01-07T18:08:22.306362Z",
     "iopub.status.idle": "2025-01-07T18:08:22.340801Z",
     "shell.execute_reply": "2025-01-07T18:08:22.340408Z",
     "shell.execute_reply.started": "2025-01-07T18:08:22.306945Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--DROP VIEW gwas_snplist;--uncomment if made previous attempts\n",
    "CREATE VIEW gwas_snplist AS\n",
    "SELECT *\n",
    "FROM gwas_cata\n",
    "WHERE snps ~ '^rs[0-9]+(; ?rs[0-9]+)*$';"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "425f51b6-b9b1-43b3-9aed-5ca3eaf43f40",
   "metadata": {},
   "source": [
    "I really don't like using regex for this application but...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "319fad35-44c2-4573-b428-93dd92a34305",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:08:30.638487Z",
     "iopub.status.busy": "2025-01-07T18:08:30.637997Z",
     "iopub.status.idle": "2025-01-07T18:08:30.644210Z",
     "shell.execute_reply": "2025-01-07T18:08:30.643848Z",
     "shell.execute_reply.started": "2025-01-07T18:08:30.638460Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "2 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>date_added</th>\n",
       "            <th>pubmed_id</th>\n",
       "            <th>first_author</th>\n",
       "            <th>date</th>\n",
       "            <th>journal</th>\n",
       "            <th>link</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_or_trait</th>\n",
       "            <th>initial_sample_size</th>\n",
       "            <th>replication_sample_size</th>\n",
       "            <th>chromosome_region</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>chromosome_pos</th>\n",
       "            <th>reported_genes</th>\n",
       "            <th>mapped_gene</th>\n",
       "            <th>upstream_gene_id</th>\n",
       "            <th>downstream_gene_id</th>\n",
       "            <th>snp_gene_ids</th>\n",
       "            <th>upstream_gene_distance</th>\n",
       "            <th>downstream_gene_distance</th>\n",
       "            <th>strongest_snp_risk_allele</th>\n",
       "            <th>snps</th>\n",
       "            <th>merged</th>\n",
       "            <th>snp_id_current</th>\n",
       "            <th>context</th>\n",
       "            <th>intergenic</th>\n",
       "            <th>risk_allele_frequency</th>\n",
       "            <th>p_value</th>\n",
       "            <th>p_value_mlog</th>\n",
       "            <th>p_value_text</th>\n",
       "            <th>odds_ratio_or_beta</th>\n",
       "            <th>perc_95_ci</th>\n",
       "            <th>platform_snps_passing_qc</th>\n",
       "            <th>cnv</th>\n",
       "            <th>id</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>1q21.3</td>\n",
       "            <td>1</td>\n",
       "            <td>152784619</td>\n",
       "            <td>LCE1E</td>\n",
       "            <td>LCE1F - LCE1E</td>\n",
       "            <td>ENSG00000240386</td>\n",
       "            <td>ENSG00000186226</td>\n",
       "            <td>None</td>\n",
       "            <td>7595</td>\n",
       "            <td>1595</td>\n",
       "            <td>rs77199844-?</td>\n",
       "            <td>rs77199844</td>\n",
       "            <td>False</td>\n",
       "            <td>77199844</td>\n",
       "            <td>intergenic_variant</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>2E-17</td>\n",
       "            <td>16.69897000433602</td>\n",
       "            <td>(Atopic dermatitis)</td>\n",
       "            <td>1.23</td>\n",
       "            <td>[1.15–1.32]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2015-12-18</td>\n",
       "            <td>25574825</td>\n",
       "            <td>Baurecht H</td>\n",
       "            <td>2015-01-08</td>\n",
       "            <td>Am J Hum Genet</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/25574825</td>\n",
       "            <td>Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.</td>\n",
       "            <td>Inflammatory skin disease</td>\n",
       "            <td>2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls</td>\n",
       "            <td>NA</td>\n",
       "            <td>6q23.3</td>\n",
       "            <td>6</td>\n",
       "            <td>137874556</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>TNFAIP3</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>ENSG00000118503</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>rs643177-?</td>\n",
       "            <td>rs643177</td>\n",
       "            <td>False</td>\n",
       "            <td>643177</td>\n",
       "            <td>intron_variant</td>\n",
       "            <td>False</td>\n",
       "            <td>NR</td>\n",
       "            <td>9E-16</td>\n",
       "            <td>15.045757490560675</td>\n",
       "            <td>(Psoriasis)</td>\n",
       "            <td>1.27</td>\n",
       "            <td>[1.20–1.34]</td>\n",
       "            <td>Affymetrix, Illumina [~ 5200000]</td>\n",
       "            <td>N</td>\n",
       "            <td>2</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '1q21.3', '1', '152784619', 'LCE1E', 'LCE1F - LCE1E', 'ENSG00000240386', 'ENSG00000186226', None, 7595, 1595, 'rs77199844-?', 'rs77199844', False, '77199844', 'intergenic_variant', True, 'NR', '2E-17', 16.69897000433602, '(Atopic dermatitis)', 1.23, '[1.15–1.32]', 'Affymetrix, Illumina [~ 5200000]', 'N', 1),\n",
       " ('2015-12-18', 25574825, 'Baurecht H', '2015-01-08', 'Am J Hum Genet', 'www.ncbi.nlm.nih.gov/pubmed/25574825', 'Genome-wide comparative analysis of atopic dermatitis and psoriasis gives insight into opposing genetic mechanisms.', 'Inflammatory skin disease', '2,079 European ancestry atopic dermatitis cases, 4,212 European ancestry psoriasis cases, 11,899 European ancestry controls', 'NA', '6q23.3', '6', '137874556', 'TNFAIP3', 'TNFAIP3', None, None, 'ENSG00000118503', None, None, 'rs643177-?', 'rs643177', False, '643177', 'intron_variant', False, 'NR', '9E-16', 15.045757490560675, '(Psoriasis)', 1.27, '[1.20–1.34]', 'Affymetrix, Illumina [~ 5200000]', 'N', 2)]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_snplist\n",
    "LIMIT 2;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "f2dce45c-87f0-4c17-bf27-63ef366f1d95",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T17:24:39.530171Z",
     "iopub.status.busy": "2025-01-07T17:24:39.529830Z",
     "iopub.status.idle": "2025-01-07T17:24:39.532502Z",
     "shell.execute_reply": "2025-01-07T17:24:39.532072Z",
     "shell.execute_reply.started": "2025-01-07T17:24:39.530153Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 12:27:04.197 EST [1949987] LOG:  checkpoint starting: time\n",
      "2025-01-07 12:27:38.631 EST [1949987] LOG:  checkpoint complete: wrote 74 buffers (0.5%); 0 WAL file(s) added, 0 removed, 3 recycled; write=32.862 s, sync=1.273 s, total=34.434 s; sync files=38, longest=1.215 s, average=0.034 s; distance=59370 kB, estimate=59370 kB; lsn=27/11387DC8, redo lsn=27/10B60128\n",
      "2025-01-07 12:32:04.068 EST [1949987] LOG:  checkpoint starting: time\n",
      "2025-01-07 12:32:35.404 EST [1949987] LOG:  checkpoint complete: wrote 37 buffers (0.2%); 0 WAL file(s) added, 0 removed, 5 recycled; write=29.643 s, sync=1.319 s, total=31.336 s; sync files=11, longest=1.318 s, average=0.120 s; distance=75333 kB, estimate=75333 kB; lsn=27/15C64AE0, redo lsn=27/154F16A8\n",
      "2025-01-07 12:33:25.080 EST [1950081] LOG:  automatic vacuum of table \"scratch.public.malin_gnomad\": index scans: 0\n",
      "\tpages: 0 removed, 20371536 remain, 2454851 scanned (12.05% of total)\n",
      "\ttuples: 0 removed, 621808782 remain, 0 are dead but not yet removable\n",
      "\tremovable cutoff: 769, which was 1 XIDs old when operation ended\n",
      "\tfrozen: 0 pages from table (0.00% of total) had 0 tuples frozen\n",
      "\tindex scan not needed: 0 pages from table (0.00% of total) had 0 dead item identifiers removed\n",
      "\tavg read rate: 31.007 MB/s, avg write rate: 30.938 MB/s\n",
      "\tbuffer usage: 2455567 hits, 2460410 misses, 2454903 dirtied\n",
      "\tWAL usage: 2454823 records, 79 full page images, 145480561 bytes\n",
      "\tsystem usage: CPU: user: 7.41 s, system: 18.82 s, elapsed: 619.91 s\n",
      "2025-01-07 12:37:05.032 EST [1949987] LOG:  checkpoint starting: time\n",
      "2025-01-07 12:37:07.967 EST [1949987] LOG:  checkpoint complete: wrote 26 buffers (0.2%); 0 WAL file(s) added, 0 removed, 1 recycled; write=2.613 s, sync=0.040 s, total=2.936 s; sync files=14, longest=0.029 s, average=0.003 s; distance=20041 kB, estimate=69804 kB; lsn=27/16883C38, redo lsn=27/16883BE0\n"
     ]
    }
   ],
   "source": [
    "#manually checked that regex by inspecting following\n",
    "#COPY (SELECT * FROM gwas_snplist)\n",
    "#TO '/home/mcn26/temp.tsv'\n",
    "#WITH (FORMAT text, DELIMITER E'\\t', HEADER true);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c10d34d1-70b0-40f7-8a14-f67664649d78",
   "metadata": {},
   "source": [
    "Create a view of those with chromosome position instead of a snp list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "29fb01f6-8a3e-4652-91db-c9cdd5f3d27f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:09:25.301542Z",
     "iopub.status.busy": "2025-01-07T18:09:25.300982Z",
     "iopub.status.idle": "2025-01-07T18:09:25.327847Z",
     "shell.execute_reply": "2025-01-07T18:09:25.327417Z",
     "shell.execute_reply.started": "2025-01-07T18:09:25.301516Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "Done.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE VIEW gwas_chrpos AS\n",
    "SELECT *\n",
    "FROM gwas_cata\n",
    "WHERE snps ~ '.*chr.*';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "dc467f2f-a478-414a-ab58-c461448937b1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:10:50.034348Z",
     "iopub.status.busy": "2025-01-07T18:10:50.033785Z",
     "iopub.status.idle": "2025-01-07T18:10:50.052420Z",
     "shell.execute_reply": "2025-01-07T18:10:50.052050Z",
     "shell.execute_reply.started": "2025-01-07T18:10:50.034323Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>date_added</th>\n",
       "            <th>pubmed_id</th>\n",
       "            <th>first_author</th>\n",
       "            <th>date</th>\n",
       "            <th>journal</th>\n",
       "            <th>link</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_or_trait</th>\n",
       "            <th>initial_sample_size</th>\n",
       "            <th>replication_sample_size</th>\n",
       "            <th>chromosome_region</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>chromosome_pos</th>\n",
       "            <th>reported_genes</th>\n",
       "            <th>mapped_gene</th>\n",
       "            <th>upstream_gene_id</th>\n",
       "            <th>downstream_gene_id</th>\n",
       "            <th>snp_gene_ids</th>\n",
       "            <th>upstream_gene_distance</th>\n",
       "            <th>downstream_gene_distance</th>\n",
       "            <th>strongest_snp_risk_allele</th>\n",
       "            <th>snps</th>\n",
       "            <th>merged</th>\n",
       "            <th>snp_id_current</th>\n",
       "            <th>context</th>\n",
       "            <th>intergenic</th>\n",
       "            <th>risk_allele_frequency</th>\n",
       "            <th>p_value</th>\n",
       "            <th>p_value_mlog</th>\n",
       "            <th>p_value_text</th>\n",
       "            <th>odds_ratio_or_beta</th>\n",
       "            <th>perc_95_ci</th>\n",
       "            <th>platform_snps_passing_qc</th>\n",
       "            <th>cnv</th>\n",
       "            <th>id</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>2024-09-20</td>\n",
       "            <td>37794183</td>\n",
       "            <td>Dhindsa RS</td>\n",
       "            <td>2023-10-04</td>\n",
       "            <td>Nature</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/37794183</td>\n",
       "            <td>Rare variant associations with plasma protein levels in the UK Biobank.</td>\n",
       "            <td>Poliovirus receptor levels</td>\n",
       "            <td>46,327 European ancestry individuals</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr19:45179618-?</td>\n",
       "            <td>chr19:45179618</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>1E-9</td>\n",
       "            <td>9.0</td>\n",
       "            <td>(genotypic)</td>\n",
       "            <td>0.09152399</td>\n",
       "            <td>[-0.121114629--0.061933356] unit decrease</td>\n",
       "            <td>NR [617073]</td>\n",
       "            <td>N</td>\n",
       "            <td>668</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2024-09-20</td>\n",
       "            <td>37794183</td>\n",
       "            <td>Dhindsa RS</td>\n",
       "            <td>2023-10-04</td>\n",
       "            <td>Nature</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/37794183</td>\n",
       "            <td>Rare variant associations with plasma protein levels in the UK Biobank.</td>\n",
       "            <td>melanoma-derived growth regulatory protein levels</td>\n",
       "            <td>46,327 European ancestry individuals</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr19:40192534-?</td>\n",
       "            <td>chr19:40192534</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>6E-13</td>\n",
       "            <td>12.221848749616356</td>\n",
       "            <td>(dominant)</td>\n",
       "            <td>0.54514116</td>\n",
       "            <td>[0.396918408-0.693363903] unit increase</td>\n",
       "            <td>NR [617073]</td>\n",
       "            <td>N</td>\n",
       "            <td>891</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>2024-09-20</td>\n",
       "            <td>37794183</td>\n",
       "            <td>Dhindsa RS</td>\n",
       "            <td>2023-10-04</td>\n",
       "            <td>Nature</td>\n",
       "            <td>www.ncbi.nlm.nih.gov/pubmed/37794183</td>\n",
       "            <td>Rare variant associations with plasma protein levels in the UK Biobank.</td>\n",
       "            <td>Poliovirus receptor levels</td>\n",
       "            <td>46,327 European ancestry individuals</td>\n",
       "            <td>NA</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>chr19:45408754-?</td>\n",
       "            <td>chr19:45408754</td>\n",
       "            <td>False</td>\n",
       "            <td>None</td>\n",
       "            <td>None</td>\n",
       "            <td>True</td>\n",
       "            <td>NR</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>7.522878745280337</td>\n",
       "            <td>(dominant)</td>\n",
       "            <td>0.5594665</td>\n",
       "            <td>[-0.757610677--0.361322235] unit decrease</td>\n",
       "            <td>NR [617073]</td>\n",
       "            <td>N</td>\n",
       "            <td>951</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('2024-09-20', 37794183, 'Dhindsa RS', '2023-10-04', 'Nature', 'www.ncbi.nlm.nih.gov/pubmed/37794183', 'Rare variant associations with plasma protein levels in the UK Biobank.', 'Poliovirus receptor levels', '46,327 European ancestry individuals', 'NA', None, None, None, None, None, None, None, None, None, None, 'chr19:45179618-?', 'chr19:45179618', False, None, None, True, 'NR', '1E-9', 9.0, '(genotypic)', 0.09152399, '[-0.121114629--0.061933356] unit decrease', 'NR [617073]', 'N', 668),\n",
       " ('2024-09-20', 37794183, 'Dhindsa RS', '2023-10-04', 'Nature', 'www.ncbi.nlm.nih.gov/pubmed/37794183', 'Rare variant associations with plasma protein levels in the UK Biobank.', 'melanoma-derived growth regulatory protein levels', '46,327 European ancestry individuals', 'NA', None, None, None, None, None, None, None, None, None, None, 'chr19:40192534-?', 'chr19:40192534', False, None, None, True, 'NR', '6E-13', 12.221848749616356, '(dominant)', 0.54514116, '[0.396918408-0.693363903] unit increase', 'NR [617073]', 'N', 891),\n",
       " ('2024-09-20', 37794183, 'Dhindsa RS', '2023-10-04', 'Nature', 'www.ncbi.nlm.nih.gov/pubmed/37794183', 'Rare variant associations with plasma protein levels in the UK Biobank.', 'Poliovirus receptor levels', '46,327 European ancestry individuals', 'NA', None, None, None, None, None, None, None, None, None, None, 'chr19:45408754-?', 'chr19:45408754', False, None, None, True, 'NR', '3E-8', 7.522878745280337, '(dominant)', 0.5594665, '[-0.757610677--0.361322235] unit decrease', 'NR [617073]', 'N', 951)]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_chrpos\n",
    "LIMIT 3;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbb51842-74e0-43af-b8d8-357c104f8f38",
   "metadata": {},
   "source": [
    "Bizzarely, despite being identified *only* by chromosome and position, the chromosome_id and chromosome_pos values are empty for at least some. So we will have to extract from the string, which has inconsistent formatting (sometimes including letter, sometimes not...)\n",
    "\n",
    "One convienient feature, however, is that there *appears* to be only one snp listed for each entry identified by a position. Let's check this in a more systematic(ish) way:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "41916ea7-de8c-4a9e-92fa-02aafbf528d3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T18:19:40.888908Z",
     "iopub.status.busy": "2025-01-07T18:19:40.888450Z",
     "iopub.status.idle": "2025-01-07T18:19:41.892570Z",
     "shell.execute_reply": "2025-01-07T18:19:41.892144Z",
     "shell.execute_reply.started": "2025-01-07T18:19:40.888887Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>snps</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>chr11:92232901:G_GTA</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr10:103214974_AT_A</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr16:232396_AGAGT_A</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr22:37462936_A_G</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr3:184091102_T_G</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('chr11:92232901:G_GTA',),\n",
       " ('chr10:103214974_AT_A',),\n",
       " ('chr16:232396_AGAGT_A',),\n",
       " ('chr22:37462936_A_G',),\n",
       " ('chr3:184091102_T_G',)]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT snps\n",
    "FROM gwas_chrpos\n",
    "ORDER BY LENGTH(snps) DESC\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ae5c60b-026a-4818-ae31-50844a2c7c8f",
   "metadata": {},
   "source": [
    "Ok perfect. If there were any ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f75eb61-13f6-427c-bff6-9d2c5bbc4ecf",
   "metadata": {},
   "source": [
    "Making sure the genome versions of the gwas catalog and malinois/gnomad are compatable:\n",
    "\n",
    "> GWAS Catalog data is currently mapped to Genome Assembly GRCh38.p14 and dbSNP Build 156.\n",
    "\n",
    "[retrieved 2024-01-07](https://www.ebi.ac.uk/gwas/)\n",
    "\n",
    "Examining the header of `gnomad.genomes.v3.1.2.sites.chr21.converted.vcf.gz`\n",
    "\n",
    "> ##reference=GRCh38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "12cb03e8-859d-4640-9357-0d6a9a9c2d18",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-07T16:51:33.115999Z",
     "iopub.status.busy": "2025-01-07T16:51:33.115787Z",
     "iopub.status.idle": "2025-01-07T16:51:35.676805Z",
     "shell.execute_reply": "2025-01-07T16:51:35.676289Z",
     "shell.execute_reply.started": "2025-01-07T16:51:33.115983Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "waiting for server to shut down...."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 11:51:33.342 EST [1936922] LOG:  received fast shutdown request\n",
      "2025-01-07 11:51:33.354 EST [1936922] LOG:  aborting any active transactions\n",
      "2025-01-07 11:51:33.354 EST [1937168] FATAL:  terminating connection due to administrator command\n",
      "2025-01-07 11:51:33.359 EST [1936922] LOG:  background worker \"logical replication launcher\" (PID 1936930) exited with exit code 1\n",
      "2025-01-07 11:51:33.392 EST [1937147] FATAL:  terminating autovacuum process due to administrator command\n",
      "2025-01-07 11:51:33.392 EST [1937147] CONTEXT:  while scanning block 17916685 of relation \"public.malin_gnomad\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 11:51:34.805 EST [1936925] LOG:  checkpoint complete: wrote 84 buffers (0.5%); 0 WAL file(s) added, 0 removed, 5 recycled; write=34.045 s, sync=1.417 s, total=35.816 s; sync files=41, longest=1.291 s, average=0.035 s; distance=75968 kB, estimate=75968 kB; lsn=27/D164EA0, redo lsn=27/C8FF300\n",
      "2025-01-07 11:51:34.823 EST [1936925] LOG:  shutting down\n",
      "2025-01-07 11:51:34.836 EST [1936925] LOG:  checkpoint starting: shutdown immediate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " done\n",
      "server stopped\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-01-07 11:51:35.477 EST [1936925] LOG:  checkpoint complete: wrote 260 buffers (1.6%); 0 WAL file(s) added, 1 removed, 0 recycled; write=0.430 s, sync=0.032 s, total=0.654 s; sync files=2, longest=0.020 s, average=0.016 s; distance=8599 kB, estimate=69231 kB; lsn=27/D164F18, redo lsn=27/D164F18\n",
      "2025-01-07 11:51:35.528 EST [1936922] LOG:  database system is shut down\n"
     ]
    }
   ],
   "source": [
    "#stop the sql db\n",
    "!pg_ctl -D ~/palmer_scratch/db stop"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
