{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 20:56:54.827 EST [4169438] LOG:  starting PostgreSQL 17.2 on x86_64-conda-linux-gnu, compiled by x86_64-conda-linux-gnu-cc (Anaconda gcc) 11.2.0, 64-bit\n",
      "2025-02-10 20:56:54.827 EST [4169438] LOG:  listening on IPv6 address \"::1\", port 5433\n",
      "2025-02-10 20:56:54.828 EST [4169438] LOG:  listening on IPv4 address \"127.0.0.1\", port 5433\n",
      "2025-02-10 20:56:54.831 EST [4169438] LOG:  listening on Unix socket \"/tmp/.s.PGSQL.5433\"\n",
      "2025-02-10 20:56:54.975 EST [4169473] LOG:  database system was shut down at 2025-02-10 20:56:38 EST\n",
      "2025-02-10 20:56:55.098 EST [4169438] LOG:  database system is ready to accept connections\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:01:54.586 EST [4169471] LOG:  checkpoint starting: time\n",
      "2025-02-10 21:02:12.314 EST [4169471] LOG:  checkpoint complete: wrote 178 buffers (1.1%); 0 WAL file(s) added, 2 removed, 0 recycled; write=17.606 s, sync=0.007 s, total=17.728 s; sync files=17, longest=0.002 s, average=0.001 s; distance=36315 kB, estimate=36315 kB; lsn=91/5C9E2038, redo lsn=91/5C9E1F90\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import requests\n",
    "#start the database\n",
    "process = subprocess.Popen([\"postgres\", \"-D\", \"/home/mcn26/palmer_scratch/db\", \"-p\", \"5433\"])\n",
    "# Load the SQL extension\n",
    "%load_ext sql\n",
    "#dealing with a weird bug\n",
    "%config SqlMagic.style = '_DEPRECATED_DEFAULT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "postgresql://mr_root:password@localhost:5433/scratch"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our emvar definition is abs(skew)>=0.5 and max(alt activity, ref activity)>=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>pos</th>\n",
       "            <th>ref</th>\n",
       "            <th>alt</th>\n",
       "            <th>k562__ref</th>\n",
       "            <th>hepg2__ref</th>\n",
       "            <th>sknsh__ref</th>\n",
       "            <th>k562__alt</th>\n",
       "            <th>hepg2__alt</th>\n",
       "            <th>sknsh__alt</th>\n",
       "            <th>k562__skew</th>\n",
       "            <th>hepg2__skew</th>\n",
       "            <th>sknsh__skew</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.83</td>\n",
       "            <td>chr21</td>\n",
       "            <td>33060745</td>\n",
       "            <td>G</td>\n",
       "            <td>C</td>\n",
       "            <td>0.4658757</td>\n",
       "            <td>0.5130802</td>\n",
       "            <td>0.39136338</td>\n",
       "            <td>0.4494594</td>\n",
       "            <td>0.51362103</td>\n",
       "            <td>0.34879774</td>\n",
       "            <td>-0.016416281</td>\n",
       "            <td>0.0005408189</td>\n",
       "            <td>-0.042565674</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr21</td>\n",
       "            <td>33060745</td>\n",
       "            <td>G</td>\n",
       "            <td>C</td>\n",
       "            <td>0.4658757</td>\n",
       "            <td>0.5130802</td>\n",
       "            <td>0.39136338</td>\n",
       "            <td>0.4494594</td>\n",
       "            <td>0.51362103</td>\n",
       "            <td>0.34879774</td>\n",
       "            <td>-0.016416281</td>\n",
       "            <td>0.0005408189</td>\n",
       "            <td>-0.042565674</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10009710</td>\n",
       "            <td>0.71</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs12640763</td>\n",
       "            <td>0.74</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs1505771</td>\n",
       "            <td>0.98</td>\n",
       "            <td>chr4</td>\n",
       "            <td>144312789</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>0.32518756</td>\n",
       "            <td>1.1361839</td>\n",
       "            <td>0.45807314</td>\n",
       "            <td>0.3284431</td>\n",
       "            <td>1.1785058</td>\n",
       "            <td>0.46938372</td>\n",
       "            <td>0.0032555754</td>\n",
       "            <td>0.042322095</td>\n",
       "            <td>0.011310599</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 'rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs7275426', '0.83', 'chr21', 33060745, 'G', 'C', 0.4658757, 0.5130802, 0.39136338, 0.4494594, 0.51362103, 0.34879774, -0.016416281, 0.0005408189, -0.042565674),\n",
       " ('AFR', 'rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs1000005', 'tag', 'chr21', 33060745, 'G', 'C', 0.4658757, 0.5130802, 0.39136338, 0.4494594, 0.51362103, 0.34879774, -0.016416281, 0.0005408189, -0.042565674),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs10009710', '0.71', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs12640763', '0.74', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599),\n",
       " ('AFR', 'rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs1505771', '0.98', 'chr4', 144312789, 'A', 'T', 0.32518756, 1.1361839, 0.45807314, 0.3284431, 1.1785058, 0.46938372, 0.0032555754, 0.042322095, 0.011310599)]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM stacked_table\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:10:39.319 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:13.159 EST [4169471] LOG:  checkpoint complete: wrote 295 buffers (1.8%); 0 WAL file(s) added, 2 removed, 31 recycled; write=32.463 s, sync=0.221 s, total=33.841 s; sync files=14, longest=0.103 s, average=0.016 s; distance=534435 kB, estimate=534435 kB; lsn=91/9C70C228, redo lsn=91/7D3CAC98\n",
      "2025-02-10 21:11:13.191 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:19.036 EST [4169471] LOG:  checkpoint complete: wrote 1214 buffers (7.4%); 0 WAL file(s) added, 10 removed, 30 recycled; write=4.085 s, sync=0.482 s, total=5.858 s; sync files=6, longest=0.362 s, average=0.081 s; distance=657312 kB, estimate=657312 kB; lsn=91/C54AD2C8, redo lsn=91/A55B3080\n",
      "2025-02-10 21:11:19.102 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:11:19.102 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:19.102 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:24.711 EST [4169471] LOG:  checkpoint complete: wrote 1057 buffers (6.5%); 0 WAL file(s) added, 12 removed, 30 recycled; write=3.981 s, sync=0.398 s, total=5.609 s; sync files=5, longest=0.284 s, average=0.080 s; distance=695451 kB, estimate=695451 kB; lsn=91/EFB15E28, redo lsn=91/CFCD9E98\n",
      "2025-02-10 21:11:24.812 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:11:24.812 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:24.812 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:30.077 EST [4169471] LOG:  checkpoint complete: wrote 790 buffers (4.8%); 0 WAL file(s) added, 10 removed, 33 recycled; write=4.168 s, sync=0.152 s, total=5.266 s; sync files=5, longest=0.104 s, average=0.031 s; distance=692411 kB, estimate=695147 kB; lsn=92/188CE898, redo lsn=91/FA108E78\n",
      "2025-02-10 21:11:30.129 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:11:30.129 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:30.129 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:35.651 EST [4169471] LOG:  checkpoint complete: wrote 516 buffers (3.1%); 0 WAL file(s) added, 6 removed, 30 recycled; write=4.063 s, sync=0.331 s, total=5.522 s; sync files=7, longest=0.166 s, average=0.048 s; distance=599652 kB, estimate=685598 kB; lsn=92/3E4010C0, redo lsn=92/1EAA2090\n",
      "2025-02-10 21:11:35.716 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:11:35.716 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:35.730 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:41.333 EST [4169471] LOG:  checkpoint complete: wrote 674 buffers (4.1%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.418 s, sync=0.266 s, total=5.618 s; sync files=5, longest=0.175 s, average=0.054 s; distance=667133 kB, estimate=683751 kB; lsn=92/66769C30, redo lsn=92/47621858\n",
      "2025-02-10 21:11:41.406 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:11:41.406 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:41.406 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:46.603 EST [4169471] LOG:  checkpoint complete: wrote 341 buffers (2.1%); 0 WAL file(s) added, 4 removed, 32 recycled; write=3.824 s, sync=0.190 s, total=5.198 s; sync files=5, longest=0.118 s, average=0.038 s; distance=594327 kB, estimate=674809 kB; lsn=92/8A442E48, redo lsn=92/6BA874B0\n",
      "2025-02-10 21:11:46.693 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:11:46.693 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:46.693 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:52.791 EST [4169471] LOG:  checkpoint complete: wrote 393 buffers (2.4%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.671 s, sync=0.302 s, total=6.099 s; sync files=5, longest=0.173 s, average=0.061 s; distance=671546 kB, estimate=674482 kB; lsn=92/B461EDD0, redo lsn=92/94A55E80\n",
      "2025-02-10 21:11:52.792 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:11:52.792 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:52.816 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:11:58.470 EST [4169471] LOG:  checkpoint complete: wrote 1003 buffers (6.1%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.114 s, sync=0.233 s, total=5.679 s; sync files=5, longest=0.199 s, average=0.047 s; distance=658593 kB, estimate=672893 kB; lsn=92/DC38F100, redo lsn=92/BCD7E548\n",
      "2025-02-10 21:11:58.549 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:11:58.549 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:11:58.549 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:04.494 EST [4169471] LOG:  checkpoint complete: wrote 1554 buffers (9.5%); 0 WAL file(s) added, 11 removed, 32 recycled; write=4.050 s, sync=0.266 s, total=5.945 s; sync files=4, longest=0.188 s, average=0.067 s; distance=690913 kB, estimate=690913 kB; lsn=93/541C148, redo lsn=92/E7036A28\n",
      "2025-02-10 21:12:04.597 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:04.597 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:04.597 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:10.946 EST [4169471] LOG:  checkpoint complete: wrote 576 buffers (3.5%); 0 WAL file(s) added, 12 removed, 31 recycled; write=4.507 s, sync=0.261 s, total=6.349 s; sync files=4, longest=0.168 s, average=0.066 s; distance=715644 kB, estimate=715644 kB; lsn=93/32357B58, redo lsn=93/12B15B50\n",
      "2025-02-10 21:12:11.040 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:12:11.040 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:11.040 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:17.159 EST [4169471] LOG:  checkpoint complete: wrote 335 buffers (2.0%); 0 WAL file(s) added, 14 removed, 31 recycled; write=4.309 s, sync=0.521 s, total=6.119 s; sync files=6, longest=0.309 s, average=0.087 s; distance=733800 kB, estimate=733800 kB; lsn=93/5E73E6A0, redo lsn=93/3F7AFE68\n",
      "2025-02-10 21:12:17.159 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:17.159 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:17.171 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:23.116 EST [4169471] LOG:  checkpoint complete: wrote 921 buffers (5.6%); 0 WAL file(s) added, 8 removed, 32 recycled; write=4.372 s, sync=0.242 s, total=5.957 s; sync files=4, longest=0.146 s, average=0.061 s; distance=654038 kB, estimate=725824 kB; lsn=93/85D562E0, redo lsn=93/67665860\n",
      "2025-02-10 21:12:23.215 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:23.215 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:23.252 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:29.440 EST [4169471] LOG:  checkpoint complete: wrote 649 buffers (4.0%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.591 s, sync=0.273 s, total=6.226 s; sync files=5, longest=0.162 s, average=0.055 s; distance=680383 kB, estimate=721280 kB; lsn=93/AFEFC730, redo lsn=93/90ED55D0\n",
      "2025-02-10 21:12:29.515 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:29.515 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:29.524 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:35.315 EST [4169471] LOG:  checkpoint complete: wrote 1021 buffers (6.2%); 0 WAL file(s) added, 8 removed, 32 recycled; write=4.328 s, sync=0.297 s, total=5.800 s; sync files=5, longest=0.202 s, average=0.060 s; distance=644773 kB, estimate=713629 kB; lsn=93/D6ECCE20, redo lsn=93/B847EAB8\n",
      "2025-02-10 21:12:35.400 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:35.400 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:35.411 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:40.998 EST [4169471] LOG:  checkpoint complete: wrote 879 buffers (5.4%); 0 WAL file(s) added, 8 removed, 32 recycled; write=3.999 s, sync=0.218 s, total=5.598 s; sync files=7, longest=0.128 s, average=0.032 s; distance=657166 kB, estimate=707983 kB; lsn=93/FF4384D8, redo lsn=93/E0642520\n",
      "2025-02-10 21:12:41.100 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:41.100 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:41.100 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:46.699 EST [4169471] LOG:  checkpoint complete: wrote 313 buffers (1.9%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.084 s, sync=0.266 s, total=5.599 s; sync files=4, longest=0.172 s, average=0.067 s; distance=664106 kB, estimate=703595 kB; lsn=94/2863AE70, redo lsn=94/8ECCF18\n",
      "2025-02-10 21:12:46.808 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:12:46.808 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:46.809 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:12:52.340 EST [4169471] LOG:  checkpoint complete: wrote 1122 buffers (6.8%); 0 WAL file(s) added, 11 removed, 31 recycled; write=4.104 s, sync=0.267 s, total=5.532 s; sync files=4, longest=0.163 s, average=0.067 s; distance=681986 kB, estimate=701434 kB; lsn=94/5212A470, redo lsn=94/328CD900\n",
      "2025-02-10 21:12:52.407 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:12:52.407 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:12:52.416 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE stacked_table ADD COLUMN k562_emvar BOOLEAN;\n",
    "UPDATE stacked_table\n",
    "SET k562_emvar = (\n",
    "    CASE\n",
    "        WHEN GREATEST(k562__ref, k562__alt) >= 1 AND ABS(k562__skew) > 0.5 THEN TRUE\n",
    "        ELSE FALSE\n",
    "    END\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:13:36.496 EST [4178338] ERROR:  canceling autovacuum task\n",
      "2025-02-10 21:13:36.496 EST [4178338] CONTEXT:  while scanning block 148515 of relation \"public.stacked_table\"\n",
      "\tautomatic vacuum of table \"scratch.public.stacked_table\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:13:53.889 EST [4169471] LOG:  checkpoint complete: wrote 760 buffers (4.6%); 0 WAL file(s) added, 9 removed, 31 recycled; write=59.665 s, sync=0.577 s, total=61.482 s; sync files=13, longest=0.418 s, average=0.045 s; distance=650714 kB, estimate=696362 kB; lsn=94/799984B0, redo lsn=94/5A444490\n",
      "2025-02-10 21:13:53.982 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:13:59.972 EST [4169471] LOG:  checkpoint complete: wrote 1232 buffers (7.5%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.189 s, sync=0.468 s, total=6.023 s; sync files=6, longest=0.300 s, average=0.078 s; distance=673808 kB, estimate=694107 kB; lsn=94/A2B373B0, redo lsn=94/836486B0\n",
      "2025-02-10 21:14:00.059 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:14:00.059 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:00.065 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:06.073 EST [4169471] LOG:  checkpoint complete: wrote 1408 buffers (8.6%); 0 WAL file(s) added, 8 removed, 32 recycled; write=4.295 s, sync=0.052 s, total=6.014 s; sync files=3, longest=0.032 s, average=0.018 s; distance=654818 kB, estimate=690178 kB; lsn=94/C96580A0, redo lsn=94/AB5C0FE8\n",
      "2025-02-10 21:14:06.190 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:14:06.190 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:06.190 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:12.700 EST [4169471] LOG:  checkpoint complete: wrote 235 buffers (1.4%); 0 WAL file(s) added, 10 removed, 32 recycled; write=4.577 s, sync=0.335 s, total=6.510 s; sync files=3, longest=0.201 s, average=0.112 s; distance=694007 kB, estimate=694007 kB; lsn=94/F47BF1A8, redo lsn=94/D5B7EF18\n",
      "2025-02-10 21:14:12.859 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:14:12.859 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:12.886 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:20.076 EST [4169471] LOG:  checkpoint complete: wrote 585 buffers (3.6%); 0 WAL file(s) added, 11 removed, 31 recycled; write=4.832 s, sync=0.474 s, total=7.217 s; sync files=7, longest=0.195 s, average=0.068 s; distance=682573 kB, estimate=692864 kB; lsn=95/1F4D1238, redo lsn=94/FF6124A8\n",
      "2025-02-10 21:14:20.121 EST [4169471] LOG:  checkpoints are occurring too frequently (8 seconds apart)\n",
      "2025-02-10 21:14:20.121 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:20.121 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:27.266 EST [4169471] LOG:  checkpoint complete: wrote 269 buffers (1.6%); 0 WAL file(s) added, 14 removed, 30 recycled; write=5.390 s, sync=0.427 s, total=7.145 s; sync files=6, longest=0.258 s, average=0.072 s; distance=718690 kB, estimate=718690 kB; lsn=95/4B639040, redo lsn=95/2B3EB008\n",
      "2025-02-10 21:14:27.412 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:14:27.412 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:27.412 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:34.677 EST [4169471] LOG:  checkpoint complete: wrote 158 buffers (1.0%); 0 WAL file(s) added, 11 removed, 31 recycled; write=5.285 s, sync=0.581 s, total=7.266 s; sync files=5, longest=0.382 s, average=0.117 s; distance=684188 kB, estimate=715240 kB; lsn=95/754913E0, redo lsn=95/550120D8\n",
      "2025-02-10 21:14:34.854 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:14:34.854 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:34.864 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:41.956 EST [4169471] LOG:  checkpoint complete: wrote 188 buffers (1.1%); 0 WAL file(s) added, 10 removed, 31 recycled; write=5.247 s, sync=0.329 s, total=7.102 s; sync files=6, longest=0.148 s, average=0.055 s; distance=685005 kB, estimate=712217 kB; lsn=95/9E6740B8, redo lsn=95/7ED054F0\n",
      "2025-02-10 21:14:41.994 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:14:41.994 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:42.044 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:48.843 EST [4169471] LOG:  checkpoint complete: wrote 585 buffers (3.6%); 0 WAL file(s) added, 11 removed, 31 recycled; write=5.053 s, sync=0.392 s, total=6.849 s; sync files=7, longest=0.188 s, average=0.056 s; distance=680645 kB, estimate=709059 kB; lsn=95/C7739B10, redo lsn=95/A85B6928\n",
      "2025-02-10 21:14:49.046 EST [4169471] LOG:  checkpoints are occurring too frequently (8 seconds apart)\n",
      "2025-02-10 21:14:49.046 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:49.076 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:14:55.928 EST [4169471] LOG:  checkpoint complete: wrote 204 buffers (1.2%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.873 s, sync=0.368 s, total=6.882 s; sync files=5, longest=0.215 s, average=0.074 s; distance=678340 kB, estimate=705987 kB; lsn=95/F06FE438, redo lsn=95/D1C27940\n",
      "2025-02-10 21:14:56.003 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:14:56.003 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:14:56.003 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:02.464 EST [4169471] LOG:  checkpoint complete: wrote 242 buffers (1.5%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.726 s, sync=0.389 s, total=6.462 s; sync files=6, longest=0.199 s, average=0.065 s; distance=653194 kB, estimate=700708 kB; lsn=96/18897F58, redo lsn=95/F9A0A4C0\n",
      "2025-02-10 21:15:02.648 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:15:02.648 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:02.686 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:09.431 EST [4169471] LOG:  checkpoint complete: wrote 562 buffers (3.4%); 0 WAL file(s) added, 12 removed, 29 recycled; write=5.222 s, sync=0.414 s, total=6.783 s; sync files=7, longest=0.155 s, average=0.060 s; distance=674866 kB, estimate=698124 kB; lsn=96/43637A30, redo lsn=96/22D16DD0\n",
      "2025-02-10 21:15:09.471 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:15:09.471 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:09.493 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:16.030 EST [4169471] LOG:  checkpoint complete: wrote 314 buffers (1.9%); 0 WAL file(s) added, 8 removed, 32 recycled; write=5.016 s, sync=0.285 s, total=6.560 s; sync files=5, longest=0.164 s, average=0.057 s; distance=648332 kB, estimate=693145 kB; lsn=96/690CE700, redo lsn=96/4A639E78\n",
      "2025-02-10 21:15:16.060 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:15:16.060 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:16.074 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:22.432 EST [4169471] LOG:  checkpoint complete: wrote 720 buffers (4.4%); 0 WAL file(s) added, 6 removed, 32 recycled; write=4.770 s, sync=0.398 s, total=6.372 s; sync files=5, longest=0.222 s, average=0.080 s; distance=619140 kB, estimate=685744 kB; lsn=96/8F66C720, redo lsn=96/702DB128\n",
      "2025-02-10 21:15:22.593 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:15:22.593 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:22.594 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:29.063 EST [4169471] LOG:  checkpoint complete: wrote 287 buffers (1.8%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.758 s, sync=0.446 s, total=6.471 s; sync files=6, longest=0.272 s, average=0.075 s; distance=631552 kB, estimate=680325 kB; lsn=96/B59ECB18, redo lsn=96/96B9B3C0\n",
      "2025-02-10 21:15:29.195 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:15:29.195 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:29.199 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:35.642 EST [4169471] LOG:  checkpoint complete: wrote 214 buffers (1.3%); 0 WAL file(s) added, 9 removed, 30 recycled; write=4.684 s, sync=0.530 s, total=6.448 s; sync files=7, longest=0.274 s, average=0.076 s; distance=640156 kB, estimate=676308 kB; lsn=96/DE037638, redo lsn=96/BDCC2788\n",
      "2025-02-10 21:15:35.677 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:15:35.677 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:35.677 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:41.798 EST [4169471] LOG:  checkpoint complete: wrote 245 buffers (1.5%); 0 WAL file(s) added, 10 removed, 32 recycled; write=4.597 s, sync=0.294 s, total=6.121 s; sync files=5, longest=0.177 s, average=0.059 s; distance=677210 kB, estimate=677210 kB; lsn=97/68EC550, redo lsn=96/E7219080\n",
      "2025-02-10 21:15:41.894 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:15:41.894 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:41.928 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:15:47.674 EST [4169471] LOG:  checkpoint complete: wrote 240 buffers (1.5%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.265 s, sync=0.356 s, total=5.780 s; sync files=5, longest=0.245 s, average=0.072 s; distance=660405 kB, estimate=675529 kB; lsn=97/2F54E508, redo lsn=97/F7065A8\n",
      "2025-02-10 21:15:47.795 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:15:47.795 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:15:47.795 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE stacked_table ADD COLUMN hepg2_emvar BOOLEAN;\n",
    "UPDATE stacked_table\n",
    "SET hepg2_emvar = (\n",
    "    CASE\n",
    "        WHEN GREATEST(hepg2__ref, hepg2__alt) >= 1 AND ABS(hepg2__skew) > 0.5 THEN TRUE\n",
    "        ELSE FALSE\n",
    "    END\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:15:52.384 EST [4178639] ERROR:  canceling autovacuum task\n",
      "2025-02-10 21:15:52.384 EST [4178639] CONTEXT:  while scanning block 185217 of relation \"public.stacked_table\"\n",
      "\tautomatic vacuum of table \"scratch.public.stacked_table\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:16:03.102 EST [4169471] LOG:  checkpoint complete: wrote 165 buffers (1.0%); 0 WAL file(s) added, 7 removed, 33 recycled; write=13.284 s, sync=0.851 s, total=15.308 s; sync files=12, longest=0.778 s, average=0.071 s; distance=662124 kB, estimate=674189 kB; lsn=97/55F61390, redo lsn=97/37DA1960\n",
      "2025-02-10 21:16:05.927 EST [4169471] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 21:16:05.927 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:05.927 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:23.247 EST [4169471] LOG:  checkpoint complete: wrote 55 buffers (0.3%); 0 WAL file(s) added, 2 removed, 31 recycled; write=15.814 s, sync=0.487 s, total=17.320 s; sync files=15, longest=0.139 s, average=0.033 s; distance=527238 kB, estimate=659494 kB; lsn=97/7864A390, redo lsn=97/58083378\n",
      "2025-02-10 21:16:23.248 EST [4169471] LOG:  checkpoints are occurring too frequently (18 seconds apart)\n",
      "2025-02-10 21:16:23.248 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:23.259 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:29.257 EST [4169471] LOG:  checkpoint complete: wrote 754 buffers (4.6%); 0 WAL file(s) added, 8 removed, 31 recycled; write=4.321 s, sync=0.335 s, total=6.010 s; sync files=4, longest=0.135 s, average=0.084 s; distance=643584 kB, estimate=657903 kB; lsn=97/9ECD83B0, redo lsn=97/7F5036D0\n",
      "2025-02-10 21:16:29.373 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:29.373 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:29.394 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:35.206 EST [4169471] LOG:  checkpoint complete: wrote 410 buffers (2.5%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.295 s, sync=0.330 s, total=5.833 s; sync files=4, longest=0.168 s, average=0.083 s; distance=660301 kB, estimate=660301 kB; lsn=97/C6934D78, redo lsn=97/A79D6BC8\n",
      "2025-02-10 21:16:35.340 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:35.340 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:35.340 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:41.527 EST [4169471] LOG:  checkpoint complete: wrote 41 buffers (0.3%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.619 s, sync=0.385 s, total=6.187 s; sync files=4, longest=0.181 s, average=0.097 s; distance=651576 kB, estimate=659428 kB; lsn=97/EF2085A8, redo lsn=97/CF624EE8\n",
      "2025-02-10 21:16:41.586 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:41.586 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:41.597 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:47.179 EST [4169471] LOG:  checkpoint complete: wrote 213 buffers (1.3%); 0 WAL file(s) added, 8 removed, 31 recycled; write=4.008 s, sync=0.485 s, total=5.594 s; sync files=6, longest=0.293 s, average=0.081 s; distance=635278 kB, estimate=657013 kB; lsn=98/164090C8, redo lsn=97/F6288848\n",
      "2025-02-10 21:16:47.221 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:47.221 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:47.221 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:53.262 EST [4169471] LOG:  checkpoint complete: wrote 43 buffers (0.3%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.608 s, sync=0.350 s, total=6.042 s; sync files=4, longest=0.198 s, average=0.088 s; distance=655349 kB, estimate=656847 kB; lsn=98/3D61BAD8, redo lsn=98/1E285DE0\n",
      "2025-02-10 21:16:53.404 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:53.404 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:53.412 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:16:59.412 EST [4169471] LOG:  checkpoint complete: wrote 201 buffers (1.2%); 0 WAL file(s) added, 6 removed, 31 recycled; write=4.592 s, sync=0.329 s, total=6.008 s; sync files=5, longest=0.189 s, average=0.066 s; distance=613522 kB, estimate=652514 kB; lsn=98/62E74900, redo lsn=98/439AA950\n",
      "2025-02-10 21:16:59.538 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:16:59.538 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:16:59.569 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:05.319 EST [4169471] LOG:  checkpoint complete: wrote 196 buffers (1.2%); 0 WAL file(s) added, 8 removed, 31 recycled; write=4.522 s, sync=0.342 s, total=5.782 s; sync files=5, longest=0.235 s, average=0.069 s; distance=635624 kB, estimate=650825 kB; lsn=98/8935D748, redo lsn=98/6A6649F8\n",
      "2025-02-10 21:17:05.421 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:05.421 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:05.425 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:10.618 EST [4169471] LOG:  checkpoint complete: wrote 268 buffers (1.6%); 0 WAL file(s) added, 5 removed, 32 recycled; write=4.082 s, sync=0.270 s, total=5.198 s; sync files=7, longest=0.168 s, average=0.039 s; distance=602859 kB, estimate=646029 kB; lsn=98/AE6FF390, redo lsn=98/8F31F648\n",
      "2025-02-10 21:17:10.655 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:17:10.655 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:10.655 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:16.664 EST [4169471] LOG:  checkpoint complete: wrote 330 buffers (2.0%); 0 WAL file(s) added, 6 removed, 31 recycled; write=4.454 s, sync=0.239 s, total=6.010 s; sync files=5, longest=0.155 s, average=0.048 s; distance=613409 kB, estimate=642767 kB; lsn=98/D44EED38, redo lsn=98/B4A27DA0\n",
      "2025-02-10 21:17:16.746 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:16.746 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:16.755 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:23.141 EST [4169471] LOG:  checkpoint complete: wrote 305 buffers (1.9%); 0 WAL file(s) added, 12 removed, 31 recycled; write=4.699 s, sync=0.291 s, total=6.395 s; sync files=6, longest=0.150 s, average=0.049 s; distance=697064 kB, estimate=697064 kB; lsn=98/FE4BE410, redo lsn=98/DF2E2080\n",
      "2025-02-10 21:17:23.248 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:17:23.248 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:23.248 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:29.433 EST [4169471] LOG:  checkpoint complete: wrote 172 buffers (1.0%); 0 WAL file(s) added, 11 removed, 30 recycled; write=4.759 s, sync=0.338 s, total=6.186 s; sync files=7, longest=0.140 s, average=0.049 s; distance=680426 kB, estimate=695400 kB; lsn=99/28653E28, redo lsn=99/8B5C8A8\n",
      "2025-02-10 21:17:29.463 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:29.463 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:29.463 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:35.281 EST [4169471] LOG:  checkpoint complete: wrote 342 buffers (2.1%); 0 WAL file(s) added, 10 removed, 30 recycled; write=4.463 s, sync=0.255 s, total=5.818 s; sync files=6, longest=0.151 s, average=0.043 s; distance=657818 kB, estimate=691642 kB; lsn=99/50DBBF78, redo lsn=99/30DC3230\n",
      "2025-02-10 21:17:35.385 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:35.385 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:35.385 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:41.598 EST [4169471] LOG:  checkpoint complete: wrote 177 buffers (1.1%); 0 WAL file(s) added, 11 removed, 29 recycled; write=4.626 s, sync=0.341 s, total=6.213 s; sync files=5, longest=0.155 s, average=0.069 s; distance=653473 kB, estimate=687825 kB; lsn=99/795FE078, redo lsn=99/58BEB8A0\n",
      "2025-02-10 21:17:41.643 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:41.643 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:41.643 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:47.781 EST [4169471] LOG:  checkpoint complete: wrote 197 buffers (1.2%); 0 WAL file(s) added, 11 removed, 31 recycled; write=4.841 s, sync=0.336 s, total=6.138 s; sync files=5, longest=0.189 s, average=0.068 s; distance=687568 kB, estimate=687799 kB; lsn=99/A1F7C530, redo lsn=99/82B5F948\n",
      "2025-02-10 21:17:47.819 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:47.819 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:47.819 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:53.464 EST [4169471] LOG:  checkpoint complete: wrote 203 buffers (1.2%); 0 WAL file(s) added, 8 removed, 30 recycled; write=4.425 s, sync=0.356 s, total=5.645 s; sync files=7, longest=0.146 s, average=0.051 s; distance=622704 kB, estimate=681290 kB; lsn=99/C94E3EA0, redo lsn=99/A8B7BCE8\n",
      "2025-02-10 21:17:53.485 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:53.485 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:53.485 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:17:59.537 EST [4169471] LOG:  checkpoint complete: wrote 716 buffers (4.4%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.500 s, sync=0.381 s, total=6.052 s; sync files=5, longest=0.198 s, average=0.077 s; distance=621245 kB, estimate=675285 kB; lsn=99/EDA1DC90, redo lsn=99/CEA2B2E0\n",
      "2025-02-10 21:17:59.725 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:17:59.725 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:17:59.725 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:18:05.584 EST [4169471] LOG:  checkpoint complete: wrote 255 buffers (1.6%); 0 WAL file(s) added, 9 removed, 30 recycled; write=4.445 s, sync=0.308 s, total=5.860 s; sync files=5, longest=0.158 s, average=0.062 s; distance=641368 kB, estimate=671894 kB; lsn=9A/1563E078, redo lsn=99/F5C81560\n",
      "2025-02-10 21:18:05.735 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:18:05.735 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:18:05.735 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:18:11.322 EST [4169471] LOG:  checkpoint complete: wrote 234 buffers (1.4%); 0 WAL file(s) added, 8 removed, 32 recycled; write=4.282 s, sync=0.342 s, total=5.587 s; sync files=5, longest=0.160 s, average=0.069 s; distance=645950 kB, estimate=669299 kB; lsn=9A/3BACC450, redo lsn=9A/1D3510E8\n",
      "2025-02-10 21:18:11.463 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:18:11.463 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:18:11.463 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE stacked_table ADD COLUMN sknsh_emvar BOOLEAN;\n",
    "UPDATE stacked_table\n",
    "SET sknsh_emvar = (\n",
    "    CASE\n",
    "        WHEN GREATEST(sknsh__ref, sknsh__alt) >= 1 AND ABS(sknsh__skew) > 0.5 THEN TRUE\n",
    "        ELSE FALSE\n",
    "    END\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "20 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>rsid</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>pos</th>\n",
       "            <th>ref</th>\n",
       "            <th>alt</th>\n",
       "            <th>k562__ref</th>\n",
       "            <th>hepg2__ref</th>\n",
       "            <th>sknsh__ref</th>\n",
       "            <th>k562__alt</th>\n",
       "            <th>hepg2__alt</th>\n",
       "            <th>sknsh__alt</th>\n",
       "            <th>k562__skew</th>\n",
       "            <th>hepg2__skew</th>\n",
       "            <th>sknsh__skew</th>\n",
       "            <th>k562_emvar</th>\n",
       "            <th>hepg2_emvar</th>\n",
       "            <th>sknsh_emvar</th>\n",
       "            <th>any_emvar</th>\n",
       "            <th>all_emvar</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10188036</td>\n",
       "            <td>31490055</td>\n",
       "            <td>A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction</td>\n",
       "            <td>Cocaine use disorder x change in residence interaction</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>rs10188036</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr2</td>\n",
       "            <td>201391971</td>\n",
       "            <td>T</td>\n",
       "            <td>C</td>\n",
       "            <td>0.8448316</td>\n",
       "            <td>0.36654082</td>\n",
       "            <td>0.2542696</td>\n",
       "            <td>1.465271</td>\n",
       "            <td>0.53526527</td>\n",
       "            <td>0.28370926</td>\n",
       "            <td>0.62043923</td>\n",
       "            <td>0.16872439</td>\n",
       "            <td>0.029439654</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10188036</td>\n",
       "            <td>31490055</td>\n",
       "            <td>A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction</td>\n",
       "            <td>Cocaine use disorder x change in residence interaction</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>rs10194931</td>\n",
       "            <td>0.71</td>\n",
       "            <td>chr2</td>\n",
       "            <td>201391971</td>\n",
       "            <td>T</td>\n",
       "            <td>C</td>\n",
       "            <td>0.8448316</td>\n",
       "            <td>0.36654082</td>\n",
       "            <td>0.2542696</td>\n",
       "            <td>1.465271</td>\n",
       "            <td>0.53526527</td>\n",
       "            <td>0.28370926</td>\n",
       "            <td>0.62043923</td>\n",
       "            <td>0.16872439</td>\n",
       "            <td>0.029439654</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10188036</td>\n",
       "            <td>31490055</td>\n",
       "            <td>A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction</td>\n",
       "            <td>Cocaine use disorder x change in residence interaction</td>\n",
       "            <td>3E-8</td>\n",
       "            <td>rs13403153</td>\n",
       "            <td>0.9</td>\n",
       "            <td>chr2</td>\n",
       "            <td>201391971</td>\n",
       "            <td>T</td>\n",
       "            <td>C</td>\n",
       "            <td>0.8448316</td>\n",
       "            <td>0.36654082</td>\n",
       "            <td>0.2542696</td>\n",
       "            <td>1.465271</td>\n",
       "            <td>0.53526527</td>\n",
       "            <td>0.28370926</td>\n",
       "            <td>0.62043923</td>\n",
       "            <td>0.16872439</td>\n",
       "            <td>0.029439654</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>EUR</td>\n",
       "            <td>rs2270894</td>\n",
       "            <td>30595370</td>\n",
       "            <td>Leveraging Polygenic Functional Enrichment to Improve GWAS Power.</td>\n",
       "            <td>Height</td>\n",
       "            <td>2E-80</td>\n",
       "            <td>rs2270894</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr3</td>\n",
       "            <td>9933702</td>\n",
       "            <td>C</td>\n",
       "            <td>G</td>\n",
       "            <td>4.2506876</td>\n",
       "            <td>3.6712945</td>\n",
       "            <td>3.5454733</td>\n",
       "            <td>5.0015254</td>\n",
       "            <td>4.239152</td>\n",
       "            <td>4.947809</td>\n",
       "            <td>0.7508383</td>\n",
       "            <td>0.56785756</td>\n",
       "            <td>1.4023354</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Monocyte count</td>\n",
       "            <td>9E-23</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Mean platelet volume</td>\n",
       "            <td>2E-15</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Mean platelet volume</td>\n",
       "            <td>2E-15</td>\n",
       "            <td>rs9941452</td>\n",
       "            <td>1</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Monocyte count</td>\n",
       "            <td>2E-20</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Monocyte count</td>\n",
       "            <td>2E-20</td>\n",
       "            <td>rs9941452</td>\n",
       "            <td>1</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs1003393</td>\n",
       "            <td>32888493</td>\n",
       "            <td>Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.</td>\n",
       "            <td>Monocyte count</td>\n",
       "            <td>9E-23</td>\n",
       "            <td>rs9941452</td>\n",
       "            <td>1</td>\n",
       "            <td>chr19</td>\n",
       "            <td>13101400</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>1.8323534</td>\n",
       "            <td>1.5378494</td>\n",
       "            <td>1.0421877</td>\n",
       "            <td>1.3106359</td>\n",
       "            <td>1.2315954</td>\n",
       "            <td>0.8987868</td>\n",
       "            <td>-0.5217174</td>\n",
       "            <td>-0.30625406</td>\n",
       "            <td>-0.14340092</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10034452</td>\n",
       "            <td>36224396</td>\n",
       "            <td>A saturated map of common genetic variants associated with human height.</td>\n",
       "            <td>Height</td>\n",
       "            <td>9E-7</td>\n",
       "            <td>rs59341143</td>\n",
       "            <td>0.91</td>\n",
       "            <td>chr4</td>\n",
       "            <td>13529200</td>\n",
       "            <td>T</td>\n",
       "            <td>G</td>\n",
       "            <td>2.5369117</td>\n",
       "            <td>1.9763436</td>\n",
       "            <td>1.4932429</td>\n",
       "            <td>1.6715355</td>\n",
       "            <td>1.3225569</td>\n",
       "            <td>0.8809182</td>\n",
       "            <td>-0.8653763</td>\n",
       "            <td>-0.65378684</td>\n",
       "            <td>-0.61232454</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10034452</td>\n",
       "            <td>36224396</td>\n",
       "            <td>A saturated map of common genetic variants associated with human height.</td>\n",
       "            <td>Height</td>\n",
       "            <td>9E-7</td>\n",
       "            <td>rs9991833</td>\n",
       "            <td>0.76</td>\n",
       "            <td>chr4</td>\n",
       "            <td>13529200</td>\n",
       "            <td>T</td>\n",
       "            <td>G</td>\n",
       "            <td>2.5369117</td>\n",
       "            <td>1.9763436</td>\n",
       "            <td>1.4932429</td>\n",
       "            <td>1.6715355</td>\n",
       "            <td>1.3225569</td>\n",
       "            <td>0.8809182</td>\n",
       "            <td>-0.8653763</td>\n",
       "            <td>-0.65378684</td>\n",
       "            <td>-0.61232454</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10034452</td>\n",
       "            <td>36224396</td>\n",
       "            <td>A saturated map of common genetic variants associated with human height.</td>\n",
       "            <td>Height</td>\n",
       "            <td>9E-7</td>\n",
       "            <td>rs10020715</td>\n",
       "            <td>0.79</td>\n",
       "            <td>chr4</td>\n",
       "            <td>13529200</td>\n",
       "            <td>T</td>\n",
       "            <td>G</td>\n",
       "            <td>2.5369117</td>\n",
       "            <td>1.9763436</td>\n",
       "            <td>1.4932429</td>\n",
       "            <td>1.6715355</td>\n",
       "            <td>1.3225569</td>\n",
       "            <td>0.8809182</td>\n",
       "            <td>-0.8653763</td>\n",
       "            <td>-0.65378684</td>\n",
       "            <td>-0.61232454</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10034452</td>\n",
       "            <td>36224396</td>\n",
       "            <td>A saturated map of common genetic variants associated with human height.</td>\n",
       "            <td>Height</td>\n",
       "            <td>9E-7</td>\n",
       "            <td>rs10034452</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr4</td>\n",
       "            <td>13529200</td>\n",
       "            <td>T</td>\n",
       "            <td>G</td>\n",
       "            <td>2.5369117</td>\n",
       "            <td>1.9763436</td>\n",
       "            <td>1.4932429</td>\n",
       "            <td>1.6715355</td>\n",
       "            <td>1.3225569</td>\n",
       "            <td>0.8809182</td>\n",
       "            <td>-0.8653763</td>\n",
       "            <td>-0.65378684</td>\n",
       "            <td>-0.61232454</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "            <td>True</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>36914875</td>\n",
       "            <td>Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.</td>\n",
       "            <td>Lung function (FEV1/FVC)</td>\n",
       "            <td>2E-9</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>tag</td>\n",
       "            <td>chr2</td>\n",
       "            <td>86223017</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>0.53966457</td>\n",
       "            <td>0.6494023</td>\n",
       "            <td>0.14900297</td>\n",
       "            <td>1.3617532</td>\n",
       "            <td>0.5386045</td>\n",
       "            <td>0.08207319</td>\n",
       "            <td>0.8220887</td>\n",
       "            <td>-0.11079783</td>\n",
       "            <td>-0.06692975</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>36914875</td>\n",
       "            <td>Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.</td>\n",
       "            <td>Lung function (FEV1/FVC)</td>\n",
       "            <td>2E-9</td>\n",
       "            <td>rs10901</td>\n",
       "            <td>0.81</td>\n",
       "            <td>chr2</td>\n",
       "            <td>86223017</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>0.53966457</td>\n",
       "            <td>0.6494023</td>\n",
       "            <td>0.14900297</td>\n",
       "            <td>1.3617532</td>\n",
       "            <td>0.5386045</td>\n",
       "            <td>0.08207319</td>\n",
       "            <td>0.8220887</td>\n",
       "            <td>-0.11079783</td>\n",
       "            <td>-0.06692975</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>36914875</td>\n",
       "            <td>Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.</td>\n",
       "            <td>Lung function (FEV1/FVC)</td>\n",
       "            <td>2E-9</td>\n",
       "            <td>rs12714178</td>\n",
       "            <td>0.75</td>\n",
       "            <td>chr2</td>\n",
       "            <td>86223017</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>0.53966457</td>\n",
       "            <td>0.6494023</td>\n",
       "            <td>0.14900297</td>\n",
       "            <td>1.3617532</td>\n",
       "            <td>0.5386045</td>\n",
       "            <td>0.08207319</td>\n",
       "            <td>0.8220887</td>\n",
       "            <td>-0.11079783</td>\n",
       "            <td>-0.06692975</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>36914875</td>\n",
       "            <td>Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.</td>\n",
       "            <td>Lung function (FEV1/FVC)</td>\n",
       "            <td>2E-9</td>\n",
       "            <td>rs13007807</td>\n",
       "            <td>0.81</td>\n",
       "            <td>chr2</td>\n",
       "            <td>86223017</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>0.53966457</td>\n",
       "            <td>0.6494023</td>\n",
       "            <td>0.14900297</td>\n",
       "            <td>1.3617532</td>\n",
       "            <td>0.5386045</td>\n",
       "            <td>0.08207319</td>\n",
       "            <td>0.8220887</td>\n",
       "            <td>-0.11079783</td>\n",
       "            <td>-0.06692975</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10195199</td>\n",
       "            <td>36914875</td>\n",
       "            <td>Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.</td>\n",
       "            <td>Lung function (FEV1/FVC)</td>\n",
       "            <td>2E-9</td>\n",
       "            <td>rs13035593</td>\n",
       "            <td>0.81</td>\n",
       "            <td>chr2</td>\n",
       "            <td>86223017</td>\n",
       "            <td>C</td>\n",
       "            <td>T</td>\n",
       "            <td>0.53966457</td>\n",
       "            <td>0.6494023</td>\n",
       "            <td>0.14900297</td>\n",
       "            <td>1.3617532</td>\n",
       "            <td>0.5386045</td>\n",
       "            <td>0.08207319</td>\n",
       "            <td>0.8220887</td>\n",
       "            <td>-0.11079783</td>\n",
       "            <td>-0.06692975</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10830947</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Smoking initiation</td>\n",
       "            <td>2E-17</td>\n",
       "            <td>rs3024236</td>\n",
       "            <td>0.95</td>\n",
       "            <td>chr11</td>\n",
       "            <td>92792449</td>\n",
       "            <td>G</td>\n",
       "            <td>A</td>\n",
       "            <td>2.1934614</td>\n",
       "            <td>0.6742889</td>\n",
       "            <td>-0.08360493</td>\n",
       "            <td>1.3143308</td>\n",
       "            <td>0.2776651</td>\n",
       "            <td>-0.22676457</td>\n",
       "            <td>-0.87913024</td>\n",
       "            <td>-0.39662376</td>\n",
       "            <td>-0.14315963</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>True</td>\n",
       "            <td>False</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 'rs10188036', 31490055, 'A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction', 'Cocaine use disorder x change in residence interaction', '3E-8', 'rs10188036', 'tag', 'chr2', 201391971, 'T', 'C', 0.8448316, 0.36654082, 0.2542696, 1.465271, 0.53526527, 0.28370926, 0.62043923, 0.16872439, 0.029439654, True, False, False, True, False),\n",
       " ('AFR', 'rs10188036', 31490055, 'A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction', 'Cocaine use disorder x change in residence interaction', '3E-8', 'rs10194931', '0.71', 'chr2', 201391971, 'T', 'C', 0.8448316, 0.36654082, 0.2542696, 1.465271, 0.53526527, 0.28370926, 0.62043923, 0.16872439, 0.029439654, True, False, False, True, False),\n",
       " ('AFR', 'rs10188036', 31490055, 'A genome-wide association study of cocaine use disorder accounting for phenotypic heterogeneity and gene–environment interaction', 'Cocaine use disorder x change in residence interaction', '3E-8', 'rs13403153', '0.9', 'chr2', 201391971, 'T', 'C', 0.8448316, 0.36654082, 0.2542696, 1.465271, 0.53526527, 0.28370926, 0.62043923, 0.16872439, 0.029439654, True, False, False, True, False),\n",
       " ('EUR', 'rs2270894', 30595370, 'Leveraging Polygenic Functional Enrichment to Improve GWAS Power.', 'Height', '2E-80', 'rs2270894', 'tag', 'chr3', 9933702, 'C', 'G', 4.2506876, 3.6712945, 3.5454733, 5.0015254, 4.239152, 4.947809, 0.7508383, 0.56785756, 1.4023354, True, True, True, True, True),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Monocyte count', '9E-23', 'rs1003393', 'tag', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Mean platelet volume', '2E-15', 'rs1003393', 'tag', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Mean platelet volume', '2E-15', 'rs9941452', '1', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Monocyte count', '2E-20', 'rs1003393', 'tag', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Monocyte count', '2E-20', 'rs9941452', '1', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs1003393', 32888493, 'Trans-ethnic and Ancestry-Specific Blood-Cell Genetics in 746,667 Individuals from 5 Global Populations.', 'Monocyte count', '9E-23', 'rs9941452', '1', 'chr19', 13101400, 'C', 'T', 1.8323534, 1.5378494, 1.0421877, 1.3106359, 1.2315954, 0.8987868, -0.5217174, -0.30625406, -0.14340092, True, False, False, True, False),\n",
       " ('AFR', 'rs10034452', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '9E-7', 'rs59341143', '0.91', 'chr4', 13529200, 'T', 'G', 2.5369117, 1.9763436, 1.4932429, 1.6715355, 1.3225569, 0.8809182, -0.8653763, -0.65378684, -0.61232454, True, True, True, True, True),\n",
       " ('AFR', 'rs10034452', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '9E-7', 'rs9991833', '0.76', 'chr4', 13529200, 'T', 'G', 2.5369117, 1.9763436, 1.4932429, 1.6715355, 1.3225569, 0.8809182, -0.8653763, -0.65378684, -0.61232454, True, True, True, True, True),\n",
       " ('AFR', 'rs10034452', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '9E-7', 'rs10020715', '0.79', 'chr4', 13529200, 'T', 'G', 2.5369117, 1.9763436, 1.4932429, 1.6715355, 1.3225569, 0.8809182, -0.8653763, -0.65378684, -0.61232454, True, True, True, True, True),\n",
       " ('AFR', 'rs10034452', 36224396, 'A saturated map of common genetic variants associated with human height.', 'Height', '9E-7', 'rs10034452', 'tag', 'chr4', 13529200, 'T', 'G', 2.5369117, 1.9763436, 1.4932429, 1.6715355, 1.3225569, 0.8809182, -0.8653763, -0.65378684, -0.61232454, True, True, True, True, True),\n",
       " ('AFR', 'rs10195199', 36914875, 'Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.', 'Lung function (FEV1/FVC)', '2E-9', 'rs10195199', 'tag', 'chr2', 86223017, 'C', 'T', 0.53966457, 0.6494023, 0.14900297, 1.3617532, 0.5386045, 0.08207319, 0.8220887, -0.11079783, -0.06692975, True, False, False, True, False),\n",
       " ('AFR', 'rs10195199', 36914875, 'Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.', 'Lung function (FEV1/FVC)', '2E-9', 'rs10901', '0.81', 'chr2', 86223017, 'C', 'T', 0.53966457, 0.6494023, 0.14900297, 1.3617532, 0.5386045, 0.08207319, 0.8220887, -0.11079783, -0.06692975, True, False, False, True, False),\n",
       " ('AFR', 'rs10195199', 36914875, 'Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.', 'Lung function (FEV1/FVC)', '2E-9', 'rs12714178', '0.75', 'chr2', 86223017, 'C', 'T', 0.53966457, 0.6494023, 0.14900297, 1.3617532, 0.5386045, 0.08207319, 0.8220887, -0.11079783, -0.06692975, True, False, False, True, False),\n",
       " ('AFR', 'rs10195199', 36914875, 'Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.', 'Lung function (FEV1/FVC)', '2E-9', 'rs13007807', '0.81', 'chr2', 86223017, 'C', 'T', 0.53966457, 0.6494023, 0.14900297, 1.3617532, 0.5386045, 0.08207319, 0.8220887, -0.11079783, -0.06692975, True, False, False, True, False),\n",
       " ('AFR', 'rs10195199', 36914875, 'Multi-ancestry genome-wide association analyses improve resolution of genes and pathways influencing lung function and chronic obstructive pulmonary disease risk.', 'Lung function (FEV1/FVC)', '2E-9', 'rs13035593', '0.81', 'chr2', 86223017, 'C', 'T', 0.53966457, 0.6494023, 0.14900297, 1.3617532, 0.5386045, 0.08207319, 0.8220887, -0.11079783, -0.06692975, True, False, False, True, False),\n",
       " ('AFR', 'rs10830947', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Smoking initiation', '2E-17', 'rs3024236', '0.95', 'chr11', 92792449, 'G', 'A', 2.1934614, 0.6742889, -0.08360493, 1.3143308, 0.2776651, -0.22676457, -0.87913024, -0.39662376, -0.14315963, True, False, False, True, False)]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:37:15.043 EST [4169471] LOG:  checkpoint starting: time\n",
      "2025-02-10 21:38:04.698 EST [4169471] LOG:  checkpoint complete: wrote 260 buffers (1.6%); 0 WAL file(s) added, 0 removed, 8 recycled; write=48.990 s, sync=0.488 s, total=49.655 s; sync files=4, longest=0.488 s, average=0.122 s; distance=135542 kB, estimate=598212 kB; lsn=A0/473707C0, redo lsn=A0/45F74A00\n",
      "2025-02-10 21:40:14.187 EST [4181400] LOG:  automatic vacuum of table \"scratch.public.stacked_table\": index scans: 0\n",
      "\tpages: 0 removed, 3273481 remain, 2740423 scanned (83.72% of total)\n",
      "\ttuples: 161759 removed, 15057192 remain, 165849 are dead but not yet removable\n",
      "\tremovable cutoff: 826, which was 1 XIDs old when operation ended\n",
      "\tnew relfrozenxid: 824, which is 12 XIDs ahead of previous value\n",
      "\tfrozen: 2530 pages from table (0.08% of total) had 2530 tuples frozen\n",
      "\tindex scan not needed: 0 pages from table (0.00% of total) had 0 dead item identifiers removed\n",
      "\tavg read rate: 23.214 MB/s, avg write rate: 19.666 MB/s\n",
      "\tbuffer usage: 2753326 hits, 2728578 misses, 2311500 dirtied\n",
      "\tWAL usage: 4516661 records, 2295667 full page images, 364848219 bytes\n",
      "\tsystem usage: CPU: user: 5.58 s, system: 22.69 s, elapsed: 918.27 s\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "--checking a few entries to make sure they make sense. \n",
    "SELECT *\n",
    "FROM stacked_table\n",
    "WHERE k562_emvar=True\n",
    "LIMIT 20;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:22:16.895 EST [4179104] ERROR:  canceling autovacuum task\n",
      "2025-02-10 21:22:16.895 EST [4179104] CONTEXT:  while scanning block 1052397 of relation \"public.stacked_table\"\n",
      "\tautomatic vacuum of table \"scratch.public.stacked_table\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:22:43.426 EST [4169471] LOG:  checkpoint complete: wrote 9396 buffers (57.3%); 0 WAL file(s) added, 0 removed, 36 recycled; write=269.439 s, sync=0.656 s, total=271.963 s; sync files=22, longest=0.320 s, average=0.030 s; distance=597183 kB, estimate=662088 kB; lsn=9A/53A8B100, redo lsn=9A/41A80FF0\n",
      "2025-02-10 21:22:46.151 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:22:51.695 EST [4169471] LOG:  checkpoint complete: wrote 2678 buffers (16.3%); 0 WAL file(s) added, 3 removed, 30 recycled; write=3.806 s, sync=0.462 s, total=5.567 s; sync files=9, longest=0.248 s, average=0.052 s; distance=538661 kB, estimate=649745 kB; lsn=9A/824189B0, redo lsn=9A/6288A630\n",
      "2025-02-10 21:22:51.696 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:22:51.696 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:22:51.728 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:22:57.247 EST [4169471] LOG:  checkpoint complete: wrote 563 buffers (3.4%); 0 WAL file(s) added, 8 removed, 33 recycled; write=3.965 s, sync=0.178 s, total=5.552 s; sync files=3, longest=0.162 s, average=0.060 s; distance=669466 kB, estimate=669466 kB; lsn=9A/A9643668, redo lsn=9A/8B651068\n",
      "2025-02-10 21:22:57.363 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:22:57.363 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:22:57.363 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:03.195 EST [4169471] LOG:  checkpoint complete: wrote 30 buffers (0.2%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.378 s, sync=0.291 s, total=5.833 s; sync files=5, longest=0.184 s, average=0.059 s; distance=666180 kB, estimate=669137 kB; lsn=9A/D3638460, redo lsn=9A/B40E21D8\n",
      "2025-02-10 21:23:03.287 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:03.287 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:03.287 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:08.899 EST [4169471] LOG:  checkpoint complete: wrote 645 buffers (3.9%); 0 WAL file(s) added, 11 removed, 29 recycled; write=4.276 s, sync=0.318 s, total=5.613 s; sync files=7, longest=0.102 s, average=0.046 s; distance=667658 kB, estimate=668989 kB; lsn=9A/FD632290, redo lsn=9A/DCCE4B90\n",
      "2025-02-10 21:23:08.911 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:23:08.911 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:08.975 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:14.761 EST [4169471] LOG:  checkpoint complete: wrote 235 buffers (1.4%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.253 s, sync=0.321 s, total=5.851 s; sync files=6, longest=0.165 s, average=0.054 s; distance=651947 kB, estimate=667285 kB; lsn=9B/230AC798, redo lsn=9B/498F938\n",
      "2025-02-10 21:23:14.902 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:14.902 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:14.902 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:20.854 EST [4169471] LOG:  checkpoint complete: wrote 33 buffers (0.2%); 0 WAL file(s) added, 10 removed, 30 recycled; write=4.083 s, sync=0.572 s, total=5.952 s; sync files=5, longest=0.341 s, average=0.115 s; distance=659917 kB, estimate=666548 kB; lsn=9B/4D4334F0, redo lsn=9B/2CE02F48\n",
      "2025-02-10 21:23:21.046 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:23:21.046 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:21.046 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:27.510 EST [4169471] LOG:  checkpoint complete: wrote 321 buffers (2.0%); 0 WAL file(s) added, 12 removed, 32 recycled; write=4.806 s, sync=0.295 s, total=6.464 s; sync files=5, longest=0.171 s, average=0.059 s; distance=710346 kB, estimate=710346 kB; lsn=9B/776298F0, redo lsn=9B/583B5B00\n",
      "2025-02-10 21:23:27.666 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:27.666 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:27.689 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:33.088 EST [4169471] LOG:  checkpoint complete: wrote 60 buffers (0.4%); 0 WAL file(s) added, 11 removed, 31 recycled; write=3.962 s, sync=0.244 s, total=5.422 s; sync files=6, longest=0.150 s, average=0.041 s; distance=692642 kB, estimate=708576 kB; lsn=9B/A170F4E8, redo lsn=9B/8281E410\n",
      "2025-02-10 21:23:33.103 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:33.103 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:33.130 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:39.064 EST [4169471] LOG:  checkpoint complete: wrote 43 buffers (0.3%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.234 s, sync=0.263 s, total=5.962 s; sync files=4, longest=0.162 s, average=0.066 s; distance=654925 kB, estimate=703211 kB; lsn=9B/CA045460, redo lsn=9B/AA7B1A68\n",
      "2025-02-10 21:23:39.190 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:39.190 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:39.193 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:45.165 EST [4169471] LOG:  checkpoint complete: wrote 41 buffers (0.3%); 0 WAL file(s) added, 12 removed, 31 recycled; write=4.176 s, sync=0.405 s, total=5.975 s; sync files=4, longest=0.268 s, average=0.102 s; distance=711973 kB, estimate=711973 kB; lsn=9B/F564BB10, redo lsn=9B/D5EFAE78\n",
      "2025-02-10 21:23:45.288 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:45.288 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:45.288 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:51.456 EST [4169471] LOG:  checkpoint complete: wrote 29 buffers (0.2%); 0 WAL file(s) added, 12 removed, 31 recycled; write=4.790 s, sync=0.348 s, total=6.168 s; sync files=6, longest=0.164 s, average=0.058 s; distance=699669 kB, estimate=710742 kB; lsn=9C/2065CEC8, redo lsn=9C/A40330\n",
      "2025-02-10 21:23:51.477 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:51.477 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:51.477 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:23:57.034 EST [4169471] LOG:  checkpoint complete: wrote 135 buffers (0.8%); 0 WAL file(s) added, 7 removed, 32 recycled; write=4.380 s, sync=0.143 s, total=5.557 s; sync files=5, longest=0.086 s, average=0.029 s; distance=636846 kB, estimate=703353 kB; lsn=9C/460C7618, redo lsn=9C/2782BD58\n",
      "2025-02-10 21:23:57.112 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:23:57.112 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:23:57.117 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:02.876 EST [4169471] LOG:  checkpoint complete: wrote 383 buffers (2.3%); 0 WAL file(s) added, 7 removed, 32 recycled; write=4.264 s, sync=0.312 s, total=5.764 s; sync files=5, longest=0.149 s, average=0.063 s; distance=634471 kB, estimate=696464 kB; lsn=9C/6D2F98D0, redo lsn=9C/4E3C5B10\n",
      "2025-02-10 21:24:03.018 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:24:03.018 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:03.044 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:08.478 EST [4169471] LOG:  checkpoint complete: wrote 192 buffers (1.2%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.219 s, sync=0.287 s, total=5.461 s; sync files=4, longest=0.153 s, average=0.072 s; distance=662346 kB, estimate=693053 kB; lsn=9C/95F9A888, redo lsn=9C/76A984A0\n",
      "2025-02-10 21:24:08.591 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:24:08.591 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:08.619 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:14.334 EST [4169471] LOG:  checkpoint complete: wrote 274 buffers (1.7%); 0 WAL file(s) added, 8 removed, 30 recycled; write=4.505 s, sync=0.269 s, total=5.743 s; sync files=6, longest=0.118 s, average=0.045 s; distance=625232 kB, estimate=686270 kB; lsn=9C/BC6594F8, redo lsn=9C/9CD2C7B0\n",
      "2025-02-10 21:24:14.364 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:24:14.364 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:14.364 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:20.110 EST [4169471] LOG:  checkpoint complete: wrote 52 buffers (0.3%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.312 s, sync=0.456 s, total=5.747 s; sync files=5, longest=0.255 s, average=0.092 s; distance=619179 kB, estimate=679561 kB; lsn=9C/E18ED408, redo lsn=9C/C29D7488\n",
      "2025-02-10 21:24:20.239 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:24:20.239 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:20.250 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:25.985 EST [4169471] LOG:  checkpoint complete: wrote 40 buffers (0.2%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.261 s, sync=0.307 s, total=5.746 s; sync files=4, longest=0.163 s, average=0.077 s; distance=628324 kB, estimate=674438 kB; lsn=9D/84BD198, redo lsn=9C/E8F70570\n",
      "2025-02-10 21:24:26.104 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:24:26.104 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:26.104 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:24:32.187 EST [4169471] LOG:  checkpoint complete: wrote 424 buffers (2.6%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.484 s, sync=0.290 s, total=6.084 s; sync files=4, longest=0.162 s, average=0.073 s; distance=669504 kB, estimate=673944 kB; lsn=9D/315FF7C8, redo lsn=9D/11D406F8\n",
      "2025-02-10 21:24:32.298 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:24:32.298 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:24:32.298 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE stacked_table ADD COLUMN any_emvar BOOLEAN;\n",
    "UPDATE stacked_table\n",
    "SET any_emvar = (\n",
    "    k562_emvar OR hepg2_emvar OR sknsh_emvar\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:24:33.613 EST [4180837] ERROR:  canceling autovacuum task\n",
      "2025-02-10 21:24:33.613 EST [4180837] CONTEXT:  while scanning block 191337 of relation \"public.stacked_table\"\n",
      "\tautomatic vacuum of table \"scratch.public.stacked_table\"\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:25:15.833 EST [4169471] LOG:  checkpoint complete: wrote 85 buffers (0.5%); 0 WAL file(s) added, 14 removed, 27 recycled; write=41.033 s, sync=1.295 s, total=43.536 s; sync files=29, longest=0.767 s, average=0.045 s; distance=659436 kB, estimate=672493 kB; lsn=9D/5D63FF40, redo lsn=9D/3A13B730\n",
      "2025-02-10 21:25:15.851 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:21.955 EST [4169471] LOG:  checkpoint complete: wrote 47 buffers (0.3%); 0 WAL file(s) added, 12 removed, 31 recycled; write=4.564 s, sync=0.282 s, total=6.104 s; sync files=7, longest=0.150 s, average=0.041 s; distance=715596 kB, estimate=715596 kB; lsn=9D/856193F0, redo lsn=9D/65C0E8F0\n",
      "2025-02-10 21:25:22.065 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:25:22.065 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:22.087 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:27.938 EST [4169471] LOG:  checkpoint complete: wrote 72 buffers (0.4%); 0 WAL file(s) added, 9 removed, 32 recycled; write=4.466 s, sync=0.271 s, total=5.873 s; sync files=4, longest=0.159 s, average=0.068 s; distance=664780 kB, estimate=710514 kB; lsn=9D/AD62C200, redo lsn=9D/8E5419E8\n",
      "2025-02-10 21:25:28.011 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:25:28.011 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:28.033 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:33.733 EST [4169471] LOG:  checkpoint complete: wrote 654 buffers (4.0%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.136 s, sync=0.311 s, total=5.723 s; sync files=4, longest=0.200 s, average=0.078 s; distance=627534 kB, estimate=702216 kB; lsn=9D/D4424978, redo lsn=9D/B4A155E0\n",
      "2025-02-10 21:25:33.870 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:25:33.870 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:33.888 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:40.237 EST [4169471] LOG:  checkpoint complete: wrote 47 buffers (0.3%); 0 WAL file(s) added, 10 removed, 31 recycled; write=4.573 s, sync=0.553 s, total=6.367 s; sync files=7, longest=0.226 s, average=0.079 s; distance=662771 kB, estimate=698272 kB; lsn=9D/FD654388, redo lsn=9D/DD1522F8\n",
      "2025-02-10 21:25:40.296 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:25:40.296 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:40.296 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:46.435 EST [4169471] LOG:  checkpoint complete: wrote 261 buffers (1.6%); 0 WAL file(s) added, 10 removed, 32 recycled; write=4.794 s, sync=0.348 s, total=6.140 s; sync files=4, longest=0.152 s, average=0.087 s; distance=688404 kB, estimate=697285 kB; lsn=9E/2650E2D0, redo lsn=9E/7197678\n",
      "2025-02-10 21:25:46.640 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:25:46.640 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:46.640 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:52.671 EST [4169471] LOG:  checkpoint complete: wrote 38 buffers (0.2%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.576 s, sync=0.248 s, total=6.031 s; sync files=4, longest=0.117 s, average=0.062 s; distance=631983 kB, estimate=690755 kB; lsn=9E/4D420C50, redo lsn=9E/2DAC3410\n",
      "2025-02-10 21:25:52.844 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:25:52.844 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:52.844 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:25:58.952 EST [4169471] LOG:  checkpoint complete: wrote 643 buffers (3.9%); 0 WAL file(s) added, 10 removed, 32 recycled; write=4.664 s, sync=0.267 s, total=6.108 s; sync files=5, longest=0.137 s, average=0.054 s; distance=678048 kB, estimate=689484 kB; lsn=9E/76650300, redo lsn=9E/570EB648\n",
      "2025-02-10 21:25:59.111 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:25:59.111 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:25:59.111 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:05.434 EST [4169471] LOG:  checkpoint complete: wrote 44 buffers (0.3%); 0 WAL file(s) added, 7 removed, 32 recycled; write=4.695 s, sync=0.364 s, total=6.324 s; sync files=6, longest=0.199 s, average=0.061 s; distance=639477 kB, estimate=684483 kB; lsn=9E/9D8C86A8, redo lsn=9E/7E168DF0\n",
      "2025-02-10 21:26:05.651 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:26:05.651 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:05.651 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:11.722 EST [4169471] LOG:  checkpoint complete: wrote 46 buffers (0.3%); 0 WAL file(s) added, 11 removed, 31 recycled; write=4.819 s, sync=0.335 s, total=6.072 s; sync files=4, longest=0.181 s, average=0.084 s; distance=689501 kB, estimate=689501 kB; lsn=9E/C7D85DC8, redo lsn=9E/A82C0420\n",
      "2025-02-10 21:26:11.759 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:26:11.759 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:11.772 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:17.558 EST [4169471] LOG:  checkpoint complete: wrote 255 buffers (1.6%); 0 WAL file(s) added, 7 removed, 31 recycled; write=4.460 s, sync=0.148 s, total=5.799 s; sync files=4, longest=0.093 s, average=0.037 s; distance=625872 kB, estimate=683138 kB; lsn=9E/ED64A3E0, redo lsn=9E/CE5F4778\n",
      "2025-02-10 21:26:17.733 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:26:17.733 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:17.733 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:24.017 EST [4169471] LOG:  checkpoint complete: wrote 338 buffers (2.1%); 0 WAL file(s) added, 8 removed, 32 recycled; write=4.674 s, sync=0.464 s, total=6.285 s; sync files=6, longest=0.191 s, average=0.078 s; distance=652622 kB, estimate=680087 kB; lsn=9F/1549B968, redo lsn=9E/F6348250\n",
      "2025-02-10 21:26:24.066 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:26:24.066 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:24.095 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:30.562 EST [4169471] LOG:  checkpoint complete: wrote 516 buffers (3.1%); 0 WAL file(s) added, 8 removed, 31 recycled; write=4.833 s, sync=0.407 s, total=6.496 s; sync files=6, longest=0.213 s, average=0.068 s; distance=645393 kB, estimate=676617 kB; lsn=9F/3D62CC50, redo lsn=9F/1D98C7B0\n",
      "2025-02-10 21:26:30.720 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:26:30.720 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:30.761 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:36.996 EST [4169471] LOG:  checkpoint complete: wrote 1151 buffers (7.0%); 0 WAL file(s) added, 9 removed, 32 recycled; write=4.890 s, sync=0.162 s, total=6.276 s; sync files=4, longest=0.081 s, average=0.041 s; distance=670907 kB, estimate=676046 kB; lsn=9F/64E11250, redo lsn=9F/468BB6D8\n",
      "2025-02-10 21:26:37.127 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:26:37.127 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:37.131 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:42.835 EST [4169471] LOG:  checkpoint complete: wrote 35 buffers (0.2%); 0 WAL file(s) added, 9 removed, 31 recycled; write=4.538 s, sync=0.180 s, total=5.709 s; sync files=4, longest=0.090 s, average=0.045 s; distance=657019 kB, estimate=674143 kB; lsn=9F/8DF3B458, redo lsn=9F/6EA5A4A0\n",
      "2025-02-10 21:26:42.991 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:26:42.991 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:42.991 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:48.409 EST [4169471] LOG:  checkpoint complete: wrote 526 buffers (3.2%); 0 WAL file(s) added, 7 removed, 32 recycled; write=4.213 s, sync=0.131 s, total=5.418 s; sync files=7, longest=0.062 s, average=0.019 s; distance=643260 kB, estimate=671055 kB; lsn=9F/B4439208, redo lsn=9F/95E894E8\n",
      "2025-02-10 21:26:48.448 EST [4169471] LOG:  checkpoints are occurring too frequently (6 seconds apart)\n",
      "2025-02-10 21:26:48.448 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:48.448 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:26:53.849 EST [4169471] LOG:  checkpoint complete: wrote 106 buffers (0.6%); 0 WAL file(s) added, 6 removed, 32 recycled; write=4.260 s, sync=0.110 s, total=5.402 s; sync files=4, longest=0.040 s, average=0.028 s; distance=608448 kB, estimate=664794 kB; lsn=9F/D9C95850, redo lsn=9F/BB0B9880\n",
      "2025-02-10 21:26:53.972 EST [4169471] LOG:  checkpoints are occurring too frequently (5 seconds apart)\n",
      "2025-02-10 21:26:53.972 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:26:53.985 EST [4169471] LOG:  checkpoint starting: wal\n",
      "2025-02-10 21:27:00.122 EST [4169471] LOG:  checkpoint complete: wrote 37 buffers (0.2%); 0 WAL file(s) added, 8 removed, 31 recycled; write=4.453 s, sync=0.421 s, total=6.151 s; sync files=4, longest=0.222 s, average=0.105 s; distance=643319 kB, estimate=662647 kB; lsn=A0/2618EE8, redo lsn=9F/E24F74E0\n",
      "2025-02-10 21:27:00.328 EST [4169471] LOG:  checkpoints are occurring too frequently (7 seconds apart)\n",
      "2025-02-10 21:27:00.328 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:27:00.328 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17769769 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:27:06.016 EST [4169471] LOG:  checkpoint complete: wrote 375 buffers (2.3%); 0 WAL file(s) added, 11 removed, 32 recycled; write=4.681 s, sync=0.201 s, total=5.688 s; sync files=4, longest=0.073 s, average=0.051 s; distance=711527 kB, estimate=711527 kB; lsn=A0/2CABFBB0, redo lsn=A0/DBD1470\n",
      "2025-02-10 21:27:15.264 EST [4169471] LOG:  checkpoints are occurring too frequently (15 seconds apart)\n",
      "2025-02-10 21:27:15.264 EST [4169471] HINT:  Consider increasing the configuration parameter \"max_wal_size\".\n",
      "2025-02-10 21:27:15.264 EST [4169471] LOG:  checkpoint starting: wal\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "ALTER TABLE stacked_table ADD COLUMN all_emvar BOOLEAN;\n",
    "UPDATE stacked_table\n",
    "SET all_emvar = (\n",
    "    k562_emvar AND hepg2_emvar AND sknsh_emvar\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "(psycopg2.ProgrammingError) can't execute an empty query\n",
      "[SQL: --checking a few entries to make sure they make sense. \n",
      "--SELECT *\n",
      "--FROM stacked_table\n",
      "--WHERE any_emvar=True\n",
      "--LIMIT 20;]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "--checking a few entries to make sure they make sense. \n",
    "--SELECT *\n",
    "--FROM stacked_table\n",
    "--WHERE any_emvar=True\n",
    "--LIMIT 20;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>total_rows</th>\n",
       "            <th>true_any_emvar_count</th>\n",
       "            <th>true_all_emvar_count</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>1262917</td>\n",
       "            <td>23572</td>\n",
       "            <td>4742</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>ASN</td>\n",
       "            <td>4804689</td>\n",
       "            <td>84765</td>\n",
       "            <td>22285</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>EUR</td>\n",
       "            <td>11702163</td>\n",
       "            <td>237597</td>\n",
       "            <td>53041</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 1262917, 23572, 4742),\n",
       " ('ASN', 4804689, 84765, 22285),\n",
       " ('EUR', 11702163, 237597, 53041)]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:31:45.833 EST [4169471] LOG:  checkpoint complete: wrote 3247 buffers (19.8%); 0 WAL file(s) added, 0 removed, 33 recycled; write=269.696 s, sync=0.584 s, total=270.570 s; sync files=33, longest=0.553 s, average=0.018 s; distance=528581 kB, estimate=693233 kB; lsn=A0/3CCB52E8, redo lsn=A0/2E002B58\n",
      "2025-02-10 21:32:15.787 EST [4169471] LOG:  checkpoint starting: time\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT\n",
    "    source_table,\n",
    "    COUNT(*) AS total_rows,\n",
    "    SUM(any_emvar::INT) AS true_any_emvar_count,\n",
    "    SUM(all_emvar::INT) AS true_all_emvar_count\n",
    "FROM\n",
    "    stacked_table\n",
    "GROUP BY\n",
    "    source_table;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:42:16.009 EST [4169471] LOG:  checkpoint starting: time\n",
      "2025-02-10 21:42:36.203 EST [4169471] LOG:  checkpoint complete: wrote 191 buffers (1.2%); 0 WAL file(s) added, 0 removed, 5 recycled; write=19.819 s, sync=0.047 s, total=20.195 s; sync files=37, longest=0.015 s, average=0.002 s; distance=77788 kB, estimate=546170 kB; lsn=A0/4AB6BDF8, redo lsn=A0/4AB6BCD8\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3647142 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE TABLE buddies AS\n",
    "SELECT\n",
    "    source_table,\n",
    "    ld_buddy,\n",
    "    SUM(any_emvar::INT) AS true_any_emvar_count\n",
    "FROM\n",
    "    stacked_table\n",
    "GROUP BY\n",
    "    source_table,\n",
    "    ld_buddy;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>true_any_emvar_count</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10000945</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10001439</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10004033</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10004388</td>\n",
       "            <td>1</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>rs10006545</td>\n",
       "            <td>3</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 'rs10000945', 1),\n",
       " ('AFR', 'rs10001439', 1),\n",
       " ('AFR', 'rs10004033', 1),\n",
       " ('AFR', 'rs10004388', 1),\n",
       " ('AFR', 'rs10006545', 3)]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:47:16.265 EST [4169471] LOG:  checkpoint starting: time\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "--examine a few rows to make sure it's OK\n",
    "SELECT *\n",
    "FROM buddies\n",
    "WHERE true_any_emvar_count <> 0\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "3 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>source_table</th>\n",
       "            <th>total_ld_buddies</th>\n",
       "            <th>ld_buddies_with_nonzero_emvar_in_any_cell_count</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>AFR</td>\n",
       "            <td>519285</td>\n",
       "            <td>12509</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>ASN</td>\n",
       "            <td>1155478</td>\n",
       "            <td>38690</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>EUR</td>\n",
       "            <td>1972379</td>\n",
       "            <td>83596</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AFR', 519285, 12509), ('ASN', 1155478, 38690), ('EUR', 1972379, 83596)]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "--Per pop,\n",
    "--Select the number of ld_buddies with any emvars (any_emvar_count is zero). \n",
    "SELECT\n",
    "    source_table,\n",
    "    COUNT(true_any_emvar_count) AS total_ld_buddies,\n",
    "    SUM(CASE WHEN true_any_emvar_count > 0 THEN 1 ELSE 0 END) AS ld_buddies_with_nonzero_emvar_in_any_cell_count\n",
    "FROM\n",
    "    buddies\n",
    "GROUP BY\n",
    "    source_table;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "waiting for server to shut down...."
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:54:29.254 EST [4169438] LOG:  received fast shutdown request\n",
      "2025-02-10 21:54:29.270 EST [4169438] LOG:  aborting any active transactions\n",
      "2025-02-10 21:54:29.270 EST [4176288] FATAL:  terminating connection due to administrator command\n",
      "2025-02-10 21:54:29.287 EST [4169438] LOG:  background worker \"logical replication launcher\" (PID 4169511) exited with exit code 1\n",
      "2025-02-10 21:54:29.365 EST [4169471] LOG:  shutting down\n",
      "2025-02-10 21:54:29.383 EST [4169471] LOG:  checkpoint starting: shutdown immediate\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " done\n",
      "server stopped\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-10 21:54:29.500 EST [4169471] LOG:  checkpoint complete: wrote 0 buffers (0.0%); 0 WAL file(s) added, 0 removed, 0 recycled; write=0.005 s, sync=0.001 s, total=0.135 s; sync files=0, longest=0.000 s, average=0.000 s; distance=0 kB, estimate=468259 kB; lsn=A0/5C409A80, redo lsn=A0/5C409A80\n",
      "2025-02-10 21:54:29.591 EST [4169438] LOG:  database system is shut down\n"
     ]
    }
   ],
   "source": [
    "!pg_ctl -D ~/palmer_scratch/db stop"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mcn_varef",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
