{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c4699403-628c-4809-b8b1-d4d1a9e2bdf6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-05T15:39:42.727011Z",
     "iopub.status.busy": "2025-02-05T15:39:42.726527Z",
     "iopub.status.idle": "2025-02-05T15:39:52.685867Z",
     "shell.execute_reply": "2025-02-05T15:39:52.685427Z",
     "shell.execute_reply.started": "2025-02-05T15:39:42.726990Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-05 17:32:32.300 EDT [2396617] LOG:  starting PostgreSQL 17.2 on x86_64-conda-linux-gnu, compiled by x86_64-conda-linux-gnu-cc (Anaconda gcc) 11.2.0, 64-bit\n",
      "2025-04-05 17:32:32.300 EDT [2396617] LOG:  listening on IPv6 address \"::1\", port 5433\n",
      "2025-04-05 17:32:32.300 EDT [2396617] LOG:  listening on IPv4 address \"127.0.0.1\", port 5433\n",
      "2025-04-05 17:32:32.303 EDT [2396617] LOG:  listening on Unix socket \"/tmp/.s.PGSQL.5433\"\n",
      "2025-04-05 17:32:32.371 EDT [2396620] LOG:  database system was shut down at 2025-04-05 16:38:47 EDT\n",
      "2025-04-05 17:32:32.422 EDT [2396617] LOG:  database system is ready to accept connections\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import requests\n",
    "#start the database\n",
    "process = subprocess.Popen([\"postgres\", \"-D\", \"/home/mcn26/palmer_scratch/db\", \"-p\", \"5433\"])\n",
    "# Load the SQL extension\n",
    "%load_ext sql\n",
    "#dealing with a weird bug\n",
    "%config SqlMagic.style = '_DEPRECATED_DEFAULT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7eb7ccd5-804b-4d4f-b8ba-c1a2dcbd8728",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-05T15:39:52.686803Z",
     "iopub.status.busy": "2025-02-05T15:39:52.686584Z",
     "iopub.status.idle": "2025-02-05T15:41:52.731830Z",
     "shell.execute_reply": "2025-02-05T15:41:52.731375Z",
     "shell.execute_reply.started": "2025-02-05T15:39:52.686789Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[+] Waiting 120s for database server to start\n",
      "[+] 2025-04-05 17:32:49.711748\n"
     ]
    }
   ],
   "source": [
    "print(\"[+] Waiting 120s for database server to start\", flush=True)\n",
    "import time\n",
    "import datetime\n",
    "print(f\"[+] {datetime.datetime.now()}\",flush=True)\n",
    "time.sleep(120)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "66afb22e-c33d-4224-8439-0f06472c2177",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-05T15:41:52.732791Z",
     "iopub.status.busy": "2025-02-05T15:41:52.732613Z",
     "iopub.status.idle": "2025-02-05T15:41:53.904054Z",
     "shell.execute_reply": "2025-02-05T15:41:53.903425Z",
     "shell.execute_reply.started": "2025-02-05T15:41:52.732777Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-05 17:37:32.373 EDT [2396618] LOG:  checkpoint starting: time\n",
      "2025-04-05 17:39:17.623 EDT [2396618] LOG:  checkpoint complete: wrote 531 buffers (3.2%); 0 WAL file(s) added, 4 removed, 0 recycled; write=104.499 s, sync=0.679 s, total=105.250 s; sync files=25, longest=0.305 s, average=0.028 s; distance=59107 kB, estimate=59107 kB; lsn=183/3DAF8940, redo lsn=183/3C197A78\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "postgresql://mr_root:password@localhost:5433/scratch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "493daffb",
   "metadata": {},
   "source": [
    "Let's check to make sure that the deduplication was successful..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b485e170-b900-462b-8f14-3b8334999d32",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-30T20:58:59.879803Z",
     "iopub.status.busy": "2025-01-30T20:58:59.879311Z",
     "iopub.status.idle": "2025-01-30T20:58:59.989886Z",
     "shell.execute_reply": "2025-01-30T20:58:59.989510Z",
     "shell.execute_reply.started": "2025-01-30T20:58:59.879777Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>chromosome_id</th>\n",
       "            <th>pos</th>\n",
       "            <th>id</th>\n",
       "            <th>ref</th>\n",
       "            <th>alt</th>\n",
       "            <th>qual</th>\n",
       "            <th>filter</th>\n",
       "            <th>info</th>\n",
       "            <th>k562__ref</th>\n",
       "            <th>hepg2__ref</th>\n",
       "            <th>sknsh__ref</th>\n",
       "            <th>k562__alt</th>\n",
       "            <th>hepg2__alt</th>\n",
       "            <th>sknsh__alt</th>\n",
       "            <th>k562__skew</th>\n",
       "            <th>hepg2__skew</th>\n",
       "            <th>sknsh__skew</th>\n",
       "            <th>ac</th>\n",
       "            <th>an</th>\n",
       "            <th>af</th>\n",
       "            <th>cadd_phred</th>\n",
       "            <th>is_in_dels</th>\n",
       "            <th>is_in_ca</th>\n",
       "            <th>is_in_pels</th>\n",
       "            <th>is_in_ca_h3k4me3</th>\n",
       "            <th>is_in_ca_ctcf</th>\n",
       "            <th>is_in_pls</th>\n",
       "            <th>is_in_tf</th>\n",
       "            <th>is_in_ca_tf</th>\n",
       "            <th>p_anno</th>\n",
       "            <th>mean_ref</th>\n",
       "            <th>mean_skew</th>\n",
       "            <th>maf</th>\n",
       "            <th>category</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>chr7</td>\n",
       "            <td>92754574</td>\n",
       "            <td>rs10</td>\n",
       "            <td>A</td>\n",
       "            <td>C</td>\n",
       "            <td>.</td>\n",
       "            <td>PASS</td>\n",
       "            <td>K562__ref=0.23715062;HepG2__ref=0.3610611;SKNSH__ref=0.18303184;K562__alt=0.2820961;HepG2__alt=0.40103868;SKNSH__alt=0.1955727;K562__skew=0.044945486;HepG2__skew=0.039977532;SKNSH__skew=0.012540861;AN_oth=2094;AN_ami=912;AN_sas=4830;AN_fin=10584;AN_eas=5198;AN_amr=15284;AN_afr=41452;AN_mid=316;AN_asj=3472;AN_nfe=68046;cadd_raw_score=-0.332981;cadd_phred=0.266;splice_ai_max_ds=0;splice_ai_consequence=no_consequence;vep=C|intron_variant|MODIFIER|CDK6|ENSG00000105810|Transcript|ENST00000265734|protein_coding||3/7|ENST00000265734.8:c.369+20122T&gt;G|||||||1||-1|SNV||HGNC|HGNC:1777||1|P1|CCDS5628.1|ENSP00000265734||||1||||||||||||,C|intron_variant|MODIFIER|CDK6|ENSG00000105810|Transcript|ENST00000424848|protein_coding||3/7|ENST00000424848.3:c.369+20122T&gt;G|||||||1||-1|SNV||HGNC|HGNC:1777|YES|1|P1|CCDS5628.1|ENSP00000397087||||1||||||||||||,C|intron_variant|MODIFIER|CDK6|1021|Transcript|NM_001145306.2|protein_coding||3/7|NM_001145306.2:c.369+20122T&gt;G|||||||1||-1|SNV||EntrezGene|HGNC:1777|YES||||NP_001138778.1||||||||||||||||,C|intron_variant|MODIFIER|CDK6|1021|Transcript|NM_001259.8|protein_coding||3/7|NM_001259.8:c.369+20122T&gt;G|||||||1||-1|SNV||EntrezGene|HGNC:1777|||||NP_001250.1||||||||||||||||,C|intron_variant|MODIFIER|CDK6|1021|Transcript|XM_006715835.2|protein_coding||3/7|XM_006715835.2:c.369+20122T&gt;G|||||||1||-1|SNV||EntrezGene|HGNC:1777|||||XP_006715898.1||||||||||||||||;AC=146019;AN=152188;AF=0.959465;AC_oth=2038;AF_oth=0.973257;AC_ami=904;AF_ami=0.991228;AC_sas=4779;AF_sas=0.989441;AC_fin=9827;AF_fin=0.928477;AC_eas=5197;AF_eas=0.999808;AC_amr=14683;AF_amr=0.960678;AC_afr=41052;AF_afr=0.99035;AC_mid=316;AF_mid=1;AC_asj=3414;AF_asj=0.983295;AC_nfe=63809;AF_nfe=0.937733;AC_popmax=5197;AN_popmax=5198;AF_popmax=0.999808</td>\n",
       "            <td>0.23715062</td>\n",
       "            <td>0.3610611</td>\n",
       "            <td>0.18303184</td>\n",
       "            <td>0.2820961</td>\n",
       "            <td>0.40103868</td>\n",
       "            <td>0.1955727</td>\n",
       "            <td>0.044945486</td>\n",
       "            <td>0.039977532</td>\n",
       "            <td>0.012540861</td>\n",
       "            <td>146019</td>\n",
       "            <td>152188</td>\n",
       "            <td>0.959465</td>\n",
       "            <td>0.266</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>-0.715</td>\n",
       "            <td>0.26041452089945477</td>\n",
       "            <td>0.032487958669662476</td>\n",
       "            <td>0.040534973</td>\n",
       "            <td>COMMON</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr12</td>\n",
       "            <td>126406434</td>\n",
       "            <td>rs1000000</td>\n",
       "            <td>G</td>\n",
       "            <td>A</td>\n",
       "            <td>.</td>\n",
       "            <td>PASS</td>\n",
       "            <td>K562__ref=0.14421938;HepG2__ref=0.21698855;SKNSH__ref=0.22241707;K562__alt=0.14721091;HepG2__alt=0.2585578;SKNSH__alt=0.22416125;K562__skew=0.0029915501;HepG2__skew=0.041569263;SKNSH__skew=0.0017441983;AN_oth=2082;AN_ami=908;AN_sas=4828;AN_fin=10554;AN_eas=5188;AN_amr=15268;AN_afr=41358;AN_mid=316;AN_asj=3468;AN_nfe=68008;cadd_raw_score=-0.028762;cadd_phred=1.165;vep=A|upstream_gene_variant|MODIFIER|LINC02825|ENSG00000256494|Transcript|ENST00000536422|lncRNA||||||||||1|906|-1|SNV||HGNC|HGNC:27477|YES|5|||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|AC007368.1|ENSG00000255595|Transcript|ENST00000642569|lncRNA||8/9|ENST00000642569.1:n.1405+175G&gt;A|||||||1||1|SNV||Clone_based_ensembl_gene||||||||||||||||||||||,A|non_coding_transcript_exon_variant|MODIFIER|AC007368.1|ENSG00000255595|Transcript|ENST00000662644|lncRNA|5/5||ENST00000662644.1:n.1019G&gt;A||1019|||||1||1|SNV||Clone_based_ensembl_gene||||||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|LOC283435|283435|Transcript|NR_147498.1|lncRNA||2/3|NR_147498.1:n.260-906C&gt;T|||||||1||-1|SNV||EntrezGene||YES||||||||||||||||||||;AC=28089;AN=151978;AF=0.184823;AC_oth=407;AF_oth=0.195485;AC_ami=183;AF_ami=0.201542;AC_sas=1144;AF_sas=0.236951;AC_fin=2528;AF_fin=0.23953;AC_eas=671;AF_eas=0.129337;AC_amr=2292;AF_amr=0.150118;AC_afr=5251;AF_afr=0.126965;AC_mid=64;AF_mid=0.202532;AC_asj=663;AF_asj=0.191176;AC_nfe=14886;AF_nfe=0.218886;AC_popmax=1144;AN_popmax=4828;AF_popmax=0.236951</td>\n",
       "            <td>0.14421938</td>\n",
       "            <td>0.21698855</td>\n",
       "            <td>0.22241707</td>\n",
       "            <td>0.14721091</td>\n",
       "            <td>0.2585578</td>\n",
       "            <td>0.22416125</td>\n",
       "            <td>0.0029915501</td>\n",
       "            <td>0.041569263</td>\n",
       "            <td>0.0017441983</td>\n",
       "            <td>28089</td>\n",
       "            <td>151978</td>\n",
       "            <td>0.184823</td>\n",
       "            <td>1.165</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>0.656</td>\n",
       "            <td>0.19454167286554971</td>\n",
       "            <td>0.015435003985961279</td>\n",
       "            <td>0.184823</td>\n",
       "            <td>COMMON</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr4</td>\n",
       "            <td>40087276</td>\n",
       "            <td>rs10000000</td>\n",
       "            <td>A</td>\n",
       "            <td>T</td>\n",
       "            <td>.</td>\n",
       "            <td>PASS</td>\n",
       "            <td>K562__ref=0.36345437;HepG2__ref=0.26544645;SKNSH__ref=0.49515116;K562__alt=0.353244;HepG2__alt=0.25292698;SKNSH__alt=0.4838382;K562__skew=-0.010210349;HepG2__skew=-0.01251948;SKNSH__skew=-0.0113130035;AN_oth=2058;AN_ami=902;AN_sas=4744;AN_fin=10380;AN_eas=5106;AN_amr=15066;AN_afr=40904;AN_mid=308;AN_asj=3428;AN_nfe=67160;cadd_raw_score=-0.252126;cadd_phred=0.394;splice_ai_max_ds=0;splice_ai_consequence=no_consequence;vep=T|intron_variant|MODIFIER|N4BP2|ENSG00000078177|Transcript|ENST00000261435|protein_coding||2/17|ENST00000261435.11:c.-114-9951A&gt;T|||||||1||1|SNV||HGNC|HGNC:29851|YES|5|P1|CCDS3457.1|ENSP00000261435||||1||||||||||||,T|downstream_gene_variant|MODIFIER|AC098591.1|ENSG00000201863|Transcript|ENST00000364993|snoRNA||||||||||1|4170|1|SNV||Clone_based_ensembl_gene||YES||||||||||||||||||||,T|intron_variant&amp;NMD_transcript_variant|MODIFIER|N4BP2|ENSG00000078177|Transcript|ENST00000511480|nonsense_mediated_decay||2/18|ENST00000511480.5:c.-114-9951A&gt;T|||||||1||1|SNV||HGNC|HGNC:29851||1|||ENSP00000422436||||1||||||||||||,T|intron_variant|MODIFIER|N4BP2|ENSG00000078177|Transcript|ENST00000515550|protein_coding||2/2|ENST00000515550.1:c.-12+13725A&gt;T|||||||1||1|SNV||HGNC|HGNC:29851||3|||ENSP00000422057||||1||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|NM_001318359.2|protein_coding||2/18|NM_001318359.2:c.-445-9951A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||NP_001305288.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|NM_018177.6|protein_coding||2/17|NM_018177.6:c.-114-9951A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|YES||||NP_060647.2||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_006714023.3|protein_coding||2/16|XM_006714023.3:c.-12+13725A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_006714086.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_011513716.2|protein_coding||1/16|XM_011513716.2:c.626-14799A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_011512018.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_011513717.2|protein_coding||1/16|XM_011513717.2:c.626-14799A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_011512019.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_011513718.2|protein_coding||1/16|XM_011513718.2:c.-114-9951A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_011512020.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_011513719.2|protein_coding||2/17|XM_011513719.2:c.-102-12760A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_011512021.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_017008397.1|protein_coding||1/15|XM_017008397.1:c.626-14799A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_016863886.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_017008398.1|protein_coding||1/16|XM_017008398.1:c.-115+6797A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_016863887.1||||||||||||||||,T|intron_variant|MODIFIER|N4BP2|55728|Transcript|XM_017008399.1|protein_coding||1/17|XM_017008399.1:c.-445-9951A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||XP_016863888.1||||||||||||||||,T|intron_variant&amp;non_coding_transcript_variant|MODIFIER|N4BP2|55728|Transcript|XR_001741283.1|misc_RNA||1/17|XR_001741283.1:n.807-14799A&gt;T|||||||1||1|SNV||EntrezGene|HGNC:29851|||||||||||||||||||||,T|regulatory_region_variant|MODIFIER|||RegulatoryFeature|ENSR00001081330|promoter_flanking_region||||||||||1|||SNV||||||||||||||||||||||||;AC=18324;AN=150056;AF=0.122114;AC_oth=243;AF_oth=0.118076;AC_ami=63;AF_ami=0.0698448;AC_sas=780;AF_sas=0.164418;AC_fin=723;AF_fin=0.0696532;AC_eas=2132;AF_eas=0.417548;AC_amr=1178;AF_amr=0.0781893;AC_afr=8745;AF_afr=0.213793;AC_mid=35;AF_mid=0.113636;AC_asj=265;AF_asj=0.0773046;AC_nfe=4160;AF_nfe=0.0619416;AC_popmax=2132;AN_popmax=5106;AF_popmax=0.417548</td>\n",
       "            <td>0.36345437</td>\n",
       "            <td>0.26544645</td>\n",
       "            <td>0.49515116</td>\n",
       "            <td>0.353244</td>\n",
       "            <td>0.25292698</td>\n",
       "            <td>0.4838382</td>\n",
       "            <td>-0.010210349</td>\n",
       "            <td>-0.01251948</td>\n",
       "            <td>-0.0113130035</td>\n",
       "            <td>18324</td>\n",
       "            <td>150056</td>\n",
       "            <td>0.122114</td>\n",
       "            <td>0.394</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>-3.541</td>\n",
       "            <td>0.37468401590983075</td>\n",
       "            <td>-0.011347610503435135</td>\n",
       "            <td>0.122114</td>\n",
       "            <td>COMMON</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr4</td>\n",
       "            <td>56695481</td>\n",
       "            <td>rs10000003</td>\n",
       "            <td>A</td>\n",
       "            <td>G</td>\n",
       "            <td>.</td>\n",
       "            <td>PASS</td>\n",
       "            <td>K562__ref=0.2162989;HepG2__ref=0.1461182;SKNSH__ref=0.1776367;K562__alt=0.22383805;HepG2__alt=0.16024604;SKNSH__alt=0.18377645;K562__skew=0.0075391266;HepG2__skew=0.014127848;SKNSH__skew=0.006139751;AN_oth=2090;AN_ami=912;AN_sas=4822;AN_fin=10566;AN_eas=5190;AN_amr=15256;AN_afr=41406;AN_mid=316;AN_asj=3470;AN_nfe=68010;cadd_raw_score=-0.982282;cadd_phred=0.008;vep=G|downstream_gene_variant|MODIFIER|LOC105377667|105377667|Transcript|XR_001741701.1|lncRNA||||||||||1|1736|-1|SNV||EntrezGene||||||||||||||||||||||,G|downstream_gene_variant|MODIFIER|LOC105377667|105377667|Transcript|XR_001741702.2|lncRNA||||||||||1|1736|-1|SNV||EntrezGene||||||||||||||||||||||,G|downstream_gene_variant|MODIFIER|LOC105377667|105377667|Transcript|XR_001741703.1|lncRNA||||||||||1|1736|-1|SNV||EntrezGene||YES||||||||||||||||||||,G|intron_variant&amp;non_coding_transcript_variant|MODIFIER|LOC105377666|105377666|Transcript|XR_941078.2|lncRNA||2/2|XR_941078.2:n.362-694A&gt;G|||||||1||1|SNV||EntrezGene||YES||||||||||||||||||||,G|downstream_gene_variant|MODIFIER|LOC105377667|105377667|Transcript|XR_941080.2|lncRNA||||||||||1|1736|-1|SNV||EntrezGene||||||||||||||||||||||,G|downstream_gene_variant|MODIFIER|LOC105377667|105377667|Transcript|XR_941081.3|lncRNA||||||||||1|1736|-1|SNV||EntrezGene||||||||||||||||||||||;AC=112032;AN=152038;AF=0.736868;AC_oth=1477;AF_oth=0.706699;AC_ami=626;AF_ami=0.686404;AC_sas=3600;AF_sas=0.746578;AC_fin=7095;AF_fin=0.671493;AC_eas=4627;AF_eas=0.891522;AC_amr=10628;AF_amr=0.696644;AC_afr=33557;AF_afr=0.810438;AC_mid=228;AF_mid=0.721519;AC_asj=2475;AF_asj=0.713256;AC_nfe=47719;AF_nfe=0.701647;AC_popmax=4627;AN_popmax=5190;AF_popmax=0.891522</td>\n",
       "            <td>0.2162989</td>\n",
       "            <td>0.1461182</td>\n",
       "            <td>0.1776367</td>\n",
       "            <td>0.22383805</td>\n",
       "            <td>0.16024604</td>\n",
       "            <td>0.18377645</td>\n",
       "            <td>0.0075391266</td>\n",
       "            <td>0.014127848</td>\n",
       "            <td>0.006139751</td>\n",
       "            <td>112032</td>\n",
       "            <td>152038</td>\n",
       "            <td>0.736868</td>\n",
       "            <td>0.008</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>-3.1</td>\n",
       "            <td>0.18001792828241983</td>\n",
       "            <td>0.009268908451000849</td>\n",
       "            <td>0.26313198</td>\n",
       "            <td>COMMON</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>chr4</td>\n",
       "            <td>84240405</td>\n",
       "            <td>rs10000005</td>\n",
       "            <td>G</td>\n",
       "            <td>A</td>\n",
       "            <td>.</td>\n",
       "            <td>PASS</td>\n",
       "            <td>K562__ref=0.11294843;HepG2__ref=0.002546789;SKNSH__ref=-0.14828461;K562__alt=0.2211035;HepG2__alt=0.007651248;SKNSH__alt=-0.036082108;K562__skew=0.10815508;HepG2__skew=0.005104457;SKNSH__skew=0.11220249;AN_oth=2092;AN_ami=906;AN_sas=4822;AN_fin=10542;AN_eas=5182;AN_amr=15266;AN_afr=41314;AN_mid=316;AN_asj=3472;AN_nfe=67982;cadd_raw_score=-0.347702;cadd_phred=0.247;vep=A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|AC079160.1|ENSG00000250546|Transcript|ENST00000508406|lncRNA||4/7|ENST00000508406.1:n.579+2706C&gt;T|||||||1||-1|SNV||Clone_based_ensembl_gene||YES|5|||||||||||||||||||,A|upstream_gene_variant|MODIFIER|AC025538.1|ENSG00000214980|Transcript|ENST00000510707|processed_pseudogene||||||||||1|3598|1|SNV||Clone_based_ensembl_gene||YES||||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|AC079160.1|ENSG00000250546|Transcript|ENST00000513489|lncRNA||2/5|ENST00000513489.5:n.228+2706C&gt;T|||||||1||-1|SNV||Clone_based_ensembl_gene|||1|||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|AC079160.1|ENSG00000250546|Transcript|ENST00000657959|lncRNA||2/5|ENST00000657959.1:n.155+2706C&gt;T|||||||1||-1|SNV||Clone_based_ensembl_gene||||||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|AC079160.1|ENSG00000250546|Transcript|ENST00000668493|lncRNA||5/6|ENST00000668493.1:n.642+2706C&gt;T|||||||1||-1|SNV||Clone_based_ensembl_gene||||||||||||||||||||||,A|intron_variant&amp;non_coding_transcript_variant|MODIFIER|LOC101928978|101928978|Transcript|NR_125909.1|lncRNA||5/8|NR_125909.1:n.622+2706C&gt;T|||||||1||-1|SNV||EntrezGene||YES||||||||||||||||||||;AC=86646;AN=151894;AF=0.570437;AC_oth=1258;AF_oth=0.601338;AC_ami=372;AF_ami=0.410596;AC_sas=3890;AF_sas=0.806719;AC_fin=5069;AF_fin=0.480839;AC_eas=4041;AF_eas=0.779815;AC_amr=10022;AF_amr=0.656492;AC_afr=22346;AF_afr=0.540882;AC_mid=224;AF_mid=0.708861;AC_asj=2151;AF_asj=0.619528;AC_nfe=37273;AF_nfe=0.548277;AC_popmax=3890;AN_popmax=4822;AF_popmax=0.806719</td>\n",
       "            <td>0.11294843</td>\n",
       "            <td>0.002546789</td>\n",
       "            <td>-0.14828461</td>\n",
       "            <td>0.2211035</td>\n",
       "            <td>0.007651248</td>\n",
       "            <td>-0.036082108</td>\n",
       "            <td>0.10815508</td>\n",
       "            <td>0.005104457</td>\n",
       "            <td>0.11220249</td>\n",
       "            <td>86646</td>\n",
       "            <td>151894</td>\n",
       "            <td>0.570437</td>\n",
       "            <td>0.247</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>False</td>\n",
       "            <td>-1.001</td>\n",
       "            <td>-0.010929798086484274</td>\n",
       "            <td>0.07515400648117065</td>\n",
       "            <td>0.429563</td>\n",
       "            <td>COMMON</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('chr7', 92754574, 'rs10', 'A', 'C', '.', 'PASS', 'K562__ref=0.23715062;HepG2__ref=0.3610611;SKNSH__ref=0.18303184;K562__alt=0.2820961;HepG2__alt=0.40103868;SKNSH__alt=0.1955727;K562__skew=0.044945486 ... (1447 characters truncated) ... fr=41052;AF_afr=0.99035;AC_mid=316;AF_mid=1;AC_asj=3414;AF_asj=0.983295;AC_nfe=63809;AF_nfe=0.937733;AC_popmax=5197;AN_popmax=5198;AF_popmax=0.999808', 0.23715062, 0.3610611, 0.18303184, 0.2820961, 0.40103868, 0.1955727, 0.044945486, 0.039977532, 0.012540861, 146019, 152188, 0.959465, 0.266, False, False, False, False, False, False, False, False, -0.715, 0.26041452089945477, 0.032487958669662476, 0.040534973, 'COMMON'),\n",
       " ('chr12', 126406434, 'rs1000000', 'G', 'A', '.', 'PASS', 'K562__ref=0.14421938;HepG2__ref=0.21698855;SKNSH__ref=0.22241707;K562__alt=0.14721091;HepG2__alt=0.2585578;SKNSH__alt=0.22416125;K562__skew=0.0029915 ... (1192 characters truncated) ... 51;AF_afr=0.126965;AC_mid=64;AF_mid=0.202532;AC_asj=663;AF_asj=0.191176;AC_nfe=14886;AF_nfe=0.218886;AC_popmax=1144;AN_popmax=4828;AF_popmax=0.236951', 0.14421938, 0.21698855, 0.22241707, 0.14721091, 0.2585578, 0.22416125, 0.0029915501, 0.041569263, 0.0017441983, 28089, 151978, 0.184823, 1.165, False, False, False, False, False, False, False, False, 0.656, 0.19454167286554971, 0.015435003985961279, 0.184823, 'COMMON'),\n",
       " ('chr4', 40087276, 'rs10000000', 'A', 'T', '.', 'PASS', 'K562__ref=0.36345437;HepG2__ref=0.26544645;SKNSH__ref=0.49515116;K562__alt=0.353244;HepG2__alt=0.25292698;SKNSH__alt=0.4838382;K562__skew=-0.01021034 ... (3508 characters truncated) ... 5;AF_afr=0.213793;AC_mid=35;AF_mid=0.113636;AC_asj=265;AF_asj=0.0773046;AC_nfe=4160;AF_nfe=0.0619416;AC_popmax=2132;AN_popmax=5106;AF_popmax=0.417548', 0.36345437, 0.26544645, 0.49515116, 0.353244, 0.25292698, 0.4838382, -0.010210349, -0.01251948, -0.0113130035, 18324, 150056, 0.122114, 0.394, False, False, False, False, False, False, False, False, -3.541, 0.37468401590983075, -0.011347610503435135, 0.122114, 'COMMON'),\n",
       " ('chr4', 56695481, 'rs10000003', 'A', 'G', '.', 'PASS', 'K562__ref=0.2162989;HepG2__ref=0.1461182;SKNSH__ref=0.1776367;K562__alt=0.22383805;HepG2__alt=0.16024604;SKNSH__alt=0.18377645;K562__skew=0.007539126 ... (1358 characters truncated) ... ;AF_afr=0.810438;AC_mid=228;AF_mid=0.721519;AC_asj=2475;AF_asj=0.713256;AC_nfe=47719;AF_nfe=0.701647;AC_popmax=4627;AN_popmax=5190;AF_popmax=0.891522', 0.2162989, 0.1461182, 0.1776367, 0.22383805, 0.16024604, 0.18377645, 0.0075391266, 0.014127848, 0.006139751, 112032, 152038, 0.736868, 0.008, False, False, False, False, False, False, False, False, -3.1, 0.18001792828241983, 0.009268908451000849, 0.26313198, 'COMMON'),\n",
       " ('chr4', 84240405, 'rs10000005', 'G', 'A', '.', 'PASS', 'K562__ref=0.11294843;HepG2__ref=0.002546789;SKNSH__ref=-0.14828461;K562__alt=0.2211035;HepG2__alt=0.007651248;SKNSH__alt=-0.036082108;K562__skew=0.10 ... (1684 characters truncated) ... ;AF_afr=0.540882;AC_mid=224;AF_mid=0.708861;AC_asj=2151;AF_asj=0.619528;AC_nfe=37273;AF_nfe=0.548277;AC_popmax=3890;AN_popmax=4822;AF_popmax=0.806719', 0.11294843, 0.002546789, -0.14828461, 0.2211035, 0.007651248, -0.036082108, 0.10815508, 0.005104457, 0.11220249, 86646, 151894, 0.570437, 0.247, False, False, False, False, False, False, False, False, -1.001, -0.010929798086484274, 0.07515400648117065, 0.429563, 'COMMON')]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-04-05 17:42:32.147 EDT [2396618] LOG:  checkpoint starting: time\n",
      "2025-04-05 17:43:56.687 EDT [2396618] LOG:  checkpoint complete: wrote 328 buffers (2.0%); 0 WAL file(s) added, 4 removed, 0 recycled; write=84.348 s, sync=0.155 s, total=84.540 s; sync files=16, longest=0.054 s, average=0.010 s; distance=74331 kB, estimate=74331 kB; lsn=183/41EA34B8, redo lsn=183/40A2E7C8\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM unique_id_malin_gnomad\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f058c5d3-859d-41a3-ba23-27e002f71c6e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-01-30T21:25:19.123486Z",
     "iopub.status.busy": "2025-01-30T21:25:19.123024Z",
     "iopub.status.idle": "2025-01-30T21:25:19.137194Z",
     "shell.execute_reply": "2025-01-30T21:25:19.136869Z",
     "shell.execute_reply.started": "2025-01-30T21:25:19.123460Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * postgresql://mr_root:***@localhost:5433/scratch\n",
      "5 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <thead>\n",
       "        <tr>\n",
       "            <th>tag_snp</th>\n",
       "            <th>pubmedid</th>\n",
       "            <th>study</th>\n",
       "            <th>disease_trait</th>\n",
       "            <th>p_value</th>\n",
       "            <th>ld_buddy</th>\n",
       "            <th>r2</th>\n",
       "        </tr>\n",
       "    </thead>\n",
       "    <tbody>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs1000005</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs1000005</td>\n",
       "            <td>36477530</td>\n",
       "            <td>Genetic diversity fuels gene discovery for tobacco and alcohol use.</td>\n",
       "            <td>Drinks per week</td>\n",
       "            <td>5E-17</td>\n",
       "            <td>rs7275426</td>\n",
       "            <td>0.83</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10000225</td>\n",
       "            <td>tag</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs10009710</td>\n",
       "            <td>0.71</td>\n",
       "        </tr>\n",
       "        <tr>\n",
       "            <td>rs10000225</td>\n",
       "            <td>26634245</td>\n",
       "            <td>A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.</td>\n",
       "            <td>Post bronchodilator FEV1</td>\n",
       "            <td>3E-6</td>\n",
       "            <td>rs12640763</td>\n",
       "            <td>0.74</td>\n",
       "        </tr>\n",
       "    </tbody>\n",
       "</table>"
      ],
      "text/plain": [
       "[('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs1000005', 'tag'),\n",
       " ('rs1000005', 36477530, 'Genetic diversity fuels gene discovery for tobacco and alcohol use.', 'Drinks per week', '5E-17', 'rs7275426', '0.83'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs10000225', 'tag'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs10009710', '0.71'),\n",
       " ('rs10000225', 26634245, 'A genome-wide association study identifies risk loci for spirometric measures among smokers of European and African ancestry.', 'Post bronchodilator FEV1', '3E-6', 'rs12640763', '0.74')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT *\n",
    "FROM gwas_AFR\n",
    "LIMIT 5;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "766fbf1b-0049-4d3d-9d66-d9aade4d75bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"[+] POP AFR\",flush=True)\n",
    "print(f\"[+] {datetime.datetime.now()}\",flush=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "473d92f4-3950-4d37-98da-092f8bf71d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "--DROP TABLE merged_AFR CASCADE;\n",
    "CREATE TABLE merged_AFR AS\n",
    "SELECT *\n",
    "FROM\n",
    "    gwas_AFR\n",
    "INNER JOIN\n",
    "    unique_id_malin_gnomad\n",
    "ON\n",
    "    unique_id_malin_gnomad.id=gwas_AFR.ld_buddy;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30cc1db8-3ee2-4c4a-b82b-5f48910d2738",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"[+] POP EUR\",flush=True)\n",
    "print(f\"[+] {datetime.datetime.now()}\",flush=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f17b057e-ef5b-4571-8c9a-4ede31a53e40",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "--DROP TABLE merged_EUR CASCADE;\n",
    "CREATE TABLE merged_EUR AS\n",
    "SELECT *\n",
    "FROM\n",
    "    gwas_EUR\n",
    "INNER JOIN\n",
    "    unique_id_malin_gnomad\n",
    "ON\n",
    "    unique_id_malin_gnomad.id=gwas_EUR.ld_buddy;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfd830f4-39f7-4dde-8c3f-b1c02517aa01",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"[+] POP ASN\",flush=True)\n",
    "print(f\"[+] {datetime.datetime.now()}\",flush=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29a9c664-2dba-4832-90d1-aa58bd03e952",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%sql\n",
    "--DROP TABLE merged_ASN CASCADE;\n",
    "CREATE TABLE merged_ASN AS\n",
    "SELECT *\n",
    "FROM\n",
    "    gwas_ASN\n",
    "INNER JOIN\n",
    "    unique_id_malin_gnomad\n",
    "ON\n",
    "    unique_id_malin_gnomad.id=gwas_ASN.ld_buddy;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed51dae-a1fb-4a72-9d94-039a2caf8bf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"[+] DONE.\",flush=True)\n",
    "print(f\"[+] {datetime.datetime.now()}\",flush=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0b66b352-e799-44a4-8722-123d52b00eb0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-05T15:41:53.905020Z",
     "iopub.status.busy": "2025-02-05T15:41:53.904856Z",
     "iopub.status.idle": "2025-02-05T15:41:55.651369Z",
     "shell.execute_reply": "2025-02-05T15:41:55.650774Z",
     "shell.execute_reply.started": "2025-02-05T15:41:53.905005Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pg_ctl: PID file \"/home/mcn26/palmer_scratch/db/postmaster.pid\" does not exist\n",
      "Is server running?\n"
     ]
    }
   ],
   "source": [
    "!pg_ctl -D ~/palmer_scratch/db stop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac1328c5-a2db-4cd8-b747-02c14c28f892",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"WAITING 60s FOR SERVER TO SHUT DOWN\",flush=True)\n",
    "time.sleep(60)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
